---
title: "Poster Graphics"
output: html_notebook
---


Load Packages and Data

```{r}
## Load packages
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(cowplot)
library(pracma)
library(RColorBrewer)`
library(EnhancedVolcano)
library(gridExtra)
library(reshape2)

memory.limit(size = 16000000)

## Integrated PDAC datasets 
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 155698\\PDAC155698.integrated")

## BCCIO dataset
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 123813\\BCCIOwithUMAP")

## Merged dataset (5000 cells from PDAC and BCC, random sample)
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.PDAC.merge")

## Merged dataset (5000 cells from PDAC and BCC pretreatment, random sample)
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.pre.PDAC.merge")

## Load functions
source("ClusteringAnalysisFunctions.R", echo=TRUE)
source("SeuratClusteringFunctions.R", echo=TRUE)
source("DifferentialExpressionFunctions.R", echo=TRUE)

```



BCC Clustering UMAP

```{r}

BCCIO <- RenameIdents(object = BCCIO, 
                      'CD8_memory' = 'CD8+ memory T cells',
                      'CD4' = 'CD4+ T cells', 
                      'Tregs' = 'Regulatory T cells',
                      'CD8_activation' = 'CD8+ activated T cells',
                      'CD8_exhausted' = 'CD8+ exhausted T cells',
                      'Plasma' = 'Plasma cells',
                      'Tumor 1' = 'Malignant cells 1',
                      'Tumor 2' = 'Malignant cells 2',
                      'T_proliferation' = 'Proliferating T cells',
                      'NK' = 'NK cells',
                      'Endothelial' = 'Endothelial cells',
                      'DCs' = 'Dendritic cells',
                      'Misc.' = 'Other')

BCC.UMAP <- DimPlot(object = BCCIO, 
                    reduction ="umap", 
                    label = TRUE, 
                    pt.size = 0.3,
                    label.size = 6) + NoLegend()
BCC.UMAP

## Default dimensions, width = 1200px
## Saved as BCC_UMAP.jpeg

```



BCC UMAP separated by patient

```{r}

BCC.UMAP.patient <- DimPlot(object = BCCIO,
                            reduction = "umap",
                            label = FALSE,
                            split.by = c("patient"),
                            pt.size = 0.1,
                            ncol = 6)
BCC.UMAP.patient

## Dimensions: width = 1500px, height = 505px
## Post-processing to remove axes in powerpoint
## Saved as BCC_UMAP_Patient.JPG

```



BCC UMAP separated by responders vs. nonresponders

```{r}

BCC.UMAP.responder <- DimPlot(object = BCCIO,
                              reduction = "umap",
                              label = FALSE,
                              split.by = c("response"),
                              pt.size = 0.1,
                              ncol = 2)

BCC.UMAP.responder

## Dimensions: width = 1000px, height = 505px
## Saved as BCC_UMAP_Responder.jpeg

```



BCC UMAP separated by patient and treatment

```{r}

BCCIO$patient.treatment <- paste0(BCCIO$patient, BCCIO$treatment)

Idents(BCCIO) <- BCCIO$patient.treatment

BCC.patient001.pre.cells <- WhichCells(BCCIO, idents = c("su001pre"))
BCC.patient001.post.cells <- WhichCells(BCCIO, idents = c("su001post"))
BCC.patient002.pre.cells <- WhichCells(BCCIO, idents = c("su002pre"))
BCC.patient002.post.cells <- WhichCells(BCCIO, idents = c("su002post"))
BCC.patient003.pre.cells <- WhichCells(BCCIO, idents = c("su003pre"))
BCC.patient003.post.cells <- WhichCells(BCCIO, idents = c("su003post"))
BCC.patient004.pre.cells <- WhichCells(BCCIO, idents = c("su004pre"))
BCC.patient004.post.cells <- WhichCells(BCCIO, idents = c("su004post"))
BCC.patient005.pre.cells <- WhichCells(BCCIO, idents = c("su005pre"))
BCC.patient005.post.cells <- WhichCells(BCCIO, idents = c("su005post"))
BCC.patient006.pre.cells <- WhichCells(BCCIO, idents = c("su006pre"))
BCC.patient006.post.cells <- WhichCells(BCCIO, idents = c("su006post"))
BCC.patient007.pre.cells <- WhichCells(BCCIO, idents = c("su007pre"))
BCC.patient007.post.cells <- WhichCells(BCCIO, idents = c("su007post"))
BCC.patient008.pre.cells <- WhichCells(BCCIO, idents = c("su008pre"))
BCC.patient008.post.cells <- WhichCells(BCCIO, idents = c("su008post"))
BCC.patient009.pre.cells <- WhichCells(BCCIO, idents = c("su009pre"))
BCC.patient009.post.cells <- WhichCells(BCCIO, idents = c("su009post"))
BCC.patient010.pre.cells <- WhichCells(BCCIO, idents = c("su010pre"))
BCC.patient010.post.cells <- WhichCells(BCCIO, idents = c("su010post"))
BCC.patient012.pre.cells <- WhichCells(BCCIO, idents = c("su012pre"))
BCC.patient012.post.cells <- WhichCells(BCCIO, idents = c("su012post"))

BCC.UMAP.patient1.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient001.pre.cells, BCC.patient001.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su001", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient2.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient002.pre.cells, BCC.patient002.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su002", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient3.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient003.pre.cells, BCC.patient003.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su003", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient4.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient004.pre.cells, BCC.patient004.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su004", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient5.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient005.pre.cells, BCC.patient005.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su005", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient6.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient006.pre.cells, BCC.patient006.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su006", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient7.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient007.pre.cells, BCC.patient007.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su007", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient8.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient008.pre.cells, BCC.patient008.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su008", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient9.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient009.pre.cells, BCC.patient009.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su009", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient10.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient010.pre.cells, BCC.patient010.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su010", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))
BCC.UMAP.patient12.treatment <- DimPlot(object = BCCIO,
                                      reduction = "umap", 
                                      label = FALSE,
                                      group.by = c("patient.treatment"),
                                      cells.highlight = list(BCC.patient012.pre.cells, BCC.patient012.post.cells),
                                      cols.highlight = c("royalblue1", "darkslategray1"),
                                      cols = c("grey90"),
                                      sizes.highlight = 0.1,
                                      pt.size = 0.1) + NoLegend() + 
                                      labs(title = "su012", x = "",  y = "") + 
                                      theme(axis.line.y = element_blank(), axis.line.x = element_blank(),
                                            axis.text.y = element_blank(), axis.text.x = element_blank(),
                                            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
                                            plot.title = element_text(hjust = 0.5))

BCC.UMAP.patient.treatment <- grid.arrange(BCC.UMAP.patient1.treatment, BCC.UMAP.patient2.treatment, BCC.UMAP.patient3.treatment, BCC.UMAP.patient4.treatment, BCC.UMAP.patient5.treatment, BCC.UMAP.patient6.treatment, BCC.UMAP.patient7.treatment, BCC.UMAP.patient8.treatment, BCC.UMAP.patient9.treatment, BCC.UMAP.patient10.treatment, BCC.UMAP.patient12.treatment,  nrow=2)

BCC.UMAP.patient.treatment

## Dimensions: width = 1500px, height = 500px
## Post-processing to add legend in Powerpoint
## Saved as: BCC_UMAP_Patient_Treatment.JPG

```



BCC Clustering Marker Genes

```{r}

BCC.cluster.order <- c('CD4+ T cells', 'CD8+ memory T cells', 'CD8+ activated T cells', 'CD8+ exhausted T cells', 'Proliferating T cells', 'Regulatory T cells', 'NK cells', 'B cells', 'Plasma cells', 'Macrophages', 'Dendritic cells', 'pDCs', 'Myofibroblasts', 'CAFs', 'Malignant cells 1', 'Malignant cells 2', 'Endothelial cells', 'Melanocytes', 'Other')

BCCIO <- RenameIdents(object = BCCIO, 
                      'CD4+ T cells' = 'CD4+ T cells', 
                      'CD8+ memory T cells' = 'CD8+ memory T cells',
                      'CD8+ activated T cells' = 'CD8+ activated T cells',
                      'CD8+ exhausted T cells' = 'CD8+ exhausted T cells',
                      'Proliferating T cells' = 'Proliferating T cells',
                      'Regulatory T cells' = 'Regulatory T cells',
                      'NK cells' = 'NK cells',
                      'B cells' = 'B cells',
                      'Plasma cells' = 'Plasma cells',
                      'Macrophages' = 'Macrophages',
                      'Dendritic cells' = 'Dendritic cells',
                      'pDCs' = 'pDCs', 
                      'Myofibroblasts' = 'Myofibroblasts',
                      'CAFs' = 'CAFs',
                      'Malignant cells 1' = 'Malignant cells 1',
                      'Malignant cells 2' = 'Malignant cells 2',
                      'Endothelial cells' = 'Endothelial cells',
                      'Melanocytes' = 'Melanocytes',
                      'Other' = 'Other')

features <- c('CD3D', 'CD3E', 'CD2', 'CD8A', 'GZMA', 'GZMB', 'CD4', 'FOXP3', 'KLRC1', 'KLRC3', 'NKG7', 'CD19', 'CD79A', 'SLAMF7', 'IGKC', 'FCGR2A', 'CSF1R', 'FLT3', 'CLEC4C', 'COL1A2', 'MCAM', 'MYLK', 'FAP', 'PDPN', 'EPCAM', 'TP63', 'PECAM1', 'VWF', 'PMEL', 'MLANA')

BCC.cluster.markers <- DotPlot(BCCIO, features = features, assay = 'RNA', cols = c('Azure', 'Blue')) + RotatedAxis() + xlab('Marker Genes')+ ylab('Clusters')

BCC.cluster.markers

## Dimensions: width = 1500px, height = 750px
## Saved as BCC_Cluster_Markers.jpeg

```



PDAC Clustering Marker Genes

```{r}

Idents(PDAC155698.integrated) <- PDAC155698.integrated$seurat_clusters

PDAC155698.integrated <- RenameIdents(object = PDAC155698.integrated,
                                      "Endothelial" = "Endothelial cells",
                                      "Tregs" = "Regulatory T cells", 
                                      "Epithelial + Endocrine" = "Epithelial cells")

features <- c('LAMP3', 'CCL22', 'TFF1', 'KRT18', 'KRT19', 'KRT8', 'CLU', 'MMP7', 'SPP1', 'REG1A', 'CTRB2', 'PRSS1', 'DCN', 'LUM', 'CPA3', 'TPSAB1', 'CDH5', 'VWF', 'PLVAP', 'IRF7', 'RGS5', 'PDGFRB', 'CD3E', 'NCAM1', 'NKG7', 'CD3D', 'CD14', 'HLA-DRA', 'GZMB', 'ITGAX', 'ITGAM', 'APOE', 'LYZ', 'IGJ', 'CD79A', 'MS4A1')

PDAC.cluster.markers <- DotPlot(PDAC155698.integrated, features = features, assay = 'RNA', cols = c('Azure', 'Blue')) + RotatedAxis() + xlab('Marker Genes')+ ylab('Clusters')

PDAC.cluster.markers

```



BCC Clustering Comparison Heatmap

```{r}

load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 123813\\BCCIOrawdata")

BCCIO.meta.data1 <- BCCIO.meta.data[order(BCCIO.meta.data[, "sort"], -xtfrm(BCCIO.meta.data[, "treatment"])),]
BCCIO.meta.data2 <- BCCIO@meta.data[order(BCCIO@meta.data[, "sort"]),]

BCCIO.meta.data1["rownames"] <- rownames(BCCIO.meta.data1)
BCCIO.meta.data2["rownames"] <- rownames(BCCIO.meta.data2)

BCCIO.meta.data.merge <- merge(BCCIO.meta.data1, BCCIO.meta.data2, by.x=c("rownames"), by.y=c("rownames"))
BCCIO.clustering.comparison <- BCCIO.meta.data.merge[c("cluster", "seurat_clusters")]

uniqueclusters1 <- as.matrix(unique(BCCIO.clustering.comparison[,1]))
uniqueclusters2 <- as.matrix(unique(BCCIO.clustering.comparison[,2]))

BCCIO.distribution.table <- dist_table(BCCIO.clustering.comparison, uniqueclusters1, uniqueclusters2, 1, 2)

BCCIO.proportion.table <- prop_table(BCCIO.distribution.table, uniqueclusters1, uniqueclusters2)

clusterlabels1 <- c('CD8+ memory T cells', 'CD8+ exhausted T cells', 'Regulatory T cells', 'CD4+ T cells', 'Proliferating T cells', 'NK cells', 'B cells 1', 'CD8+ activated T cells', 'Macrophages', 'pDCs', 'Plasma cells', 'Dendritic cells', 'CAFs', 'Endothelial cells', 'Myofibroblasts', 'B cells 2', 'Melanocytes', 'Malignant cells 1', 'Malignant cells 2')

clusterlabels2 <- c('CD8+ memory T cells', 'CD8+ exhausted T cells', 'Regulatory T cells', 'CD4+ T cells', 'Proliferating T cells', 'NK cells', 'Other', 'CD8+ activated T cells', 'B cells', 'Macrophages', 'pDCs', 'Plasma cells', 'Dendritic cells', 'CAFs', 'Endothelial cells', 'Myofibroblasts', 'Melanocytes', 'Malignant cells 1', 'Malignant cells 2')

BCCIO.heatmap <- ggplot_heatmap(BCCIO.proportion.table, clusterlabels1, clusterlabels2, c("Metadata Clustering"), c("Unsupervised Clustering")) + 
  theme(axis.text = element_text(size=10), axis.title = element_text(size=14)) + 
  scale_fill_viridis(discrete=FALSE, name = "Proportion of \nsimilar cells")

BCCIO.heatmap

## Dimensions: width = 1000px (scale up)
## Saved as BCC_Clustering_Heatmap.jpeg

```



BCC Pretreatment vs. Posttreatment Barcharts 

```{r}

## Separate into pretreatment and posttreatment batches

Idents(BCCIO) <- BCCIO$treatment
BCCIO.pre <- subset(BCCIO, idents=c("pre"))
BCCIO.post <- subset(BCCIO, idents=c("post"))

## Pre-treatment barchart

dataframe <- data.frame("Batch" = BCCIO.pre@meta.data$patient, "Clusters" = BCCIO.pre@meta.data$seurat_clusters)
tibble <- as_tibble(dataframe)
  
summary <- tibble %>% group_by(Clusters, Batch) %>% count()

summary$cluster <- as.vector(summary$Clusters)

summary[summary == "CD8_memory"] <- "CD8+ memory T cells"
summary[summary == "CD4"] <- "CD4+ T cells"
summary[summary == "Tregs"] <- "Regulatory T cells"
summary[summary == "CD8_activation"] <- "CD8+ activated T cells"
summary[summary == "CD8_exhausted"] <- "CD8+ exhausted T cells"
summary[summary == "Plasma"] <- "Plasma cells"
summary[summary == "Tumor 1"] <- "Malignant cells 1"
summary[summary == "Tumor 2"] <- "Malignant cells 2"
summary[summary == "T_proliferation"] <- "Proliferating T cells"
summary[summary == "NK"] <- "NK cells"
summary[summary == "Endothelial"] <- "Endothelial cells"
summary[summary == "DCs"] <- "Dendritic cells"
summary[summary == "Misc."] <- "Other"

summary$Clusters <- as.factor(summary$cluster)
summary <- subset(summary, select = -cluster)

normalized_dataset <- summary %>% group_by(Batch) %>% mutate(normalize = n/sum(n))

normalized_dataset$Clusters <- factor(normalized_dataset$Clusters,
                                      levels = c("CD4+ T cells", "CD8+ memory T cells", "CD8+ activated T cells", "CD8+ exhausted T cells", "Regulatory T cells", "Proliferating T cells", "NK cells", "B cells", "Macrophages", "Dendritic cells", "pDCs", "Plasma cells", "CAFs", "Endothelial cells", "Myofibroblasts", "Melanocytes", "Malignant cells 1", "Malignant cells 2", "Other"))

BCC.pre.patient.barchart <- ggplot(normalized_dataset, aes(x=Clusters, y=normalize, fill=Batch)) + 
      geom_col(position = 'fill') + 
      scale_fill_viridis(discrete=TRUE) +
      theme_classic() +
      theme(axis.text.x = element_text(angle=45, hjust=1)) + 
      ylab("Cell Proportion Normalized by Batch") + 
      ggtitle("Pre-treatment") +
      theme(plot.title = element_text(hjust = 0.5))

## Post-treatment barchart

dataframe <- data.frame("Batch" = BCCIO.post@meta.data$patient, "Clusters" = BCCIO.post@meta.data$seurat_clusters)
tibble <- as_tibble(dataframe)
  
summary <- tibble %>% group_by(Clusters, Batch) %>% count()

summary$cluster <- as.vector(summary$Clusters)

summary[summary == "CD8_memory"] <- "CD8+ memory T cells"
summary[summary == "CD4"] <- "CD4+ T cells"
summary[summary == "Tregs"] <- "Regulatory T cells"
summary[summary == "CD8_activation"] <- "CD8+ activated T cells"
summary[summary == "CD8_exhausted"] <- "CD8+ exhausted T cells"
summary[summary == "Plasma"] <- "Plasma cells"
summary[summary == "Tumor 1"] <- "Malignant cells 1"
summary[summary == "Tumor 2"] <- "Malignant cells 2"
summary[summary == "T_proliferation"] <- "Proliferating T cells"
summary[summary == "NK"] <- "NK cells"
summary[summary == "Endothelial"] <- "Endothelial cells"
summary[summary == "DCs"] <- "Dendritic cells"
summary[summary == "Misc."] <- "Other"

summary$Clusters <- as.factor(summary$cluster)
summary <- subset(summary, select = -cluster)

normalized_dataset <- summary %>% group_by(Batch) %>% mutate(normalize = n/sum(n))

normalized_dataset$Clusters <- factor(normalized_dataset$Clusters,
                                      levels = c("CD4+ T cells", "CD8+ memory T cells", "CD8+ activated T cells", "CD8+ exhausted T cells", "Regulatory T cells", "Proliferating T cells", "NK cells", "B cells", "Macrophages", "Dendritic cells", "pDCs", "Plasma cells", "CAFs", "Endothelial cells", "Myofibroblasts", "Melanocytes", "Malignant cells 1", "Malignant cells 2", "Other"))

BCC.post.patient.barchart <- ggplot(normalized_dataset, aes(x=Clusters, y=normalize, fill=Batch)) + 
      geom_col(position = 'fill') + 
      theme_classic() +
      theme(axis.text.x = element_text(angle=45, hjust=1)) + 
      ylab("Cell Proportion Normalized by Patient") + 
      scale_fill_viridis(discrete=TRUE, name = "Patient") + 
      ggtitle("Post-treatment") +
      theme(plot.title = element_text(hjust = 0.5))


BCC.pre.patient.barchart + NoLegend() +
  BCC.post.patient.barchart

## Dimensions: width= 1000px, height = 500px
## Saved as BCC_Barcharts_Treatment.jpeg

```



BCC Responder vs. Non-responder Barcharts

```{r}

## Separate into pretreatment and posttreatment batches

Idents(BCCIO) <- BCCIO$patient
BCCIO.responders <- subset(BCCIO, idents=c("su001", "su002", "su003", "su004", "su009", "su012"))
BCCIO.nonresponders <- subset(BCCIO, idents=c("su005", "su006", "su007", "su008", "su010"))

## Responders barchart

dataframe <- data.frame("Batch" = BCCIO.responders@meta.data$treatment, "Clusters" = BCCIO.responders@meta.data$seurat_clusters)
tibble <- as_tibble(dataframe)
  
summary <- tibble %>% group_by(Clusters, Batch) %>% count()

summary$cluster <- as.vector(summary$Clusters)
summary$batches <- as.vector(summary$Batch)

summary[summary == "CD8_memory"] <- "CD8+ memory T cells"
summary[summary == "CD4"] <- "CD4+ T cells"
summary[summary == "Tregs"] <- "Regulatory T cells"
summary[summary == "CD8_activation"] <- "CD8+ activated T cells"
summary[summary == "CD8_exhausted"] <- "CD8+ exhausted T cells"
summary[summary == "Plasma"] <- "Plasma cells"
summary[summary == "Tumor 1"] <- "Malignant cells 1"
summary[summary == "Tumor 2"] <- "Malignant cells 2"
summary[summary == "T_proliferation"] <- "Proliferating T cells"
summary[summary == "NK"] <- "NK cells"
summary[summary == "Endothelial"] <- "Endothelial cells"
summary[summary == "DCs"] <- "Dendritic cells"
summary[summary == "Misc."] <- "Other"
summary[summary == "post"] <- "Post-treatment"
summary[summary == "pre"] <- "Pre-treatment"

summary$Clusters <- as.factor(summary$cluster)
summary <- subset(summary, select = -cluster)
summary$Batch <- as.factor(summary$batches)
summary <- subset(summary, select = -batches)

normalized_dataset <- summary %>% group_by(Batch) %>% mutate(normalize = n/sum(n))

normalized_dataset$Clusters <- factor(normalized_dataset$Clusters,
                                      levels = c("CD4+ T cells", "CD8+ memory T cells", "CD8+ activated T cells", "CD8+ exhausted T cells", "Regulatory T cells", "Proliferating T cells", "NK cells", "B cells", "Macrophages", "Dendritic cells", "pDCs", "Plasma cells", "CAFs", "Endothelial cells", "Myofibroblasts", "Melanocytes", "Malignant cells 1", "Malignant cells 2", "Other"))

normalized_dataset$Batch <- factor(normalized_dataset$Batch, 
                                   levels = c("Pre-treatment", "Post-treatment"))

BCC.responders.barchart <- ggplot(normalized_dataset, aes(x=Clusters, y=normalize, fill=Batch)) + 
      geom_col(position = 'fill') + 
      scale_fill_manual(values = c("royalblue3", "lightskyblue1")) +
      theme_classic() +
      theme(axis.text.x = element_text(angle=45, hjust=1)) + 
      ylab("Cell Proportion Normalized by Batch") + 
      ggtitle("Responders") +
      theme(plot.title = element_text(hjust = 0.5))

## Nonresponders barchart

dataframe <- data.frame("Batch" = BCCIO.nonresponders@meta.data$treatment, "Clusters" = BCCIO.nonresponders@meta.data$seurat_clusters)
tibble <- as_tibble(dataframe)
  
summary <- tibble %>% group_by(Clusters, Batch) %>% count()

summary$cluster <- as.vector(summary$Clusters)
summary$batches <- as.vector(summary$Batch)

summary[summary == "CD8_memory"] <- "CD8+ memory T cells"
summary[summary == "CD4"] <- "CD4+ T cells"
summary[summary == "Tregs"] <- "Regulatory T cells"
summary[summary == "CD8_activation"] <- "CD8+ activated T cells"
summary[summary == "CD8_exhausted"] <- "CD8+ exhausted T cells"
summary[summary == "Plasma"] <- "Plasma cells"
summary[summary == "Tumor 1"] <- "Malignant cells 1"
summary[summary == "Tumor 2"] <- "Malignant cells 2"
summary[summary == "T_proliferation"] <- "Proliferating T cells"
summary[summary == "NK"] <- "NK cells"
summary[summary == "Endothelial"] <- "Endothelial cells"
summary[summary == "DCs"] <- "Dendritic cells"
summary[summary == "Misc."] <- "Other"
summary[summary == "post"] <- "Post-treatment"
summary[summary == "pre"] <- "Pre-treatment"

summary$Clusters <- as.factor(summary$cluster)
summary <- subset(summary, select = -cluster)
summary$Batch <- as.factor(summary$batches)
summary <- subset(summary, select = -batches)

normalized_dataset <- summary %>% group_by(Batch) %>% mutate(normalize = n/sum(n))

normalized_dataset$Clusters <- factor(normalized_dataset$Clusters,
                                      levels = c("CD4+ T cells", "CD8+ memory T cells", "CD8+ activated T cells", "CD8+ exhausted T cells", "Regulatory T cells", "Proliferating T cells", "NK cells", "B cells", "Macrophages", "Dendritic cells", "pDCs", "Plasma cells", "CAFs", "Endothelial cells", "Myofibroblasts", "Melanocytes", "Malignant cells 1", "Malignant cells 2", "Other"))

normalized_dataset$Batch <- factor(normalized_dataset$Batch, 
                                   levels = c("Pre-treatment", "Post-treatment"))

BCC.nonresponders.barchart <- ggplot(normalized_dataset, aes(x=Clusters, y=normalize, fill=Batch)) + 
      geom_col(position = 'fill') + 
      scale_fill_manual(values = c("tomato4", "salmon1")) +
      theme_classic() +
      theme(axis.text.x = element_text(angle=45, hjust=1)) + 
      ylab("Cell Proportion Normalized by Batch") + 
      ggtitle("Nonresponders") +
      theme(plot.title = element_text(hjust = 0.5))


BCC.responders.barchart + BCC.nonresponders.barchart

## Dimensions: width= 1200px, height = 500px
## Post-processing to modify legend in Powerpoint
## Saved as BCC_Barcharts_Responders.jpeg

```



BCC Macrophage Pretreatment Enhanced Volcano Plot

```{r}

Idents(BCCIO.pre) <- BCCIO.pre$seurat_clusters

macrophage.markers <- FindMarkers(BCCIO.pre, subset.ident='Macrophages', group.by='response', ident.1='Responsive')
macrophage.markers_tibble <- as_tibble(macrophage.markers, rownames = "GeneNames")
macrophage.markers_tibble <- arrange(macrophage.markers_tibble, avg_logFC)

macrophage.volcano <- EnhancedVolcano(macrophage.markers_tibble,
                               lab = macrophage.markers_tibble$GeneNames,
                               x = 'avg_logFC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'Differential Expression of Macrophages: Responders vs. Nonresponders ',
                               subtitle = 'Enhanced Volcano Plot',
                               col = c('black', 'black', 'black', 'red3'), colAlpha = 1)
macrophage.volcano

## Dimensions: width = 1000px, height = 750px
## Saved as BCC_macrophage_volcano.jpeg

```



BCC Macrophage Responder Pretreatment vs. PDAC 

```{r}

Idents(bcc.pre.responders.pdac.merge) <- bcc.pre.responders.pdac.merge$seurat_clusters

macrophage.comparison.markers <- FindMarkers(bcc.pre.responders.pdac.merge, subset.ident='Macrophages', group.by='response', ident.1='Responsive')
macrophage.comparison.markers_tibble <- as_tibble(macrophage.comparison.markers, rownames = "GeneNames")
macrophage.comparison.markers_tibble <- arrange(macrophage.comparison.markers_tibble, avg_logFC)

macrophage.comparison.markers_tibble$p_val_adj[macrophage.comparison.markers_tibble$p_val_adj < 1.00e-50] <- 1.00e-50
macrophage.comparison.markers_tibble$avg_logFC[macrophage.comparison.markers_tibble$avg_logFC < -2] <- -2
macrophage.comparison.markers_tibble$avg_logFC[macrophage.comparison.markers_tibble$avg_logFC > 2] <- 2

macrophage.comparison.volcano <- EnhancedVolcano(macrophage.comparison.markers_tibble,
                               lab = macrophage.comparison.markers_tibble$GeneNames,
                               x = 'avg_logFC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'Differential Expression of Macrophages: BCC Responders vs. PDAC',
                               subtitle = 'Enhanced Volcano Plot',
                               col = c('black', 'black', 'black', 'red3'), colAlpha = 1)
macrophage.comparison.volcano

## Dimensions: width = 1000px, height = 750px
## Saved as BCC_PDAC_macrophage_volcano.jpeg

```



PDAC Clustering UMAP

```{r}

PDAC155698.integrated <- RenameIdents(object = PDAC155698.integrated, 
                      'Epithelial + Endocrine' = 'Epithelial and Endocrine cells',
                      'Endothelial' = 'Endothelial cells',
                      'Dendritic' = 'Dendritic cells',
                      'Tregs' = 'Regulatory T cells')

pdac.umap <- DimPlot(object = PDAC155698.integrated, 
                    reduction ="umap", 
                    label = TRUE, 
                    pt.size = 0.3,
                    label.size = 6) + NoLegend()
pdac.umap

## Dimensions: width= 1200px, height = 900px
## Saved as PDAC_UMAP.jpeg

```



PDAC UMAPs by Patiet

```{r}

PDAC.UMAP.patient <- DimPlot(object = PDAC155698.integrated,
                            reduction = "umap",
                               label = FALSE,
                            split.by = c("orig.ident"),
                            pt.size = 0.1,
                            ncol = 8) + NoAxes()
PDAC.UMAP.patient

## Dimensions: width = 1500px, height = 500px
## Saved as PDAC_UMAP_patient.jpeg

```



PDAC Barchart by Patient 

```{r}

dataframe <- data.frame("Batch" = PDAC155698.integrated@meta.data$orig.ident, "Clusters" = PDAC155698.integrated@meta.data$seurat_clusters)
tibble <- as_tibble(dataframe)
  
summary <- tibble %>% group_by(Clusters, Batch) %>% count()

summary$cluster <- as.vector(summary$Clusters)

summary$cluster[summary$cluster == 'Epithelial + Endocrine'] <- 'Epithelial and Endocrine cells'
summary$cluster[summary$cluster == 'Endothelial'] <- 'Endothelial cells'
summary$cluster[summary$cluster == 'Dendritic'] <- 'Dendritic cells'
summary$cluster[summary$cluster == 'Tregs'] <- 'Regulatory T cells'

summary$Clusters <- as.factor(summary$cluster)
summary <- subset(summary, select = -cluster)

normalized_dataset <- summary %>% group_by(Batch) %>% mutate(normalize = n/sum(n))

normalized_dataset$Clusters <- factor(normalized_dataset$Clusters,
                                      levels = c("CD4+ T cells", "CD8+ T cells", "Regulatory T cells", "NK cells", "B cells", "Macrophages", "Dendritic cells", "Plasma cells", "Endothelial cells", "Epithelial and Endocrine cells", "Fibroblasts", "Mast cells", "Pericytes", "Granulocytes", "Acinar cells"))

pdac.patient.barchart <- ggplot(normalized_dataset, aes(x=Clusters, y=normalize, fill=Batch)) + 
      geom_col(position = 'fill') + 
      theme_classic() +
      theme(axis.text.x = element_text(angle=45, hjust=1)) + 
      ylab("Cell Proportion Normalized by Patient") + 
      scale_fill_viridis(discrete=TRUE, name = "Patient") + 
      ggtitle("Integrated PDAC Dataset") +
      theme(plot.title = element_text(hjust = 0.5, size=16), 
            axis.title = element_text(size=13), 
            axis.text = element_text(size=11))

pdac.patient.barchart

## Dimensions: width = 1000px, height = 750px
## Saved as PDAC_Barchart.jpeg

```



ANOVA Boxplots for Gene Scores on Macrophages

```{r}

## HSP Score

hsp.averages.score.graphics <- hsp.averages.score

colnames(hsp.averages.score.graphics) <- c("HSP Score", "Patients")

hsp.averages.score.graphics$Patients[hsp.averages.score.graphics$Patients == "R"] <- "BCC Responders"
hsp.averages.score.graphics$Patients[hsp.averages.score.graphics$Patients == "NR"] <- "BCC Nonresponders"


comparisons <- list(c("BCC Responders", "BCC Nonresponders"), c("BCC Responders", "PDAC"), c("BCC Nonresponders", "PDAC"))

HSP.boxplot <- ggboxplot(hsp.averages.score.graphics, x = "Patients", y = "HSP Score", color = "Patients", palette = c("blue", "red", "purple"), add = "jitter") +
  stat_compare_means(comparisons = comparisons)

HSP.boxplot <- ggpar(p = HSP.boxplot, legend = "none", main = "HSP Genes", font.main = c(20, "bold", "black"), font.caption = c(9, "italic", "black"), font.x = "bold", font.y = "bold", font.tickslab = 10) +
  theme(plot.title = element_text(hjust = 0.5), plot.caption = element_text(hjust = 0.5))


## MHC Scores

mhc1.averages.score.graphics <- hla.averages.score[, c(1,3)]
mhc2.averages.score.graphics <- hla.averages.score[, c(2,3)]

colnames(mhc1.averages.score.graphics) <- c("MHC Class I Score", "Patients")
colnames(mhc2.averages.score.graphics) <- c("MHC Class II Score", "Patients")

mhc1.averages.score.graphics$Patients <- as.character(mhc1.averages.score.graphics$Patients)

mhc1.averages.score.graphics$Patients[mhc1.averages.score.graphics$Patients == "R"] <- "BCC Responders"
mhc1.averages.score.graphics$Patients[mhc1.averages.score.graphics$Patients == "NR"] <- "BCC Nonresponders"

MHC1.boxplot <- ggboxplot(mhc1.averages.score.graphics, x = "Patients", y = "MHC Class I Score", color = "Patients", palette = c("blue", "red", "purple"), add = "jitter") +
  stat_compare_means(comparisons = comparisons)

MHC1.boxplot <- ggpar(p = MHC1.boxplot, legend = "none", main = "MHC Class I Genes", font.main = c(20, "bold", "black"), font.caption = c(9, "italic", "black"), font.x = "bold", font.y = "bold", font.tickslab = 10) +
  theme(plot.title = element_text(hjust = 0.5), plot.caption = element_text(hjust = 0.5))
 

mhc2.averages.score.graphics$Patients <- as.character(mhc2.averages.score.graphics$Patients)

mhc2.averages.score.graphics$Patients[mhc2.averages.score.graphics$Patients == "R"] <- "BCC Responders"
mhc2.averages.score.graphics$Patients[mhc2.averages.score.graphics$Patients == "NR"] <- "BCC Nonresponders"

MHC2.boxplot <- ggboxplot(mhc2.averages.score.graphics, x = "Patients", y = "MHC Class II Score", color = "Patients", palette = c("blue", "red", "purple"), add = "jitter") +
  stat_compare_means(comparisons = comparisons)

MHC2.boxplot <- ggpar(p = MHC2.boxplot, legend = "none", main = "MHC Class II Genes", font.main = c(20, "bold", "black"), font.caption = c(9, "italic", "black"), font.x = "bold", font.y = "bold", font.tickslab = 10) +
  theme(plot.title = element_text(hjust = 0.5), plot.caption = element_text(hjust = 0.5))
 

## Combine all three boxplots

grid.arrange(MHC1.boxplot, MHC2.boxplot, HSP.boxplot, ncol=3, top = textGrob("Macrophages", gp = gpar(fontsize=32, font=2)), padding = unit(3, "line"))


## Dimensions: Width = 1800px, height = 600px
## Saved as Gene_Expression_Boxplots.jpeg

```



ANOVA Boxplots for Gene Scores on Malignant Cells

```{r}

## HSP Score

hsp.malignant.averages.score.graphics <- hsp.malignant.averages.score

colnames(hsp.malignant.averages.score.graphics) <- c("HSP Score", "Patients")

hsp.malignant.averages.score.graphics$Patients[hsp.malignant.averages.score.graphics$Patients == "R"] <- "BCC Responders"
hsp.malignant.averages.score.graphics$Patients[hsp.malignant.averages.score.graphics$Patients == "NR"] <- "BCC Nonresponders"


comparisons <- list(c("BCC Responders", "BCC Nonresponders"), c("BCC Responders", "PDAC"), c("BCC Nonresponders", "PDAC"))

HSP.malignant.boxplot <- ggboxplot(hsp.malignant.averages.score.graphics, x = "Patients", y = "HSP Score", color = "Patients", palette = c("blue", "red", "purple"), add = "jitter") +
  stat_compare_means(comparisons = comparisons)

HSP.malignant.boxplot <- ggpar(p = HSP.malignant.boxplot, legend = "none", main = "HSP Genes", font.main = c(20, "bold", "black"), font.caption = c(9, "italic", "black"), font.x = "bold", font.y = "bold", font.tickslab = 10) +
  theme(plot.title = element_text(hjust = 0.5), plot.caption = element_text(hjust = 0.5))


## MHC Scores

mhc1.malignant.averages.score.graphics <- hla.malignant.averages.score[, c(1,3)]
mhc2.malignant.averages.score.graphics <- hla.malignant.averages.score[, c(2,3)]

colnames(mhc1.malignant.averages.score.graphics) <- c("MHC Class I Score", "Patients")
colnames(mhc2.malignant.averages.score.graphics) <- c("MHC Class II Score", "Patients")

mhc1.malignant.averages.score.graphics$Patients <- as.character(mhc1.malignant.averages.score.graphics$Patients)

mhc1.malignant.averages.score.graphics$Patients[mhc1.malignant.averages.score.graphics$Patients == "R"] <- "BCC Responders"
mhc1.malignant.averages.score.graphics$Patients[mhc1.malignant.averages.score.graphics$Patients == "NR"] <- "BCC Nonresponders"

MHC1.malignant.boxplot <- ggboxplot(mhc1.malignant.averages.score.graphics, x = "Patients", y = "MHC Class I Score", color = "Patients", palette = c("blue", "red", "purple"), add = "jitter") +
  stat_compare_means(comparisons = comparisons)

MHC1.malignant.boxplot <- ggpar(p = MHC1.malignant.boxplot, legend = "none", main = "MHC Class I Genes", font.main = c(20, "bold", "black"), font.caption = c(9, "italic", "black"), font.x = "bold", font.y = "bold", font.tickslab = 10) +
  theme(plot.title = element_text(hjust = 0.5), plot.caption = element_text(hjust = 0.5))
 

mhc2.malignant.averages.score.graphics$Patients <- as.character(mhc2.malignant.averages.score.graphics$Patients)

mhc2.malignant.averages.score.graphics$Patients[mhc2.malignant.averages.score.graphics$Patients == "R"] <- "BCC Responders"
mhc2.malignant.averages.score.graphics$Patients[mhc2.malignant.averages.score.graphics$Patients == "NR"] <- "BCC Nonresponders"

MHC2.malignant.boxplot <- ggboxplot(mhc2.malignant.averages.score.graphics, x = "Patients", y = "MHC Class II Score", color = "Patients", palette = c("blue", "red", "purple"), add = "jitter") +
  stat_compare_means(comparisons = comparisons)

MHC2.malignant.boxplot <- ggpar(p = MHC2.malignant.boxplot, legend = "none", main = "MHC Class II Genes", font.main = c(20, "bold", "black"), font.caption = c(9, "italic", "black"), font.x = "bold", font.y = "bold", font.tickslab = 10) +
  theme(plot.title = element_text(hjust = 0.5), plot.caption = element_text(hjust = 0.5))
 

## Combine all three boxplots

grid.arrange(MHC1.malignant.boxplot, MHC2.malignant.boxplot, HSP.malignant.boxplot, ncol=3, top = textGrob("Malignant Cells", gp = gpar(fontsize=32, font=2)), padding = unit(3, "line"))

## Dimensions: Width = 1800px, height = 600px
## Saved as Gene_Expression_Boxplots_Malignant.jpeg

   
# ggplot(df, aes(x = Gene Expression, y = Cell Type, fill = Cancer) + geom_boxplot()


```





HLA Markers Dot Plot

```{r}

hla.markers.de <- hla.markers.by.patient

hla.markers.de$`p-value` <- rep(NA, nrow(hla.markers.de)) 
hla.markers.de$`p-value` <- cut(hla.markers.de$p_val_adj, c(0, 0.05, Inf), labels = c("<0.05", ">0.05"), right = FALSE)

hla.markers.de <- hla.markers.de[, c("GeneNames", "avg_log2FC", "p-value", "patient")]

color <- factor(hla.markers.de$`p-value`, labels = c("black", "grey"))

hla.markers.boxplot <- ggplot(hla.markers.de, aes(x = GeneNames, y = avg_log2FC), fill = "red") +
  geom_boxplot(alpha= 0.8) +
  geom_point(color = color, size = 3) + 
  theme(legend.position = "right", axis.text.x = element_text(angle = 45)) +
  theme_bw()
    
hla.markers.boxplot

hla.markers.boxplot.patient <- ggplot(hla.markers.de, aes(x = patient, y = avg_log2FC), fill = "red") +
  geom_boxplot(alpha= 0.8) +
  geom_point(color = color, size = 3) + 
  theme(legend.position = "right", axis.text.x = element_text(angle = 45)) +
  theme_bw()
    
hla.markers.boxplot.patient



```



PD1 Pathway Gene Expression Boxplots

```{r}

## Code for objects starts on line 429 of BCCIO analysis

bcc.tcells.pd1.averages.melt <- melt(bcc.tcells.pd1.averages, id.vars = c("Patient.ID", "Patients"), measure.vars = c("PDCD1", "CD274", "PDCD1LG2", "TCF7"))
bcc.bcells.pd1.averages.melt <- melt(bcc.bcells.pd1.averages, id.vars = c("Patient.ID", "Patients"), measure.vars = c("PDCD1", "CD274", "PDCD1LG2", "TCF7"))
bcc.macrophages.pd1.averages.melt <- melt(bcc.macrophages.pd1.averages, id.vars = c("Patient.ID", "Patients"), measure.vars = c("PDCD1", "CD274", "PDCD1LG2", "TCF7"))
bcc.tumor.pd1.averages.melt <- melt(bcc.tumor.pd1.averages, id.vars = c("Patient.ID", "Patients"), measure.vars = c("PDCD1", "CD274", "PDCD1LG2", "TCF7"))

bcc.tcells.pd1.averages.melt$variable <- as.character(bcc.tcells.pd1.averages.melt$variable)
bcc.tcells.pd1.averages.melt$variable[bcc.tcells.pd1.averages.melt$variable == "PDCD1"] <- "PD-1"
bcc.tcells.pd1.averages.melt$variable[bcc.tcells.pd1.averages.melt$variable == "CD274"] <- "PD-L1"
bcc.tcells.pd1.averages.melt$variable[bcc.tcells.pd1.averages.melt$variable == "PDCD1LG2"] <- "PD-L2"
bcc.tcells.pd1.averages.melt$variable <- as.factor(bcc.tcells.pd1.averages.melt$variable)
                                                   
bcc.bcells.pd1.averages.melt$variable <- as.character(bcc.bcells.pd1.averages.melt$variable)
bcc.bcells.pd1.averages.melt$variable[bcc.bcells.pd1.averages.melt$variable == "PDCD1"] <- "PD-1"
bcc.bcells.pd1.averages.melt$variable[bcc.bcells.pd1.averages.melt$variable == "CD274"] <- "PD-L1"
bcc.bcells.pd1.averages.melt$variable[bcc.bcells.pd1.averages.melt$variable == "PDCD1LG2"] <- "PD-L2"
bcc.bcells.pd1.averages.melt$variable <- as.factor(bcc.bcells.pd1.averages.melt$variable)
                                                   
bcc.macrophages.pd1.averages.melt$variable <- as.character(bcc.macrophages.pd1.averages.melt$variable)
bcc.macrophages.pd1.averages.melt$variable[bcc.macrophages.pd1.averages.melt$variable == "PDCD1"] <- "PD-1"
bcc.macrophages.pd1.averages.melt$variable[bcc.macrophages.pd1.averages.melt$variable == "CD274"] <- "PD-L1"
bcc.macrophages.pd1.averages.melt$variable[bcc.macrophages.pd1.averages.melt$variable == "PDCD1LG2"] <- "PD-L2"
bcc.macrophages.pd1.averages.melt$variable <- as.factor(bcc.macrophages.pd1.averages.melt$variable)
                                                   
bcc.tumor.pd1.averages.melt$variable <- as.character(bcc.tumor.pd1.averages.melt$variable)
bcc.tumor.pd1.averages.melt$variable[bcc.tumor.pd1.averages.melt$variable == "PDCD1"] <- "PD-1"
bcc.tumor.pd1.averages.melt$variable[bcc.tumor.pd1.averages.melt$variable == "CD274"] <- "PD-L1"
bcc.tumor.pd1.averages.melt$variable[bcc.tumor.pd1.averages.melt$variable == "PDCD1LG2"] <- "PD-L2"
bcc.tumor.pd1.averages.melt$variable <- as.factor(bcc.tumor.pd1.averages.melt$variable)

bcc.tcells.pd1.graphic <- ggboxplot(bcc.tcells.pd1.averages.melt, x = "variable", y = "value", color = "Patients", palette = c("Red", "Blue"), add = "jitter") + stat_compare_means(aes(group = Patients), method = "t.test")
bcc.bcells.pd1.graphic <- ggboxplot(bcc.bcells.pd1.averages.melt, x = "variable", y = "value", color = "Patients", palette = c("Red", "Blue"), add = "jitter") + stat_compare_means(aes(group = Patients), method = "t.test")
bcc.macrophages.pd1.graphic <- ggboxplot(bcc.macrophages.pd1.averages.melt, x = "variable", y = "value", color = "Patients", palette = c("Red", "Blue"), add = "jitter") + stat_compare_means(aes(group = Patients), method = "t.test")
bcc.tumor.pd1.graphic <- ggboxplot(bcc.tumor.pd1.averages.melt, x = "variable", y = "value", color = "Patients", palette = c("Red", "Blue"), add = "jitter") + stat_compare_means(aes(group = Patients), method = "t.test")

bcc.tcells.pd1.graphic <- ggpar(p = bcc.tcells.pd1.graphic, legend = "none", main = "T Cells", ylab = "Expression", font.main = c(22, "bold", "black"), font.x = "bold", font.y = "bold") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  rremove("xlab")

bcc.bcells.pd1.graphic <- ggpar(p = bcc.bcells.pd1.graphic, legend = "none", main = "B Cells", ylab = "Expression", font.main = c(22, "bold", "black"), font.x = "bold", font.y = "bold") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  rremove("xlab")

bcc.macrophages.pd1.graphic <- ggpar(p = bcc.macrophages.pd1.graphic, legend = "none", main = "Macrophages", ylab = "Expression", font.main = c(22, "bold", "black"), font.x = "bold", font.y = "bold") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  rremove("xlab")

bcc.tumor.pd1.graphic <- ggpar(p = bcc.tumor.pd1.graphic, legend = "none", main = "Malignant Cells", ylab = "Expression", font.main = c(22, "bold", "black"), font.x = "bold", font.y = "bold") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  rremove("xlab")

grid.arrange(bcc.tcells.pd1.graphic, bcc.bcells.pd1.graphic, bcc.macrophages.pd1.graphic, bcc.tumor.pd1.graphic, ncol=2)

```



################################################################################

Spring 2021 update presentation graphics

Reclustered PDAC UMAP 

```{r}

colors <- c("#d071db", "#ad4d4b", "#d78eca", "#715b9e", "#aebe6d", "#d77c34", "#9447b5", "#5c6bd7", "#db3a5b", "#d1aa34", "#5bc15b", "#d8a36a", "#91652b", "#9395de", "#468e4b", "#d44392", "#e5818b", "#9fc43f", "#a14b76", "#64d2af", "#4fa5d7", "#35927d", "#6c7a2d", "#d14d2f")

Idents(PDAC.all) <- PDAC.all$cluster
PDAC.all.UMAP <- DimPlot(PDAC.all, group.by="cluster", label="TRUE", repel="TRUE", cols=colors, pt.size=0.1, label.size=4)

ggarrange(PDAC.all.UMAP, ncol=1, legend="bottom")

PDAC.all.UMAP.cancerous <- DimPlot(PDAC.all, group.by="cluster", label="FALSE", cols=colors, pt.size=0.1, label.size=4, split.by="status") + NoLegend()

```



Boxplots comparing MHC Class I and Class II Scores

```{r}

`malignant.PDAC M` <- merge(`malignant.PDAC M1`, `malignant.PDAC M2`)

clusters <- list(`malignant.PDAC M`, `malignant.PDAC N`, `malignant.PDAC H`, `malignant.BCC RPre`, `malignant.BCC RPost`, `malignant.BCC NRPre`, `malignant.BCC NRPost`)

clusternames <- c('PDAC.M', 'PDAC.N', 'PDAC.H', 'BCC.RPre', 'BCC.RPost', 'BCC.NRPre', 'BCC.NRPost')

pdacclusters <- list(`malignant.PDAC M`, `malignant.PDAC N`, `malignant.PDAC H`)

for (i in 1:length(pdacclusters)) {
  clusters[[i]]$patient <- clusters[[i]]$orig.ident
}

for (i in 1:length(clusters)) {
  Idents(clusters[[i]]) <- clusters[[i]]$patient
  assign(paste0(clusternames[i], ".exp"), AverageExpression(clusters[[i]]))
}

cluster.averages <- list(PDAC.M.exp, PDAC.N.exp, PDAC.H.exp, BCC.RPre.exp, BCC.RPost.exp, BCC.NRPre.exp, BCC.NRPost.exp)

for (i in 1:length(cluster.averages)) {
  assign(paste0(clusternames[i], ".exp.mhc1"), cluster.averages[[i]]$RNA[mhc1.genes, ])
  assign(paste0(clusternames[i], ".exp.mhc2"), cluster.averages[[i]]$RNA[mhc2.genes, ])
}

cluster.averages.mhc1 <- list(PDAC.M.exp.mhc1, PDAC.N.exp.mhc1, PDAC.H.exp.mhc1, BCC.RPre.exp.mhc1, BCC.RPost.exp.mhc1, BCC.NRPre.exp.mhc1, BCC.NRPost.exp.mhc1)

cluster.averages.mhc2 <- list(PDAC.M.exp.mhc2, PDAC.N.exp.mhc2, PDAC.H.exp.mhc2, BCC.RPre.exp.mhc2, BCC.RPost.exp.mhc2, BCC.NRPre.exp.mhc2, BCC.NRPost.exp.mhc2)

for (i in 1:length(cluster.averages.mhc1)) {
  cluster.averages.mhc1[[i]] <- as.data.frame(colMeans(cluster.averages.mhc1[[i]]))
  cluster.averages.mhc1[[i]]$patient <- rownames(cluster.averages.mhc1[[i]])
  cluster.averages.mhc1[[i]]$cluster <- clusternames[i]
  colnames(cluster.averages.mhc1[[i]]) <- c("score", "patient", "cluster")
}

for (i in 1:length(cluster.averages.mhc2)) {
  cluster.averages.mhc2[[i]] <- as.data.frame(colMeans(cluster.averages.mhc2[[i]]))
  cluster.averages.mhc2[[i]]$patient <- rownames(cluster.averages.mhc2[[i]])
  cluster.averages.mhc2[[i]]$cluster <- clusternames[i]
  colnames(cluster.averages.mhc2[[i]]) <- c("score", "patient", "cluster")
}

cluster.averages.mhc1 <- rbind(cluster.averages.mhc1[[1]], cluster.averages.mhc1[[2]], cluster.averages.mhc1[[3]], cluster.averages.mhc1[[4]], cluster.averages.mhc1[[5]], cluster.averages.mhc1[[6]], cluster.averages.mhc1[[7]])

cluster.averages.mhc2 <- rbind(cluster.averages.mhc2[[1]], cluster.averages.mhc2[[2]], cluster.averages.mhc2[[3]], cluster.averages.mhc2[[4]], cluster.averages.mhc2[[5]], cluster.averages.mhc2[[6]], cluster.averages.mhc2[[7]])

comparisons <- list(c("BCC.RPre", "BCC.RPost"), c("BCC.NRPre", "BCC.NRPost"), c("BCC.RPre", "BCC.NRPre"), c("BCC.RPost", "BCC.NRPost"), c("PDAC.M", "PDAC.N"), c("PDAC.M", "PDAC.H"), c("PDAC.N", "PDAC.H"))

mhc1.score.comparison <- ggboxplot(cluster.averages.mhc1, x="cluster", y="score", color="cluster", title="MHC Class I Score", add="jitter") +
  stat_compare_means(comparisons = comparisons) +
  theme(legend.position="none", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5)) +
  labs(x="Cell Population", y="Score")

mhc2.score.comparison <- ggboxplot(cluster.averages.mhc2, x="cluster", y="score", color="cluster", title="MHC Class II Score", add="jitter") +
  stat_compare_means(comparisons = comparisons) +
  theme(legend.position="none", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5)) +
  labs(x="Cell Population", y="Score")

ggarrange(mhc1.score.comparison, mhc2.score.comparison, ncol=2)

```



PDF of MHC Class I and Class II Score

```{r}

mhc1.cdf.data.filtered <- mhc1.cdf.data
mhc1.cdf.data.filtered$id <- as.character(mhc1.cdf.data.filtered$id)
mhc1.cdf.data.filtered$id[mhc1.cdf.data.filtered$id == "PDAC M1" | mhc1.cdf.data.filtered$id == "PDAC M2"] <- "PDAC M"
mhc1.cdf.data.filtered$id <- as.factor(mhc1.cdf.data.filtered$id)
delete <- which(mhc1.cdf.data$id == "BCC RPost" | mhc1.cdf.data$id == "BCC NRPost")
mhc1.cdf.data.filtered <- mhc1.cdf.data.filtered[-c(delete),]

mhc2.cdf.data.filtered <- mhc2.cdf.data
mhc2.cdf.data.filtered$id <- as.character(mhc2.cdf.data.filtered$id)
mhc2.cdf.data.filtered$id[mhc2.cdf.data.filtered$id == "PDAC M1" | mhc2.cdf.data.filtered$id == "PDAC M2"] <- "PDAC M"
mhc2.cdf.data.filtered$id <- as.factor(mhc2.cdf.data.filtered$id)
delete <- which(mhc2.cdf.data$id == "BCC RPost" | mhc2.cdf.data$id == "BCC NRPost")
mhc2.cdf.data.filtered <- mhc2.cdf.data.filtered[!delete,]


mhc1.pdf <- ggplot(mhc1.cdf.data.filtered, aes(x=MHC.Class.I, fill=id)) + geom_density(alpha=0.3) +
  labs(title="MHC Class I", x="Score",y = "Probability Density") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))
  
mhc2.pdf <- ggplot(mhc2.cdf.data.filtered, aes(x=MHC.Class.II, fill=id)) + geom_density(alpha=0.3) +
  labs(title="MHC Class II", x="Score",y = "Probability Density") +
  theme_light() + xlim(0, 10) + ylim(0, 1) +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

ggarrange(mhc1.pdf, mhc2.pdf, ncol=2, common.legend=TRUE, legend="bottom")


mhc1.vln <- ggplot(mhc1.cdf.data.filtered, aes(x=MHC.Class.I/6, factor(id))) + geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
  labs(title="MHC Class I Score", x = "Sum of Normalized Count Matrix", y = "Cluster") +
  aes(fill=factor(id)) + theme_bw() + xlab("Score") +
  theme(legend.position="none", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

mhc2.vln <- ggplot(mhc2.cdf.data.filtered, aes(x=MHC.Class.II/12, factor(id))) + geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
  labs(title="MHC Class II Score", x = "Sum of Normalized Count Matrix", y = "Cluster") +
  aes(fill=factor(id)) + theme_bw() + xlim(0, 1.2) + xlab("Score") +
  theme(legend.position="none", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

ggarrange(mhc1.vln, mhc2.vln, ncol=2)

```



Marker Genes for Identification of PDAC CD8+ T Cells

```{r}

Idents(PDAC.all) <- PDAC.all$cluster
pdac.cd8 <- subset(PDAC.all, idents=c('CD8+ Effector', 'CD8+ Memory', 'CD8+ Exhausted'))

cd8markergenes <- c('PFN1', 'TMEM66', 'CXCR4', 'H1FX', 'HERPUD1', 'NCR3', 'DRAXIN', 'BTG1', 'LGALS3', 'LTB', 'RORA', 'APBA2', 'AIM1', 'TMEM173', 'PRF1', 'GFPT2', 'LGALS1', 'GABARAPL1', 'KLRB1', 'TOB1','S100A4', 'CCR6', 'KLRC1', 'SLC4A10', 'S100A6', 'CSF1', 'TNFRSF25', 'PTGER4', 'IL7R', 'GZMB', 'CD40LG','PTGER2', 'VNN2', 'FAM65B', 'NLRC5', 'LAT2', 'BCL2A1', 'NBEAL2', 'ATHL1', 'TRIM56', 'IRS2', 'IFITM3', 'RNF130', 'CD74', 'ADAM28', 'TCF7', 'GZMK', 'HLA-DPB1', 'CLDND1', 'HAVCR2', 'KLRG1', 'IFI44L','ITGB2', 'KLF2', 'EOMES', 'TNFRSF18', 'PPP1R14B', 'KLRC3', 'GPR183', 'CD27', 'KLRF1', 'TIGIT', 'HLA-DPA1')

pdac.cd8.markers <- DotPlot(pdac.cd8, features=cd8markergenes)
pdac.cd8.markers.data <- pdac.cd8.markers$data

ggplot(data=pdac.cd8.markers.data, aes(x=features.plot, y=id, fill=avg.exp.scaled)) +
  geom_tile(color="white") + 
  scale_fill_gradient2(low="black", high="#EFE350FF", mid="#593D9CFF", midpoint=0) +
  theme_minimal() + theme(axis.text.x = element_text(angle=90, vjust=1, hjust=1)) +
  labs(x="Gene", y="Cluster", fill="Expression") + coord_equal()

pdac.cd8.markers1 <- DotPlot(pdac.cd8, features=cd8markergenes[1:31])
pdac.cd8.markers2 <- DotPlot(pdac.cd8, features=cd8markergenes[32:62])


```



Final UMAPs for BCCIO and PDAC

```{r}


DimPlot(BCCIO, label=TRUE, label.size=10) + NoLegend()
## Saved at 2000 x 1374

DimPlot(PDAC.all, label=TRUE, label.size=10, repel=TRUE) + NoLegend()


```