---
title: "BCCIO / PDAC Comparison"
output: html_notebook
---

Load Packages and Data

```{r}
## Load packages
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(cowplot)
library(pracma)
library(RColorBrewer)
library(EnhancedVolcano)
library(gridExtra)
library(reshape2)

memory.limit(size = 16000000)

## Integrated PDAC datasets 
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 155698\\PDAC155698.integrated")

## BCCIO dataset
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 123813\\BCCIOwithUMAP")

## Merged dataset (5000 cells from PDAC and BCC, random sample)
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.PDAC.merge")

## Merged dataset (5000 cells from PDAC and BCC pretreatment, random sample)
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.pre.PDAC.merge")

## Load functions
source("ClusteringAnalysisFunctions.R", echo=TRUE)
source("SeuratClusteringFunctions.R", echo=TRUE)
source("DifferentialExpressionFunctions.R", echo=TRUE)

```



Merge PDAC and BCC datasets (subset)

```{r}

## Subset data 

BCCIO@meta.data$cell.names <- rownames(BCCIO@meta.data)
PDAC155698.integrated$cell.names <- rownames(PDAC155698.integrated@meta.data)

set.seed(111)
BCCIO.sampled.cells <- sample(x = BCCIO@meta.data$cell.names, size=5000, replace=F)
PDAC.sampled.cells <- sample(x = PDAC155698.integrated@meta.data$cell.names, size=5000, replace=F)

BCCIO.subset <- subset(BCCIO, cells = BCCIO.sampled.cells)
PDAC155698.integrated.subset <- subset(PDAC155698.integrated, cells = PDAC.sampled.cells)

BCCIO.subset$cancer <- c("bcc")
PDAC155698.integrated.subset$cancer <- c("pdac")

## Merge datasets

bcc.pdac.merge <- merge(BCCIO.subset, PDAC155698.integrated.subset, add.cell.ids = c("bcc", "pdac"))
Idents(bcc.pdac.merge) <- bcc.pdac.merge@meta.data$seurat_clusters

bcc.pdac.merge@meta.data <- bcc.pdac.merge@meta.data[, -which(colnames(bcc.pdac.merge@meta.data) %in% c('cell', 'cell.names', 'sub_cluster'))]
bcc.pdac.merge$orig.clusters <- bcc.pdac.merge$seurat_clusters

## Save merged dataset

save(bcc.pdac.merge, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.PDAC.merge")

```



Enhanced Volcano Plots (PDAC vs. BCC)

```{r}

## How to read volcano plots: 
## X-axis: difference in expression (log2 fold change) of gene between ident.1 and ident.2
## Y-axis: statistical significance of difference in expression (-Log10 probability) 
## Genes of interest: located on upper corners of plot


## CD8 T cells 

bcc.pdac.merge <- RenameIdents(bcc.pdac.merge, 'CD8_memory' = 'CD8+ T cells', 'CD8_activation' = 'CD8+ T cells', 'CD8_exhausted' = 'CD8+ T cells', 'CD4' = 'CD4+ T cells')

CD8.markers <- FindMarkers(bcc.pdac.merge, subset.ident='CD8+ T cells', group.by='cancer', ident.1='bcc')
CD8.markers_tibble <- as_tibble(CD8.markers, rownames = "GeneNames")
CD8.markers_tibble <- arrange(CD8.markers_tibble, avg_log2FC)

CD8.volcano <- EnhancedVolcano(CD8.markers_tibble,
                               lab = CD8.markers_tibble$GeneNames,
                               x = 'avg_log2FC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'CD8+ T Cell DE: BCC vs. PDAC')
CD8.volcano

CD8.volcano.pd1 <- EnhancedVolcano(CD8.markers_tibble,
                                   lab = CD8.markers_tibble$GeneNames,
                                   x = 'avg_log2FC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'CD8+ T Cell DE: BCC vs. PDAC')
CD8.volcano.pd1


## CD4 T cells

CD4.markers <- FindMarkers(bcc.pdac.merge, subset.ident='CD4+ T cells', group.by='cancer', ident.1='bcc')
CD4.markers_tibble <- as_tibble(CD4.markers, rownames = "GeneNames")
CD4.markers_tibble <- arrange(CD4.markers_tibble, avg_log2FC)

CD4.volcano <- EnhancedVolcano(CD4.markers_tibble,
                               lab = CD4.markers_tibble$GeneNames,
                               x = 'avg_log2FC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'CD4+ T Cell DE: BCC vs. PDAC')
CD4.volcano

CD4.volcano.pd1 <- EnhancedVolcano(CD4.markers_tibble,
                                   lab = CD4.markers_tibble$GeneNames,
                                   x = 'avg_log2FC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'CD4+ T Cell DE: BCC vs. PDAC')
CD4.volcano.pd1


## Macrophages

macrophage.markers <- FindMarkers(bcc.pdac.merge, subset.ident='Macrophages', group.by='cancer', ident.1='bcc')
macrophage.markers_tibble <- as_tibble(macrophage.markers, rownames = "GeneNames")
macrophage.markers_tibble <- arrange(macrophage.markers_tibble, avg_log2FC)

macrophage.volcano <- EnhancedVolcano(macrophage.markers_tibble,
                               lab = macrophage.markers_tibble$GeneNames,
                               x = 'avg_log2FC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'Macrophage DE: BCC vs. PDAC')
macrophage.volcano

macrophage.volcano.pd1 <- EnhancedVolcano(macrophage.markers_tibble,
                                   lab = macrophage.markers_tibble$GeneNames,
                                   x = 'avg_log2FC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'Macrophage DE: BCC vs. PDAC')
macrophage.volcano.pd1

```



Enhanced Volcano Plots (PDAC)

```{r}

pdac.macrophage.markers <- FindMarkers(PDAC155698.integrated, subset.ident='Macrophages', group.by='orig.ident', ident.1='PDAC155698.1')
pdac.macrophage.markers_tibble <- as_tibble(pdac.macrophage.markers, rownames = "GeneNames")
pdac.macrophage.markers_tibble <- arrange(pdac.macrophage.markers_tibble, avg_log2FC)

pdac.macrophage.volcano <- EnhancedVolcano(pdac.macrophage.markers_tibble,
                               lab = pdac.macrophage.markers_tibble$GeneNames,
                               x = 'avg_log2FC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'Macrophage DE: PDAC Patient 1 vs. Rest of Patients')
pdac.macrophage.volcano

```



Compare BCC Responders vs. Nonresponders

```{r}

## Label BCC dataset with responders vs. nonresponders

responders <- c("su001", "su002", "su003", "su004", "su009", "su012")
nonresponders <- c("su005", "su006", "su007", "su008", "su010")

for (patient in responders) {
  BCCIO@meta.data$response[BCCIO$patient == patient] = "Responsive"
}
for (patient in nonresponders) {
  BCCIO@meta.data$response[BCCIO$patient == patient] = "Nonresponsive"
}

## Enhanced Volcano Plots for T cells, B cells, and macrophages

Idents(BCCIO) <- BCCIO$seurat_clusters

bcc.cd4.responder.volcano <- bioc_volcano(BCCIO, c("CD4"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd4.responder.volcano

bcc.cd8_memory.responder.volcano <- bioc_volcano(BCCIO, c("CD8_memory"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_memory.responder.volcano

bcc.cd8_exhausted.responder.volcano <- bioc_volcano(BCCIO, c("CD8_exhausted"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_exhausted.responder.volcano

bcc.cd8_activation.responder.volcano <- bioc_volcano(BCCIO, c("CD8_activation"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_activation.responder.volcano

bcc.tregs.responder.volcano <- bioc_volcano(BCCIO, c("Tregs"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.tregs.responder.volcano

bcc.bcell.responder.volcano <- bioc_volcano(BCCIO, c("B cells"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.bcell.responder.volcano

bcc.macrophage.responder.volcano <- bioc_volcano(BCCIO, c("Macrophages"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.macrophage.responder.volcano

grid.arrange(bcc.cd4.responder.volcano, bcc.cd8_memory.responder.volcano, bcc.cd8_exhausted.responder.volcano, bcc.cd8_activation.responder.volcano, bcc.tregs.responder.volcano, bcc.bcell.responder.volcano, bcc.macrophage.responder.volcano, ncol=3)

```



Look specifically at BCC pre-treatment 
Goal: Determine predictors of PD1-blockade responsiveness

```{r}

## Split dataset into pre-treatment and post-treatment

Idents(BCCIO) <- BCCIO$treatment

BCCIO.pre <- subset(BCCIO, idents=c("pre"))
BCCIO.post <- subset(BCCIO, idents=c("post"))

## Enhanced Volcano plots (CD4, CD8, B, macrophage)

interestingcelltypes <- c("CD8_memory", "CD8_activation", "CD8_exhausted", "CD4", "Tregs", "T_proliferation", "B cells", "Macrophages")

Idents(BCCIO.pre) <- BCCIO.pre$seurat_clusters

for(celltype in interestingcelltypes) {
  bcc.pretreatment.responder.volcano <- bioc_volcano(BCCIO.pre, celltype, c("response"), c("Responsive"), c("Nonresponsive"))
  assign(paste0("bcc.pretreatment.", celltype, ".responder.volcano"), bcc.pretreatment.responder.volcano)
  bcc.pretreatment.responder.volcano
}

`bcc.pretreatment.B cells.responder.volcano`
## Pretty much no difference

`bcc.pretreatment.CD4.responder.volcano`
## Nothing of interest (CXCL13 higher in responders -> bad clustering?)

`bcc.pretreatment.CD8_activation.responder.volcano`
## Nothing of interest

`bcc.pretreatment.CD8_exhausted.responder.volcano`
## Lots of significant differences
## Higher amongst responders: CCL5 (associated with higher proliferation in melanoma)
## Higher amongst nonresponders: HSP group 

`bcc.pretreatment.CD8_memory.responder.volcano`
## Higher amongst responders: VIM (EMT cell marker -> leads to metastasis)
## Higher amongst nonresponders: HSP group, CCL4 (similar to CCL5)

`bcc.pretreatment.Macrophages.responder.volcano`
## Higher amongst responders: HLA genes (major role in immune regulation)

`bcc.pretreatment.Tregs.responder.volcano`
## Nothing of interest

`bcc.pretreatment.T_proliferation.responder.volcano`
## Higher amongst responders: HLA genes 

```



Conclusion?: look for higher expression of HLA genes in macrophages and T_proliferation 

Use Gene Ontology to investigate markers and gene pathways 

```{r}

## Macrophage markers

macrophage.markers <- FindMarkers(BCCIO.pre, subset.ident='Macrophages', group.by='response', ident.1='Responsive')
macrophage.markers_tibble <- as_tibble(macrophage.markers, rownames = "GeneNames")

macrophage.markers_tibble <- arrange(macrophage.markers_tibble, avg_log2FC)
 
macrophage.nonresponder.markers <- macrophage.markers_tibble[macrophage.markers_tibble$p_val_adj < 0.05 &  macrophage.markers_tibble$avg_log2FC < -0.5, ]

macrophage.responder.markers <- macrophage.markers_tibble[macrophage.markers_tibble$p_val_adj < 0.05 &  macrophage.markers_tibble$avg_log2FC > 0.5, ]

## Proliferated T cells markers 

T_proliferation.markers <- FindMarkers(BCCIO.pre, subset.ident='T_proliferation', group.by='response', ident.1='Responsive')
T_proliferation.markers_tibble <- as_tibble(T_proliferation.markers, rownames = "GeneNames")

T_proliferation.markers_tibble <- arrange(T_proliferation.markers_tibble, avg_log2FC)
 
T_proliferation.nonresponder.markers <- T_proliferation.markers_tibble[T_proliferation.markers_tibble$p_val_adj < 0.05 &  T_proliferation.markers_tibble$avg_log2FC < -0.5, ]

T_proliferation.responder.markers <- T_proliferation.markers_tibble[T_proliferation.markers_tibble$p_val_adj < 0.05 &  T_proliferation.markers_tibble$avg_log2FC > 0.5, ]

```



Merge BCC pretreatment data with PDAC data

```{r}

BCCIO.pre@meta.data$cell.names <- rownames(BCCIO.pre@meta.data)
PDAC155698.integrated$cell.names <- rownames(PDAC155698.integrated@meta.data)

cells.to.sample <- 5000
set.seed(111)
BCCIO.pre.sampled.cells <- sample(x = BCCIO.pre@meta.data$cell.names, size=cells.to.sample, replace=F)
PDAC.sampled.cells <- sample(x = PDAC155698.integrated@meta.data$cell.names, size=cells.to.sample, replace=F)

BCCIO.pre.subset <- subset(BCCIO.pre, cells = BCCIO.pre.sampled.cells)
PDAC155698.integrated.subset <- subset(PDAC155698.integrated, cells = PDAC.sampled.cells)

BCCIO.pre.subset$cancer <- c("bcc.pre")
PDAC155698.integrated.subset$cancer <- c("pdac")

## Merge datasets

bcc.pre.pdac.merge <- merge(BCCIO.pre.subset, PDAC155698.integrated.subset, add.cell.ids = c("bcc.pre", "pdac"))
Idents(bcc.pre.pdac.merge) <- bcc.pdac.merge@meta.data$seurat_clusters

bcc.pre.pdac.merge@meta.data <- bcc.pre.pdac.merge@meta.data[, -which(colnames(bcc.pre.pdac.merge@meta.data) %in% c('cell', 'cell.names', 'sub_cluster'))]
bcc.pre.pdac.merge$orig.clusters <- bcc.pre.pdac.merge$seurat_clusters

## Save merged dataset

save(bcc.pre.pdac.merge, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.pre.PDAC.merge")

```


Enhanced Volcano plots for BCC pretreatment (responder) vs. PDAC

```{r}

## Subset dataset to only include BCC pretreatment repsonders

Idents(bcc.pre.pdac.merge) <- bcc.pre.pdac.merge$response
bcc.pre.pdac.merge$response[is.na(bcc.pre.pdac.merge$response)] <- "PDAC"
Idents(bcc.pre.pdac.merge) <- bcc.pre.pdac.merge$response

bcc.pre.responders.pdac.merge <- subset(bcc.pre.pdac.merge, idents = c("PDAC", "Responsive"))

interestingcelltypes <- c("T_proliferation", "Macrophages")

Idents(bcc.pre.responders.pdac.merge) <- bcc.pre.responders.pdac.merge$seurat_clusters

for(celltype in interestingcelltypes) {
  bcc.pdac.responder.volcano <- bioc_volcano(bcc.pre.responders.pdac.merge, celltype, c("response"), c("Responsive"), c("PDAC"))
  assign(paste0("bcc.pdac.", celltype, ".responder.volcano"), bcc.pdac.responder.volcano)
  bcc.pdac.responder.volcano
}

## Macrophages
`bcc.pdac.Macrophages.responder.volcano`

## Proliferated T cells
`bcc.pdac.T_proliferation.responder.volcano`


```



Enhanced Volcano plots for PDAC by patient vs. BCC responders

```{r}

bcc.pre.pdac.merge$patient[is.na(bcc.pre.pdac.merge$patient)] <- c("pdac")
Idents(bcc.pre.pdac.merge) <- bcc.pre.pdac.merge@meta.data$patient

bcc.pre.responders.pdac.merge <- subset(bcc.pre.pdac.merge, idents=c("pdac", "su001", "su002", "su003", "su004", "su009", "su012"))

## Add PDAC patient column

bcc.pre.responders.pdac.merge$pdac.patient <- as.vector(bcc.pre.responders.pdac.merge$orig.ident)

bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre.tumor.cd45'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su002.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su003.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su004.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su004.pre.tumor.cd45'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su009.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su012.pre.tcell'] <- 'BCC'

## Create separate objects for each PDAC patient

Idents(bcc.pre.responders.pdac.merge) <- bcc.pre.responders.pdac.merge$pdac.patient

bcc.pdac.1 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.1', 'BCC'))
bcc.pdac.2 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.2', 'BCC'))
bcc.pdac.3 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.3', 'BCC'))
bcc.pdac.4 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.4', 'BCC'))
bcc.pdac.5 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.5', 'BCC'))
bcc.pdac.6 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.6', 'BCC'))
bcc.pdac.7 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.7', 'BCC'))
bcc.pdac.8 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.8', 'BCC'))
bcc.pdac.9 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.9', 'BCC'))
bcc.pdac.10 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.10', 'BCC'))
bcc.pdac.11A <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.11A', 'BCC'))
bcc.pdac.11B <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.11B', 'BCC'))
bcc.pdac.12 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.12', 'BCC'))
bcc.pdac.13 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.13', 'BCC'))
bcc.pdac.15 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.15', 'BCC'))
bcc.pdac.16 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.16', 'BCC'))

Idents(bcc.pdac.1) <- bcc.pdac.1$seurat_clusters
Idents(bcc.pdac.2) <- bcc.pdac.2$seurat_clusters
Idents(bcc.pdac.3) <- bcc.pdac.3$seurat_clusters
Idents(bcc.pdac.4) <- bcc.pdac.4$seurat_clusters
Idents(bcc.pdac.5) <- bcc.pdac.5$seurat_clusters
Idents(bcc.pdac.6) <- bcc.pdac.6$seurat_clusters
Idents(bcc.pdac.7) <- bcc.pdac.7$seurat_clusters
Idents(bcc.pdac.8) <- bcc.pdac.8$seurat_clusters
Idents(bcc.pdac.9) <- bcc.pdac.9$seurat_clusters
Idents(bcc.pdac.10) <- bcc.pdac.10$seurat_clusters
Idents(bcc.pdac.11A) <- bcc.pdac.11A$seurat_clusters
Idents(bcc.pdac.11B) <- bcc.pdac.11B$seurat_clusters
Idents(bcc.pdac.12) <- bcc.pdac.12$seurat_clusters
Idents(bcc.pdac.13) <- bcc.pdac.13$seurat_clusters
Idents(bcc.pdac.15) <- bcc.pdac.15$seurat_clusters
Idents(bcc.pdac.16) <- bcc.pdac.16$seurat_clusters

## Enhanced Volcano Plots (for loop doesn't work)

bcc.pdac.1.volcano <- bioc_volcano(bcc.pdac.1, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 1"))
bcc.pdac.2.volcano <- bioc_volcano(bcc.pdac.2, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 2"))
bcc.pdac.3.volcano <- bioc_volcano(bcc.pdac.3, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 3"))
bcc.pdac.4.volcano <- bioc_volcano(bcc.pdac.4, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 4"))
bcc.pdac.5.volcano <- bioc_volcano(bcc.pdac.5, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 5"))
bcc.pdac.6.volcano <- bioc_volcano(bcc.pdac.6, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 6"))
bcc.pdac.7.volcano <- bioc_volcano(bcc.pdac.7, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 7"))
bcc.pdac.8.volcano <- bioc_volcano(bcc.pdac.8, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 8"))
bcc.pdac.9.volcano <- bioc_volcano(bcc.pdac.9, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 9"))
bcc.pdac.10.volcano <- bioc_volcano(bcc.pdac.10, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 10"))
bcc.pdac.11A.volcano <- bioc_volcano(bcc.pdac.11A, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 11A"))
bcc.pdac.11B.volcano <- bioc_volcano(bcc.pdac.11B, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 11B"))
bcc.pdac.12.volcano <- bioc_volcano(bcc.pdac.12, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 12"))
bcc.pdac.13.volcano <- bioc_volcano(bcc.pdac.13, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 13"))
bcc.pdac.15.volcano <- bioc_volcano(bcc.pdac.15, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 15"))
bcc.pdac.16.volcano <- bioc_volcano(bcc.pdac.16, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 16"))

```



Check DE gene expression of HLA genes by patient 

```{r}
## Differentially expressed genes
`
bcc.pdac.subsets <- list(bcc.pdac.1, bcc.pdac.2, bcc.pdac.3, bcc.pdac.4, bcc.pdac.5, bcc.pdac.6, bcc.pdac.7, bcc.pdac.8, bcc.pdac.9, bcc.pdac.10, bcc.pdac.11A, bcc.pdac.11B, bcc.pdac.12, bcc.pdac.13, bcc.pdac.15, bcc.pdac.16)

bcc.pdac.markers <- lapply(bcc.pdac.subsets, function(i) bioc_markers(i, c("Macrophages"), c("pdac.patient"), c("BCC"), pCutoff = 1, FCcutoff = 0.01))

for (i in 1:length(bcc.pdac.markers)) {
  assign(paste0("bcc.pdac.", i, ".markers"), bcc.pdac.markers[[i]])
}

bcc.pdac.11A.markers <- bcc.pdac.11.markers
bcc.pdac.11B.markers <- bcc.pdac.12.markers
bcc.pdac.12.markers <- bcc.pdac.13.markers
bcc.pdac.13.markers <- bcc.pdac.14.markers

## Segregate the HLA genes

bcc.pdac.markers <- list(bcc.pdac.1.markers, bcc.pdac.2.markers, bcc.pdac.3.markers, bcc.pdac.4.markers, bcc.pdac.5.markers, bcc.pdac.6.markers, bcc.pdac.7.markers, bcc.pdac.8.markers, bcc.pdac.9.markers, bcc.pdac.10.markers, bcc.pdac.11A.markers, bcc.pdac.11B.markers, bcc.pdac.12.markers, bcc.pdac.13.markers, bcc.pdac.15.markers, bcc.pdac.16.markers)

patient.ids <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11A", "11B", "12", "13", "15", "16")

hla.markers <- NULL

for (i in 1:length(bcc.pdac.markers)) {
  bcc.pdac.markers[[i]]$gene.header <- substr(bcc.pdac.markers[[i]]$GeneNames, 0, 4)
  bcc.pdac.markers[[i]]$patient <- (patient.ids[i])
  hla.markers[[i]] <- subset(bcc.pdac.markers[[i]], gene.header == "HLA-")
}

for (i in 1:length(hla.markers)) {
  hla.markers[[i]]$patient
  assign(paste0("hla.markers.", i), hla.markers[[i]])
}

## Merge datasets together

hla.markers <- list(hla.markers.1, hla.markers.2, hla.markers.3, hla.markers.4, hla.markers.5, hla.markers.6, hla.markers.7, hla.markers.8, hla.markers.9, hla.markers.10, hla.markers.11, hla.markers.12, hla.markers.13, hla.markers.14, hla.markers.15, hla.markers.16)

patient.ids <- c('1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11A', '11B', '12', '13', '15', '16')

for (i in 1:length(hla.markers)) {
  hla.markers[[i]]$patient <- patient.ids[i]
}

hla.markers.by.patient <- do.call("rbind", hla.markers)

## Plot results

hla.markers.by.patient$log_pvaladj <- -log(hla.markers.by.patient$p_val_adj)
#hla.markers.by.patient$loglog_pvaladj <- log(hla.markers.by.patient$log_pvaladj + 1)

hla.markers.plot <- ggplot(hla.markers.by.patient, aes(x=log_pvaladj, y=avg_log2FC, color=GeneNames, size=1)) + theme_minimal() + ggtitle("Differential Expression of HLA genes by patient: \n BCC Responders vs. PDAC") + xlab("-log10(P)") + ylab("Log2 fold change") + geom_point() + theme(plot.title = element_text(size=20), legend.text = element_text(size = 10), legend.title = element_text(size = 14))
hla.markers.plot

```

Result: HLA-DRB5 and HLA-G have the greatest difference in response

Look at difference in HLA expression between BCC nonresponders and responders pretreatment 

```{r}

Idents(BCCIO) <- BCCIO@meta.data$treatment
BCCIO.pre <- subset(BCCIO, idents=c("pre"))

Idents(BCCIO.pre) <- BCCIO.pre$patient
bcc.pre.responder.5 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su005"))
bcc.pre.responder.6 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su006"))
bcc.pre.responder.7 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su007"))
bcc.pre.responder.8 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su008"))
bcc.pre.responder.10 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su010"))

Idents(bcc.pre.responder.5) <- bcc.pre.responder.5$seurat_clusters
Idents(bcc.pre.responder.6) <- bcc.pre.responder.6$seurat_clusters
Idents(bcc.pre.responder.7) <- bcc.pre.responder.7$seurat_clusters
Idents(bcc.pre.responder.8) <- bcc.pre.responder.8$seurat_clusters
Idents(bcc.pre.responder.10) <- bcc.pre.responder.10$seurat_clusters


## Enhanced volcano plots by patient

bcc.pre.response.volcano.5 <- bioc_volcano(bcc.pre.responder.5, c("Macrophages"), c("patient"), c("su005"), c("Responders"))
bcc.pre.response.volcano.6 <- bioc_volcano(bcc.pre.responder.6, c("Macrophages"), c("patient"), c("su006"), c("Responders"))
bcc.pre.response.volcano.7 <- bioc_volcano(bcc.pre.responder.7, c("Macrophages"), c("patient"), c("su007"), c("Responders"))
bcc.pre.response.volcano.8 <- bioc_volcano(bcc.pre.responder.8, c("Macrophages"), c("patient"), c("su008"), c("Responders"))
bcc.pre.response.volcano.10 <- bioc_volcano(bcc.pre.responder.10, c("Macrophages"), c("patient"), c("su010"), c("Responders")) # Doesn't work for some reason (very little cells anyways)

bcc.pre.response.volcano.5
bcc.pre.response.volcano.6
bcc.pre.response.volcano.7
bcc.pre.response.volcano.8

## Differentially expressed genes by patient

bcc.pre.response.markers.5 <- bioc_markers(bcc.pre.responder.5, c("Macrophages"), c("patient"), c("su005")) 
bcc.pre.response.markers.6 <- bioc_markers(bcc.pre.responder.6, c("Macrophages"), c("patient"), c("su006")) 
bcc.pre.response.markers.7 <- bioc_markers(bcc.pre.responder.7, c("Macrophages"), c("patient"), c("su007")) 
bcc.pre.response.markers.8 <- bioc_markers(bcc.pre.responder.8, c("Macrophages"), c("patient"), c("su008")) 

bcc.pre.response.markers <- list(bcc.pre.response.markers.5, bcc.pre.response.markers.6, bcc.pre.response.markers.7, bcc.pre.response.markers.8)

patient.ids <- c("su005", "su006", "su007", "su008")
bcc.hla.markers <- NULL

for (i in 1:length(bcc.pre.response.markers)) {
  bcc.pre.response.markers[[i]]$gene.header <- substr(bcc.pre.response.markers[[i]]$GeneNames, 0, 4)
  bcc.pre.response.markers[[i]]$patient <- (patient.ids[i])
  bcc.hla.markers[[i]] <- subset(bcc.pdac.markers[[i]], gene.header == "HLA-")
}

for (i in 1:length(bcc.hla.markers)) {
  assign(paste0("bcc.hla.", patient.ids[i], ".markers"), bcc.hla.markers[[i]])
}

bcc.hla.markers <- list(bcc.hla.su005.markers, bcc.hla.su006.markers, bcc.hla.su007.markers, bcc.hla.su008.markers)
bcc.hla.markers <- do.call("rbind", bcc.hla.markers)

## Plot results

bcc.hla.markers$log_pvaladj <- -log(bcc.hla.markers$p_val_adj)
#hla.markers.by.patient$loglog_pvaladj <- log(hla.markers.by.patient$log_pvaladj + 1)

bcc.hla.markers.plot <- ggplot(bcc.hla.markers, aes(x=log_pvaladj, y=avg_log2FC, color=GeneNames, size=1.5)) + theme_minimal() + ggtitle("Differential Expression of HLA genes by patient: \n BCC Responders vs. Nonresponders") + xlab("-log10(P)") + ylab("Log2 fold change") + geom_point() + theme(plot.title = element_text(size=20), legend.text = element_text(size=10), legend.title = element_text(size = 14))
bcc.hla.markers.plot



```

Result: HLA-DRB5 has greatest differential expression -- check to make sure same direction 

Violin Plots for Gene Expression 

```{r}

bcc.pdac.merge$patient[is.na(bcc.pdac.merge$patient)] <- "pdac"
Idents(bcc.pdac.merge) <- bcc.pdac.merge@meta.data$patient

Idents(BCCIO) <- BCCIO@meta.data$patient

HLADRB5.bcc.vlnplot <- VlnPlot(BCCIO, "HLA-DRB5")
HLADRB5.bcc.vlnplot

HLADRB5.pdac.vlnplot <- VlnPlot(PDAC155698.integrated, "HLA-DRB5")

View(PDAC155698.integrated@meta.data)

## Violin plots

Idents(BCCIO) <- BCCIO@meta.data$seurat_clusters

bcc.macrophages <- subset(BCCIO, idents = c("Macrophages"))
Idents(bcc.macrophages) <- bcc.macrophages@meta.data$patient


HLADRB5.bcc.macrophages.vlnplot <- VlnPlot(bcc.macrophages, "HLA-DRB5")
HLADRB5.bcc.macrophages.vlnplot

## Average expression for BCC

set.seed(001)

Idents(BCCIO) <- BCCIO@meta.data$patient
BCCIO.subset <- BCCIO[, sample(colnames(BCCIO), size=5000, replace=F)]

BCCIO.averages <- AverageExpression(BCCIO.subset)

head(BCCIO.averages[["RNA"]][c("HLA-DRB5"),])

## Average expression for PDAC

Idents(PDAC155698.integrated) <- PDAC155698.integrated@meta.data$orig.ident
PDAC.averages <- AverageExpression(PDAC155698.integrated)

head(PDAC.averages[["RNA"]][c("HLA-DRB5"),])

```



Conduct ANOVA tests

```{r}

## Create Seurat object with all macrophages from both datasets

Idents(BCCIO.pre) <- BCCIO.pre$seurat_clusters
bcc.macrophages <- subset(BCCIO.pre, idents = c("Macrophages"))

Idents(PDAC155698.integrated) <- PDAC155698.integrated$seurat_clusters
pdac.macrophages <- subset(PDAC155698.integrated, idents= c("Macrophages"))

bcc.pdac.macrophages <- merge(bcc.macrophages, pdac.macrophages, add.cell.ids = c("bcc", "pdac"))

## Average gene expression per patient for HLA 

hla.genes <- unique(hla.markers.by.patient$GeneNames)

Idents(bcc.pdac.macrophages) <- bcc.pdac.macrophages$patient
bcc.pdac.averages <- AverageExpression(bcc.pdac.macrophages)
hla.averages <- bcc.pdac.averages[["RNA"]][hla.genes,]

hla.averages <- rbind(hla.averages, c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC"))

rownames(hla.averages)[rownames(hla.averages) == "18"] <- "category"
hla.averages <- as.data.frame(t(hla.averages))

for (i in 1:18) {
  hla.averages[,i] <- as.character(hla.averages[,i])
  hla.averages[,i] <- as.numeric(hla.averages[,i])
}


save(hla.averages, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\hla.averages")

## ANOVA test for all HLA genes at 95% confidence

hla.anova.plots <- vector("list", length=17)

for (i in 1:length(hla.genes)) {
  colnames(hla.averages)[i] <- "gene"
  plot <- anova_test(hla.averages, hla.genes[i], "Category")
  hla.anova.plots[[i]] <- recordPlot()
  assign(paste0("hla.anova.plots.", hla.genes[i]), hla.anova.plots[[i]])
  plot.new()
  colnames(hla.averages)[i] <- hla.genes[i]
}

names(hla.anova.plots) <- hla.genes

hla.anova.plots <- list(hla.anova.plots[[1]], hla.anova.plots[[2]], hla.anova.plots[[3]], hla.anova.plots[[4]], hla.anova.plots[[5]], hla.anova.plots[[6]], hla.anova.plots[[7]], hla.anova.plots[[8]], hla.anova.plots[[9]], hla.anova.plots[[10]], hla.anova.plots[[11]], hla.anova.plots[[12]], hla.anova.plots[[13]], hla.anova.plots[[14]], hla.anova.plots[[15]], hla.anova.plots[[16]], hla.anova.plots[[17]])

grid.arrange(`hla.anova.plotsHLA-A`, `hla.anova.plotsHLA-B`)

## ANOVA test for all HLA genes at 90% confidence

hla.anova.90.plots <- vector("list", length=17)

for (i in 1:length(hla.genes)) {
  colnames(hla.averages)[i] <- "gene"
  plot <- anova_test(hla.averages, hla.genes[i], "Category", confidence = 0.90)
  hla.anova.90.plots[[i]] <- recordPlot()
  plot.new()
  colnames(hla.averages)[i] <- hla.genes[i]
}

names(hla.anova.90.plots) <- hla.genes

```

Result: no significant different in mean expression between BCC responders and nonresponders for any HLA genes individually

Analyze MHC I/II genes together 

```{r}

## MHC Class 1: HLA-A, HLA-B, HLA-C, HLA-E, HLA-G
## MHC Class 2: Everything else (HLA-D__)

## Arithmetic average score 

hla.averages.score <- data.frame(matrix(ncol=2, nrow=26))
colnames(hla.averages.score) <- c("MHC1", "MHC2")
rownames(hla.averages.score) <- rownames(hla.averages)

hla.averages.score$MHC1 <- (hla.averages$`HLA-A` + hla.averages$`HLA-B`+ hla.averages$`HLA-C` + hla.averages$`HLA-E` + hla.averages$`HLA-G`) / 5

hla.averages.score$MHC2 <- (hla.averages$`HLA-DOA`+ hla.averages$`HLA-DQB2` + hla.averages$`HLA-DMA` + hla.averages$`HLA-DQA2` + hla.averages$`HLA-DRA` + hla.averages$`HLA-DPA1` + hla.averages$`HLA-DQB1` + hla.averages$`HLA-DPB1` + hla.averages$`HLA-DRB1` + hla.averages$`HLA-DRB5` + hla.averages$`HLA-DQA1` + hla.averages$`HLA-DMB`) / 12

hla.averages.score$category <- hla.averages$category

hla.scores <- melt(hla.averages.score, id.var="category")
ggplot(data=hla.scores, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

## Normalized arithmetic average score

hla.averages.normalized <- hla.averages %>% mutate_at(hla.genes, ~(scale(.) %>% as.vector))

hla.averages.score.norm <- data.frame(matrix(ncol=2, nrow=26))
colnames(hla.averages.score.norm) <- c("MHC1", "MHC2")
rownames(hla.averages.score.norm) <- rownames(hla.averages.normalized)

hla.averages.score.norm$MHC1 <- (hla.averages.normalized$`HLA-A` + hla.averages.normalized$`HLA-B`+ hla.averages.normalized$`HLA-C` + hla.averages.normalized$`HLA-E` + hla.averages.normalized$`HLA-G`) / 5

hla.averages.score.norm$MHC2 <- (hla.averages.normalized$`HLA-DOA`+ hla.averages.normalized$`HLA-DQB2` + hla.averages.normalized$`HLA-DMA` + hla.averages.normalized$`HLA-DQA2` + hla.averages.normalized$`HLA-DRA` + hla.averages.normalized$`HLA-DPA1` + hla.averages.normalized$`HLA-DQB1` + hla.averages.normalized$`HLA-DPB1` + hla.averages.normalized$`HLA-DRB1` + hla.averages.normalized$`HLA-DRB5` + hla.averages.normalized$`HLA-DQA1` + hla.averages.normalized$`HLA-DMB`) / 12

hla.averages.score.norm$category <- hla.averages$category

hla.scores.norm <- melt(hla.averages.score.norm, id.var="category")
ggplot(data=hla.scores.norm, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

## Boxplots for each gene

hla.norm <- melt(hla.averages.normalized, id.var="category")
ggplot(data=hla.norm, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

## ANOVA Test on MHC Scores

mhc1.anova <- aov(MHC1~category, data=hla.averages.score)
mhc2.anova <- aov(MHC2~category, data=hla.averages.score)

mhc1.tukey <- TukeyHSD(mhc1.anova, conf.level = 0.95)
mhc2.tukey <- TukeyHSD(mhc2.anova, conf.level = 0.95)

tuk_plot(mhc1.tukey, title = "MHC I") + tuk_plot(mhc2.tukey, title = "MHC II")

## ANOVA Test on Normalized MHC Scores

mhc1.normalized.anova <- aov(MHC1~category, data=hla.averages.score.norm)
mhc2.normalized.anova <- aov(MHC2~category, data=hla.averages.score.norm)

mhc1.normalized.tukey <- TukeyHSD(mhc1.normalized.anova, conf.level=0.95)
mhc2.normalized.tukey <- TukeyHSD(mhc2.normalized.anova, conf.level=0.95)

tuk_plot(mhc1.normalized.tukey, title = "MHC I") + tuk_plot(mhc2.normalized.tukey, title = "MHC II")



```

Result: NR and PDAC have lower scores on MHC I compared to R; differences inconclusive for MHC II
Function of MHC Class I genes: display intracellular proteins to cytotoxic T cells (function to kill cancer cells)

Check MHC scores for malignant cell populations

```{r}

## Subset malignant cell populations for BCC, combine with all PDAC cells

Idents(BCCIO) <- BCCIO$seurat_clusters
bcc.malignant <- subset(BCCIO, idents = c("Macrophages", "Tumor 1", "Tumor 2"))

Idents(PDAC155698.integrated) <- PDAC155698.integrated$seurat_clusters
PDAC.subset <- subset(PDAC155698.integrated, idents = c("Acinar cells", "Epithelial + Endocrine"))
PDAC.subset$response <- "PDAC"

bcc.pdac.malignant <- merge(bcc.malignant, PDAC.subset, add.cell.ids = c("bcc", "pdac"))
bcc.pdac.malignant$category <- paste0(bcc.pdac.malignant$response, " ", bcc.pdac.malignant$seurat_clusters)


## Average gene expression per patient for HLA 

Idents(bcc.pdac.malignant) <- bcc.pdac.malignant$patient
bcc.pdac.malignant.averages <- AverageExpression(bcc.pdac.malignant)
hla.malignant.averages <- bcc.pdac.malignant.averages[["RNA"]][hla.genes,]

rownames(hla.malignant.averages)[rownames(hla.malignant.averages) == "18"] <- "category"
hla.malignant.averages <- as.data.frame(t(hla.malignant.averages))

for (i in 1:17) {
  hla.malignant.averages[,i] <- as.character(hla.malignant.averages[,i])
  hla.malignant.averages[,i] <- as.numeric(hla.malignant.averages[,i])
}

hla.malignant.averages <- cbind(hla.malignant.averages, c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC"))

colnames(hla.malignant.averages)[18] <- "category"

save(hla.malignant.averages, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\hla.malignant.averages")

## MHC Class I Score 

hla.malignant.averages.score <- data.frame(matrix(ncol=2, nrow=27))
colnames(hla.malignant.averages.score) <- c("MHC1", "MHC2")
rownames(hla.malignant.averages.score) <- rownames(hla.malignant.averages)[1:27]

hla.malignant.averages.score$MHC1 <- (hla.malignant.averages$`HLA-A` + hla.malignant.averages$`HLA-B`+ hla.malignant.averages$`HLA-C` + hla.malignant.averages$`HLA-E` + hla.malignant.averages$`HLA-G`) / 5

hla.malignant.averages.score$MHC2 <- (hla.malignant.averages$`HLA-DOA`+ hla.malignant.averages$`HLA-DQB2` + hla.malignant.averages$`HLA-DMA` + hla.malignant.averages$`HLA-DQA2` + hla.malignant.averages$`HLA-DRA` + hla.malignant.averages$`HLA-DPA1` + hla.malignant.averages$`HLA-DQB1` + hla.malignant.averages$`HLA-DPB1` + hla.malignant.averages$`HLA-DRB1` + hla.malignant.averages$`HLA-DRB5` + hla.malignant.averages$`HLA-DQA1` + hla.malignant.averages$`HLA-DMB`) / 12

hla.malignant.averages.score$category <- c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC")

hla.malignant.scores <- melt(hla.malignant.averages.score, id.var="category")
ggplot(data=hla.malignant.scores, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

## ANOVA tests

mhc1.malignant.anova <- aov(MHC1~category, data=hla.malignant.averages.score)
mhc2.malignant.anova <- aov(MHC2~category, data=hla.malignant.averages.score)

mhc1.malignant.tukey <- TukeyHSD(mhc1.malignant.anova, conf.level = 0.95)
mhc2.malignant.tukey <- TukeyHSD(mhc2.malignant.anova, conf.level = 0.95)

tuk_plot(mhc1.malignant.tukey, title = "MHC I") + tuk_plot(mhc2.malignant.tukey, title = "MHC II")

mhc1.malignant.tukey
mhc2.malignant.tukey

```

Result: MHC Class I and Class II lower in PDAC vs. responders and nonresponders 
MHC Class II in PDAC --> essentially no expression, much lower than responders or nonresponders



HSP Genes Differential Expression: Responders vs. PDAC 

```{r}

## Segregate the HSP genes

bcc.pdac.markers <- list(bcc.pdac.1.markers, bcc.pdac.2.markers, bcc.pdac.3.markers, bcc.pdac.4.markers, bcc.pdac.5.markers, bcc.pdac.6.markers, bcc.pdac.7.markers, bcc.pdac.8.markers, bcc.pdac.9.markers, bcc.pdac.10.markers, bcc.pdac.11A.markers, bcc.pdac.11B.markers, bcc.pdac.12.markers, bcc.pdac.13.markers, bcc.pdac.15.markers, bcc.pdac.16.markers)

patient.ids <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11A", "11B", "12", "13", "15", "16")

hsp.markers <- NULL

for (i in 1:length(bcc.pdac.markers)) {
  bcc.pdac.markers[[i]]$gene.header <- substr(bcc.pdac.markers[[i]]$GeneNames, 0, 3)
  bcc.pdac.markers[[i]]$patient <- (patient.ids[i])
  hsp.markers[[i]] <- subset(bcc.pdac.markers[[i]], gene.header == "HSP")
}

for (i in 1:length(hsp.markers)) {
  hsp.markers[[i]]$patient
  assign(paste0("hsp.markers.", i), hsp.markers[[i]])
}

## Merge datasets together

hsp.markers <- list(hsp.markers.1, hsp.markers.2, hsp.markers.3, hsp.markers.4, hsp.markers.5, hsp.markers.6, hsp.markers.7, hsp.markers.8, hsp.markers.9, hsp.markers.10, hsp.markers.11, hsp.markers.12, hsp.markers.13, hsp.markers.14, hsp.markers.15, hsp.markers.16)

hsp.markers.by.patient <- do.call("rbind", hsp.markers)

## Plot results

hsp.markers.by.patient$log_pvaladj <- -log(hsp.markers.by.patient$p_val_adj)
#hsp.markers.by.patient$loglog_pvaladj <- log(hsp.markers.by.patient$log_pvaladj + 1)

hsp.markers.plot <- ggplot(hsp.markers.by.patient, aes(x=log_pvaladj, y=avg_log2FC, color=GeneNames, size=1)) + theme_minimal() + ggtitle("Differential Expression of HSP genes by patient: \n BCC Responders vs. PDAC") + xlab("-log10(P)") + ylab("Log2 fold change") + geom_point() + theme(plot.title = element_text(size=20), legend.text = element_text(size = 10), legend.title = element_text(size = 14))
hsp.markers.plot

```



HSP Genes Differential Expression: Responders vs. Nonresponders

```{r}

## Differentially expressed genes by patient

patient.ids <- c("su005", "su006", "su007", "su008")
bcc.hsp.markers <- NULL

for (i in 1:length(bcc.pre.response.markers)) {
  bcc.pre.response.markers[[i]]$gene.header <- substr(bcc.pre.response.markers[[i]]$GeneNames, 0, 3)
  bcc.pre.response.markers[[i]]$patient <- (patient.ids[i])
  bcc.hsp.markers[[i]] <- subset(bcc.pre.response.markers[[i]], gene.header == "HSP")
}



for (i in 1:length(bcc.hsp.markers)) {
  assign(paste0("bcc.hsp.", patient.ids[i], ".markers"), bcc.hsp.markers[[i]])
}

bcc.hsp.markers <- list(bcc.hsp.su005.markers, bcc.hsp.su006.markers, bcc.hsp.su007.markers, bcc.hsp.su008.markers)
bcc.hsp.markers <- do.call("rbind", bcc.hsp.markers)

## Plot results

bcc.hsp.markers$log_pvaladj <- -log(bcc.hsp.markers$p_val_adj)

bcc.hsp.markers.plot <- ggplot(bcc.hsp.markers, aes(x=log_pvaladj, y=avg_log2FC, color=GeneNames, size=1.5)) + theme_minimal() + ggtitle("Differential Expression of HSP genes by patient: \n BCC Responders vs. Nonresponders") + xlab("-log10(P)") + ylab("Log2 fold change") + geom_point() + theme(plot.title = element_text(size=20), legend.text = element_text(size=10), legend.title = element_text(size = 14))
bcc.hsp.markers.plot

```

Result: no clear trends, opposite direction from responder vs. PDAC

HSP Score Analysis

```{r}

## Average gene expression per patient for HSP

hsp.genes <- unique(hsp.markers.by.patient$GeneNames)

Idents(bcc.pdac.macrophages) <- bcc.pdac.macrophages$patient
bcc.pdac.averages <- AverageExpression(bcc.pdac.macrophages)
hsp.averages <- bcc.pdac.averages[["RNA"]][hsp.genes,]

hsp.averages <- rbind(hsp.averages, c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC"))

rownames(hsp.averages)[rownames(hsp.averages) == "15"] <- "category"
hsp.averages <- as.data.frame(t(hsp.averages))

for (i in 1:14) {
  hsp.averages[,i] <- as.character(hsp.averages[,i])
  hsp.averages[,i] <- as.numeric(hsp.averages[,i])
}

colnames(hsp.averages)[15] <- "category"

save(hsp.averages, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\hsp.averages")

## Calculate HSP score (arithmetic average)

hsp.averages.score <- data.frame(matrix(ncol=2, nrow=25))
colnames(hsp.averages.score) <- c("HSP", "category")
rownames(hsp.averages.score) <- rownames(hsp.averages)

hsp.averages.score$HSP <- rowSums(hsp.averages[,-15]) / 14

hsp.averages.score$category <- hsp.averages$category

hsp.scores <- melt(hsp.averages.score, id.var="category")
ggplot(data=hsp.scores, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

hsp.anova <- aov(HSP~category, data=hsp.averages.score)
hsp.tukey <- TukeyHSD(hsp.anova, conf.level = 0.95)

tuk_plot(hsp.tukey, title = "HSP")

```



HSP Score Analysis: Malignant Cells

```{r}

## Average gene expression for HSP malignant cells

hsp.genes <- unique(hsp.markers.by.patient$GeneNames)

Idents(bcc.pdac.malignant) <- bcc.pdac.malignant$patient
bcc.pdac.malignant.averages <- AverageExpression(bcc.pdac.malignant)
hsp.malignant.averages <- bcc.pdac.malignant.averages[["RNA"]][hsp.genes,]

hsp.malignant.averages <- rbind(hsp.malignant.averages, c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC"))

rownames(hsp.malignant.averages)[rownames(hsp.malignant.averages) == "18"] <- "category"
hsp.malignant.averages <- as.data.frame(t(hsp.malignant.averages))

for (i in 1:17) {
  hsp.malignant.averages[,i] <- as.character(hsp.malignant.averages[,i])
  hsp.malignant.averages[,i] <- as.numeric(hsp.malignant.averages[,i])
}

colnames(hsp.malignant.averages)[18] <- "category"

save(hsp.malignant.averages, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\hsp.malignant.averages")

## Calculate HSP score (arithmetic average)

hsp.malignant.averages.score <- data.frame(matrix(ncol=2, nrow=27))
colnames(hsp.malignant.averages.score) <- c("HSP", "category")
rownames(hsp.malignant.averages.score) <- rownames(hsp.malignant.averages)

hsp.malignant.averages.score$HSP <- rowSums(hsp.malignant.averages[,-18]) / 17

hsp.malignant.averages.score$category <- hsp.malignant.averages$category

hsp.malignant.scores <- melt(hsp.malignant.averages.score, id.var="category")
ggplot(data=hsp.malignant.scores, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

hsp.malignant.anova <- aov(HSP~category, data=hsp.malignant.averages.score)
hsp.malignant.tukey <- TukeyHSD(hsp.malignant.anova, conf.level = 0.95)

tuk_plot(hsp.malignant.tukey, title = "HSP")

```




####################################################################################################################################


```{r}

Idents(BCCIO) <- BCCIO$seurat_clusters
bcc.averages.celltype <- AverageExpression(BCCIO)
bcc.hla.averages.celltype <- bcc.averages.celltype[["RNA"]][hla.genes,]
colnames(bcc.hla.averages.celltype) <- paste0("bcc ", colnames(bcc.hla.averages.celltype))
bcc.hla.averages.celltype$gene <- rownames(bcc.hla.averages.celltype)

Idents(PDAC155698.integrated) <- PDAC155698.integrated$seurat_clusters
pdac.averages.celltype <- AverageExpression(PDAC155698.integrated)
pdac.hla.averages.celltype <- pdac.averages.celltype[["RNA"]][hla.genes,]
colnames(pdac.hla.averages.celltype) <- paste0("pdac ", colnames(pdac.hla.averages.celltype))
pdac.hla.averages.celltype$gene <- rownames(pdac.hla.averages.celltype)

hla.averages.celltype <- merge(bcc.hla.averages.celltype, pdac.hla.averages.celltype, by="gene")
hla.averages.celltype <- t(hla.averages.celltype)
colnames(hla.averages.celltype) <- hla.averages.celltype[1,]
hla.averages.celltype <- hla.averages.celltype[-1,]




```

####################################################################################################################################




wilcoxon test
ks test
