---
title: "BCCIO / PDAC Comparison"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

Load Packages and Data

```{r}
## Load packages
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(cowplot)
library(pracma)
library(RColorBrewer)
library(EnhancedVolcano)
library(gridExtra)
library(reshape2)
library(ggpubr)

memory.limit(size = 16000000)

## PDAC infected patients dataset 
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 155698\\PDAC.integrated")

## PDAC infected + healthy patients dataset
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 155698\\PDAC.all")

## BCCIO dataset
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 123813\\BCCIOwithUMAP")

## Merged dataset (5000 cells from PDAC and BCC, random sample)
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.PDAC.merge")

## Merged dataset (5000 cells from PDAC and BCC pretreatment, random sample)
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.pre.PDAC.merge")

## Load functions
source("ClusteringAnalysisFunctions.R", echo=FALSE)
source("SeuratClusteringFunctions.R", echo=FALSE)
source("DifferentialExpressionFunctions.R", echo=FALSE)

```



Merge PDAC and BCC datasets (subset)

```{r}

## Subset data 

BCCIO@meta.data$cell.names <- rownames(BCCIO@meta.data)
PDAC155698.integrated$cell.names <- rownames(PDAC155698.integrated@meta.data)

set.seed(111)
BCCIO.sampled.cells <- sample(x = BCCIO@meta.data$cell.names, size=5000, replace=F)
PDAC.sampled.cells <- sample(x = PDAC155698.integrated@meta.data$cell.names, size=5000, replace=F)

BCCIO.subset <- subset(BCCIO, cells = BCCIO.sampled.cells)
PDAC155698.integrated.subset <- subset(PDAC155698.integrated, cells = PDAC.sampled.cells)

BCCIO.subset$cancer <- c("bcc")
PDAC155698.integrated.subset$cancer <- c("pdac")

## Merge datasets

bcc.pdac.merge <- merge(BCCIO.subset, PDAC155698.integrated.subset, add.cell.ids = c("bcc", "pdac"))
Idents(bcc.pdac.merge) <- bcc.pdac.merge@meta.data$seurat_clusters

bcc.pdac.merge@meta.data <- bcc.pdac.merge@meta.data[, -which(colnames(bcc.pdac.merge@meta.data) %in% c('cell', 'cell.names', 'sub_cluster'))]
bcc.pdac.merge$orig.clusters <- bcc.pdac.merge$seurat_clusters

## Save merged dataset

save(bcc.pdac.merge, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.PDAC.merge")

```



Enhanced Volcano Plots (PDAC vs. BCC)

```{r}

## How to read volcano plots: 
## X-axis: difference in expression (log2 fold change) of gene between ident.1 and ident.2
## Y-axis: statistical significance of difference in expression (-Log10 probability) 
## Genes of interest: located on upper corners of plot


## CD8 T cells 

bcc.pdac.merge <- RenameIdents(bcc.pdac.merge, 'CD8_memory' = 'CD8+ T cells', 'CD8_activation' = 'CD8+ T cells', 'CD8_exhausted' = 'CD8+ T cells', 'CD4' = 'CD4+ T cells')

CD8.markers <- FindMarkers(bcc.pdac.merge, subset.ident='CD8+ T cells', group.by='cancer', ident.1='bcc')
CD8.markers_tibble <- as_tibble(CD8.markers, rownames = "GeneNames")
CD8.markers_tibble <- arrange(CD8.markers_tibble, avg_log2FC)

CD8.volcano <- EnhancedVolcano(CD8.markers_tibble,
                               lab = CD8.markers_tibble$GeneNames,
                               x = 'avg_log2FC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'CD8+ T Cell DE: BCC vs. PDAC')
CD8.volcano

CD8.volcano.pd1 <- EnhancedVolcano(CD8.markers_tibble,
                                   lab = CD8.markers_tibble$GeneNames,
                                   x = 'avg_log2FC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'CD8+ T Cell DE: BCC vs. PDAC')
CD8.volcano.pd1


## CD4 T cells

CD4.markers <- FindMarkers(bcc.pdac.merge, subset.ident='CD4+ T cells', group.by='cancer', ident.1='bcc')
CD4.markers_tibble <- as_tibble(CD4.markers, rownames = "GeneNames")
CD4.markers_tibble <- arrange(CD4.markers_tibble, avg_log2FC)

CD4.volcano <- EnhancedVolcano(CD4.markers_tibble,
                               lab = CD4.markers_tibble$GeneNames,
                               x = 'avg_log2FC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'CD4+ T Cell DE: BCC vs. PDAC')
CD4.volcano

CD4.volcano.pd1 <- EnhancedVolcano(CD4.markers_tibble,
                                   lab = CD4.markers_tibble$GeneNames,
                                   x = 'avg_log2FC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'CD4+ T Cell DE: BCC vs. PDAC')
CD4.volcano.pd1


## Macrophages

macrophage.markers <- FindMarkers(bcc.pdac.merge, subset.ident='Macrophages', group.by='cancer', ident.1='bcc')
macrophage.markers_tibble <- as_tibble(macrophage.markers, rownames = "GeneNames")
macrophage.markers_tibble <- arrange(macrophage.markers_tibble, avg_log2FC)

macrophage.volcano <- EnhancedVolcano(macrophage.markers_tibble,
                               lab = macrophage.markers_tibble$GeneNames,
                               x = 'avg_log2FC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'Macrophage DE: BCC vs. PDAC')
macrophage.volcano

macrophage.volcano.pd1 <- EnhancedVolcano(macrophage.markers_tibble,
                                   lab = macrophage.markers_tibble$GeneNames,
                                   x = 'avg_log2FC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'Macrophage DE: BCC vs. PDAC')
macrophage.volcano.pd1

```



Enhanced Volcano Plots (PDAC)

```{r}

pdac.macrophage.markers <- FindMarkers(PDAC155698.integrated, subset.ident='Macrophages', group.by='orig.ident', ident.1='PDAC155698.1')
pdac.macrophage.markers_tibble <- as_tibble(pdac.macrophage.markers, rownames = "GeneNames")
pdac.macrophage.markers_tibble <- arrange(pdac.macrophage.markers_tibble, avg_log2FC)

pdac.macrophage.volcano <- EnhancedVolcano(pdac.macrophage.markers_tibble,
                               lab = pdac.macrophage.markers_tibble$GeneNames,
                               x = 'avg_log2FC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'Macrophage DE: PDAC Patient 1 vs. Rest of Patients')
pdac.macrophage.volcano

```



Compare BCC Responders vs. Nonresponders

```{r}

## Label BCC dataset with responders vs. nonresponders

responders <- c("su001", "su002", "su003", "su004", "su009", "su012")
nonresponders <- c("su005", "su006", "su007", "su008", "su010")

for (patient in responders) {
  BCCIO@meta.data$response[BCCIO$patient == patient] = "Responsive"
}
for (patient in nonresponders) {
  BCCIO@meta.data$response[BCCIO$patient == patient] = "Nonresponsive"
}

## Enhanced Volcano Plots for T cells, B cells, and macrophages

Idents(BCCIO) <- BCCIO$seurat_clusters

bcc.cd4.responder.volcano <- bioc_volcano(BCCIO, c("CD4"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd4.responder.volcano

bcc.cd8_memory.responder.volcano <- bioc_volcano(BCCIO, c("CD8_memory"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_memory.responder.volcano

bcc.cd8_exhausted.responder.volcano <- bioc_volcano(BCCIO, c("CD8_exhausted"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_exhausted.responder.volcano

bcc.cd8_activation.responder.volcano <- bioc_volcano(BCCIO, c("CD8_activation"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_activation.responder.volcano

bcc.tregs.responder.volcano <- bioc_volcano(BCCIO, c("Tregs"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.tregs.responder.volcano

bcc.bcell.responder.volcano <- bioc_volcano(BCCIO, c("B cells"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.bcell.responder.volcano

bcc.macrophage.responder.volcano <- bioc_volcano(BCCIO, c("Macrophages"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.macrophage.responder.volcano

grid.arrange(bcc.cd4.responder.volcano, bcc.cd8_memory.responder.volcano, bcc.cd8_exhausted.responder.volcano, bcc.cd8_activation.responder.volcano, bcc.tregs.responder.volcano, bcc.bcell.responder.volcano, bcc.macrophage.responder.volcano, ncol=3)

```



Look specifically at BCC pre-treatment 
Goal: Determine predictors of PD1-blockade responsiveness

```{r}

## Split dataset into pre-treatment and post-treatment

Idents(BCCIO) <- BCCIO$treatment

BCCIO.pre <- subset(BCCIO, idents=c("pre"))
BCCIO.post <- subset(BCCIO, idents=c("post"))

## Enhanced Volcano plots (CD4, CD8, B, macrophage)

interestingcelltypes <- c("CD8_memory", "CD8_activation", "CD8_exhausted", "CD4", "Tregs", "T_proliferation", "B cells", "Macrophages")

Idents(BCCIO.pre) <- BCCIO.pre$seurat_clusters

for(celltype in interestingcelltypes) {
  bcc.pretreatment.responder.volcano <- bioc_volcano(BCCIO.pre, celltype, c("response"), c("Responsive"), c("Nonresponsive"))
  assign(paste0("bcc.pretreatment.", celltype, ".responder.volcano"), bcc.pretreatment.responder.volcano)
  bcc.pretreatment.responder.volcano
}

`bcc.pretreatment.B cells.responder.volcano`
## Pretty much no difference

`bcc.pretreatment.CD4.responder.volcano`
## Nothing of interest (CXCL13 higher in responders -> bad clustering?)

`bcc.pretreatment.CD8_activation.responder.volcano`
## Nothing of interest

`bcc.pretreatment.CD8_exhausted.responder.volcano`
## Lots of significant differences
## Higher amongst responders: CCL5 (associated with higher proliferation in melanoma)
## Higher amongst nonresponders: HSP group 

`bcc.pretreatment.CD8_memory.responder.volcano`
## Higher amongst responders: VIM (EMT cell marker -> leads to metastasis)
## Higher amongst nonresponders: HSP group, CCL4 (similar to CCL5)

`bcc.pretreatment.Macrophages.responder.volcano`
## Higher amongst responders: HLA genes (major role in immune regulation)

`bcc.pretreatment.Tregs.responder.volcano`
## Nothing of interest

`bcc.pretreatment.T_proliferation.responder.volcano`
## Higher amongst responders: HLA genes 

```



Conclusion?: look for higher expression of HLA genes in macrophages and T_proliferation 

Use Gene Ontology to investigate markers and gene pathways 

```{r}

## Macrophage markers

macrophage.markers <- FindMarkers(BCCIO.pre, subset.ident='Macrophages', group.by='response', ident.1='Responsive')
macrophage.markers_tibble <- as_tibble(macrophage.markers, rownames = "GeneNames")

macrophage.markers_tibble <- arrange(macrophage.markers_tibble, avg_log2FC)
 
macrophage.nonresponder.markers <- macrophage.markers_tibble[macrophage.markers_tibble$p_val_adj < 0.05 &  macrophage.markers_tibble$avg_log2FC < -0.5, ]

macrophage.responder.markers <- macrophage.markers_tibble[macrophage.markers_tibble$p_val_adj < 0.05 &  macrophage.markers_tibble$avg_log2FC > 0.5, ]

## Proliferated T cells markers 

T_proliferation.markers <- FindMarkers(BCCIO.pre, subset.ident='T_proliferation', group.by='response', ident.1='Responsive')
T_proliferation.markers_tibble <- as_tibble(T_proliferation.markers, rownames = "GeneNames")

T_proliferation.markers_tibble <- arrange(T_proliferation.markers_tibble, avg_log2FC)
 
T_proliferation.nonresponder.markers <- T_proliferation.markers_tibble[T_proliferation.markers_tibble$p_val_adj < 0.05 &  T_proliferation.markers_tibble$avg_log2FC < -0.5, ]

T_proliferation.responder.markers <- T_proliferation.markers_tibble[T_proliferation.markers_tibble$p_val_adj < 0.05 &  T_proliferation.markers_tibble$avg_log2FC > 0.5, ]

```



Merge BCC pretreatment data with PDAC data

```{r}

BCCIO.pre@meta.data$cell.names <- rownames(BCCIO.pre@meta.data)
PDAC155698.integrated$cell.names <- rownames(PDAC155698.integrated@meta.data)

cells.to.sample <- 5000
set.seed(111)
BCCIO.pre.sampled.cells <- sample(x = BCCIO.pre@meta.data$cell.names, size=cells.to.sample, replace=F)
PDAC.sampled.cells <- sample(x = PDAC155698.integrated@meta.data$cell.names, size=cells.to.sample, replace=F)

BCCIO.pre.subset <- subset(BCCIO.pre, cells = BCCIO.pre.sampled.cells)
PDAC155698.integrated.subset <- subset(PDAC155698.integrated, cells = PDAC.sampled.cells)

BCCIO.pre.subset$cancer <- c("bcc.pre")
PDAC155698.integrated.subset$cancer <- c("pdac")

## Merge datasets

bcc.pre.pdac.merge <- merge(BCCIO.pre.subset, PDAC155698.integrated.subset, add.cell.ids = c("bcc.pre", "pdac"))
Idents(bcc.pre.pdac.merge) <- bcc.pdac.merge@meta.data$seurat_clusters

bcc.pre.pdac.merge@meta.data <- bcc.pre.pdac.merge@meta.data[, -which(colnames(bcc.pre.pdac.merge@meta.data) %in% c('cell', 'cell.names', 'sub_cluster'))]
bcc.pre.pdac.merge$orig.clusters <- bcc.pre.pdac.merge$seurat_clusters

## Save merged dataset

save(bcc.pre.pdac.merge, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.pre.PDAC.merge")

```


Enhanced Volcano plots for BCC pretreatment (responder) vs. PDAC

```{r}

## Subset dataset to only include BCC pretreatment repsonders

Idents(bcc.pre.pdac.merge) <- bcc.pre.pdac.merge$response
bcc.pre.pdac.merge$response[is.na(bcc.pre.pdac.merge$response)] <- "PDAC"
Idents(bcc.pre.pdac.merge) <- bcc.pre.pdac.merge$response

bcc.pre.responders.pdac.merge <- subset(bcc.pre.pdac.merge, idents = c("PDAC", "Responsive"))

interestingcelltypes <- c("T_proliferation", "Macrophages")

Idents(bcc.pre.responders.pdac.merge) <- bcc.pre.responders.pdac.merge$seurat_clusters

for(celltype in interestingcelltypes) {
  bcc.pdac.responder.volcano <- bioc_volcano(bcc.pre.responders.pdac.merge, celltype, c("response"), c("Responsive"), c("PDAC"))
  assign(paste0("bcc.pdac.", celltype, ".responder.volcano"), bcc.pdac.responder.volcano)
  bcc.pdac.responder.volcano
}

## Macrophages
`bcc.pdac.Macrophages.responder.volcano`

## Proliferated T cells
`bcc.pdac.T_proliferation.responder.volcano`


```



Enhanced Volcano plots for PDAC by patient vs. BCC responders

```{r}

bcc.pre.pdac.merge$patient[is.na(bcc.pre.pdac.merge$patient)] <- c("pdac")
Idents(bcc.pre.pdac.merge) <- bcc.pre.pdac.merge@meta.data$patient

bcc.pre.responders.pdac.merge <- subset(bcc.pre.pdac.merge, idents=c("pdac", "su001", "su002", "su003", "su004", "su009", "su012"))

## Add PDAC patient column

bcc.pre.responders.pdac.merge$pdac.patient <- as.vector(bcc.pre.responders.pdac.merge$orig.ident)

bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre.tumor.cd45'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su002.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su003.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su004.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su004.pre.tumor.cd45'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su009.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su012.pre.tcell'] <- 'BCC'

## Create separate objects for each PDAC patient

Idents(bcc.pre.responders.pdac.merge) <- bcc.pre.responders.pdac.merge$pdac.patient

bcc.pdac.1 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.1', 'BCC'))
bcc.pdac.2 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.2', 'BCC'))
bcc.pdac.3 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.3', 'BCC'))
bcc.pdac.4 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.4', 'BCC'))
bcc.pdac.5 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.5', 'BCC'))
bcc.pdac.6 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.6', 'BCC'))
bcc.pdac.7 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.7', 'BCC'))
bcc.pdac.8 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.8', 'BCC'))
bcc.pdac.9 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.9', 'BCC'))
bcc.pdac.10 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.10', 'BCC'))
bcc.pdac.11A <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.11A', 'BCC'))
bcc.pdac.11B <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.11B', 'BCC'))
bcc.pdac.12 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.12', 'BCC'))
bcc.pdac.13 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.13', 'BCC'))
bcc.pdac.15 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.15', 'BCC'))
bcc.pdac.16 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.16', 'BCC'))

Idents(bcc.pdac.1) <- bcc.pdac.1$seurat_clusters
Idents(bcc.pdac.2) <- bcc.pdac.2$seurat_clusters
Idents(bcc.pdac.3) <- bcc.pdac.3$seurat_clusters
Idents(bcc.pdac.4) <- bcc.pdac.4$seurat_clusters
Idents(bcc.pdac.5) <- bcc.pdac.5$seurat_clusters
Idents(bcc.pdac.6) <- bcc.pdac.6$seurat_clusters
Idents(bcc.pdac.7) <- bcc.pdac.7$seurat_clusters
Idents(bcc.pdac.8) <- bcc.pdac.8$seurat_clusters
Idents(bcc.pdac.9) <- bcc.pdac.9$seurat_clusters
Idents(bcc.pdac.10) <- bcc.pdac.10$seurat_clusters
Idents(bcc.pdac.11A) <- bcc.pdac.11A$seurat_clusters
Idents(bcc.pdac.11B) <- bcc.pdac.11B$seurat_clusters
Idents(bcc.pdac.12) <- bcc.pdac.12$seurat_clusters
Idents(bcc.pdac.13) <- bcc.pdac.13$seurat_clusters
Idents(bcc.pdac.15) <- bcc.pdac.15$seurat_clusters
Idents(bcc.pdac.16) <- bcc.pdac.16$seurat_clusters

## Enhanced Volcano Plots (for loop doesn't work)

bcc.pdac.1.volcano <- bioc_volcano(bcc.pdac.1, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 1"))
bcc.pdac.2.volcano <- bioc_volcano(bcc.pdac.2, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 2"))
bcc.pdac.3.volcano <- bioc_volcano(bcc.pdac.3, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 3"))
bcc.pdac.4.volcano <- bioc_volcano(bcc.pdac.4, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 4"))
bcc.pdac.5.volcano <- bioc_volcano(bcc.pdac.5, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 5"))
bcc.pdac.6.volcano <- bioc_volcano(bcc.pdac.6, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 6"))
bcc.pdac.7.volcano <- bioc_volcano(bcc.pdac.7, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 7"))
bcc.pdac.8.volcano <- bioc_volcano(bcc.pdac.8, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 8"))
bcc.pdac.9.volcano <- bioc_volcano(bcc.pdac.9, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 9"))
bcc.pdac.10.volcano <- bioc_volcano(bcc.pdac.10, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 10"))
bcc.pdac.11A.volcano <- bioc_volcano(bcc.pdac.11A, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 11A"))
bcc.pdac.11B.volcano <- bioc_volcano(bcc.pdac.11B, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 11B"))
bcc.pdac.12.volcano <- bioc_volcano(bcc.pdac.12, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 12"))
bcc.pdac.13.volcano <- bioc_volcano(bcc.pdac.13, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 13"))
bcc.pdac.15.volcano <- bioc_volcano(bcc.pdac.15, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 15"))
bcc.pdac.16.volcano <- bioc_volcano(bcc.pdac.16, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 16"))

```



Check DE gene expression of HLA genes by patient 

```{r}
## Differentially expressed genes
`
bcc.pdac.subsets <- list(bcc.pdac.1, bcc.pdac.2, bcc.pdac.3, bcc.pdac.4, bcc.pdac.5, bcc.pdac.6, bcc.pdac.7, bcc.pdac.8, bcc.pdac.9, bcc.pdac.10, bcc.pdac.11A, bcc.pdac.11B, bcc.pdac.12, bcc.pdac.13, bcc.pdac.15, bcc.pdac.16)

bcc.pdac.markers <- lapply(bcc.pdac.subsets, function(i) bioc_markers(i, c("Macrophages"), c("pdac.patient"), c("BCC"), pCutoff = 1, FCcutoff = 0.01))

for (i in 1:length(bcc.pdac.markers)) {
  assign(paste0("bcc.pdac.", i, ".markers"), bcc.pdac.markers[[i]])
}

bcc.pdac.11A.markers <- bcc.pdac.11.markers
bcc.pdac.11B.markers <- bcc.pdac.12.markers
bcc.pdac.12.markers <- bcc.pdac.13.markers
bcc.pdac.13.markers <- bcc.pdac.14.markers

## Segregate the HLA genes

bcc.pdac.markers <- list(bcc.pdac.1.markers, bcc.pdac.2.markers, bcc.pdac.3.markers, bcc.pdac.4.markers, bcc.pdac.5.markers, bcc.pdac.6.markers, bcc.pdac.7.markers, bcc.pdac.8.markers, bcc.pdac.9.markers, bcc.pdac.10.markers, bcc.pdac.11A.markers, bcc.pdac.11B.markers, bcc.pdac.12.markers, bcc.pdac.13.markers, bcc.pdac.15.markers, bcc.pdac.16.markers)

patient.ids <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11A", "11B", "12", "13", "15", "16")

hla.markers <- NULL

for (i in 1:length(bcc.pdac.markers)) {
  bcc.pdac.markers[[i]]$gene.header <- substr(bcc.pdac.markers[[i]]$GeneNames, 0, 4)
  bcc.pdac.markers[[i]]$patient <- (patient.ids[i])
  hla.markers[[i]] <- subset(bcc.pdac.markers[[i]], gene.header == "HLA-")
}

for (i in 1:length(hla.markers)) {
  hla.markers[[i]]$patient
  assign(paste0("hla.markers.", i), hla.markers[[i]])
}

## Merge datasets together

hla.markers <- list(hla.markers.1, hla.markers.2, hla.markers.3, hla.markers.4, hla.markers.5, hla.markers.6, hla.markers.7, hla.markers.8, hla.markers.9, hla.markers.10, hla.markers.11, hla.markers.12, hla.markers.13, hla.markers.14, hla.markers.15, hla.markers.16)

patient.ids <- c('1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11A', '11B', '12', '13', '15', '16')

for (i in 1:length(hla.markers)) {
  hla.markers[[i]]$patient <- patient.ids[i]
}

hla.markers.by.patient <- do.call("rbind", hla.markers)

## Plot results

hla.markers.by.patient$log_pvaladj <- -log(hla.markers.by.patient$p_val_adj)
#hla.markers.by.patient$loglog_pvaladj <- log(hla.markers.by.patient$log_pvaladj + 1)

hla.markers.plot <- ggplot(hla.markers.by.patient, aes(x=log_pvaladj, y=avg_log2FC, color=GeneNames, size=1)) + theme_minimal() + ggtitle("Differential Expression of HLA genes by patient: \n BCC Responders vs. PDAC") + xlab("-log10(P)") + ylab("Log2 fold change") + geom_point() + theme(plot.title = element_text(size=20), legend.text = element_text(size = 10), legend.title = element_text(size = 14))
hla.markers.plot

```

Result: HLA-DRB5 and HLA-G have the greatest difference in response

Look at difference in HLA expression between BCC nonresponders and responders pretreatment 

```{r}

Idents(BCCIO) <- BCCIO@meta.data$treatment
BCCIO.pre <- subset(BCCIO, idents=c("pre"))

Idents(BCCIO.pre) <- BCCIO.pre$patient
bcc.pre.responder.5 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su005"))
bcc.pre.responder.6 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su006"))
bcc.pre.responder.7 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su007"))
bcc.pre.responder.8 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su008"))
bcc.pre.responder.10 <- subset(BCCIO.pre, idents=c("su001", "su002", "su003", "su004", "su009", "su012", "su010"))

Idents(bcc.pre.responder.5) <- bcc.pre.responder.5$seurat_clusters
Idents(bcc.pre.responder.6) <- bcc.pre.responder.6$seurat_clusters
Idents(bcc.pre.responder.7) <- bcc.pre.responder.7$seurat_clusters
Idents(bcc.pre.responder.8) <- bcc.pre.responder.8$seurat_clusters
Idents(bcc.pre.responder.10) <- bcc.pre.responder.10$seurat_clusters


## Enhanced volcano plots by patient

bcc.pre.response.volcano.5 <- bioc_volcano(bcc.pre.responder.5, c("Macrophages"), c("patient"), c("su005"), c("Responders"))
bcc.pre.response.volcano.6 <- bioc_volcano(bcc.pre.responder.6, c("Macrophages"), c("patient"), c("su006"), c("Responders"))
bcc.pre.response.volcano.7 <- bioc_volcano(bcc.pre.responder.7, c("Macrophages"), c("patient"), c("su007"), c("Responders"))
bcc.pre.response.volcano.8 <- bioc_volcano(bcc.pre.responder.8, c("Macrophages"), c("patient"), c("su008"), c("Responders"))
bcc.pre.response.volcano.10 <- bioc_volcano(bcc.pre.responder.10, c("Macrophages"), c("patient"), c("su010"), c("Responders")) # Doesn't work for some reason (very little cells anyways)

bcc.pre.response.volcano.5
bcc.pre.response.volcano.6
bcc.pre.response.volcano.7
bcc.pre.response.volcano.8

## Differentially expressed genes by patient

bcc.pre.response.markers.5 <- bioc_markers(bcc.pre.responder.5, c("Macrophages"), c("patient"), c("su005")) 
bcc.pre.response.markers.6 <- bioc_markers(bcc.pre.responder.6, c("Macrophages"), c("patient"), c("su006")) 
bcc.pre.response.markers.7 <- bioc_markers(bcc.pre.responder.7, c("Macrophages"), c("patient"), c("su007")) 
bcc.pre.response.markers.8 <- bioc_markers(bcc.pre.responder.8, c("Macrophages"), c("patient"), c("su008")) 

bcc.pre.response.markers <- list(bcc.pre.response.markers.5, bcc.pre.response.markers.6, bcc.pre.response.markers.7, bcc.pre.response.markers.8)

patient.ids <- c("su005", "su006", "su007", "su008")
bcc.hla.markers <- NULL

for (i in 1:length(bcc.pre.response.markers)) {
  bcc.pre.response.markers[[i]]$gene.header <- substr(bcc.pre.response.markers[[i]]$GeneNames, 0, 4)
  bcc.pre.response.markers[[i]]$patient <- (patient.ids[i])
  bcc.hla.markers[[i]] <- subset(bcc.pdac.markers[[i]], gene.header == "HLA-")
}

for (i in 1:length(bcc.hla.markers)) {
  assign(paste0("bcc.hla.", patient.ids[i], ".markers"), bcc.hla.markers[[i]])
}

bcc.hla.markers <- list(bcc.hla.su005.markers, bcc.hla.su006.markers, bcc.hla.su007.markers, bcc.hla.su008.markers)
bcc.hla.markers <- do.call("rbind", bcc.hla.markers)

## Plot results

bcc.hla.markers$log_pvaladj <- -log(bcc.hla.markers$p_val_adj)
#hla.markers.by.patient$loglog_pvaladj <- log(hla.markers.by.patient$log_pvaladj + 1)

bcc.hla.markers.plot <- ggplot(bcc.hla.markers, aes(x=log_pvaladj, y=avg_log2FC, color=GeneNames, size=1.5)) + theme_minimal() + ggtitle("Differential Expression of HLA genes by patient: \n BCC Responders vs. Nonresponders") + xlab("-log10(P)") + ylab("Log2 fold change") + geom_point() + theme(plot.title = element_text(size=20), legend.text = element_text(size=10), legend.title = element_text(size = 14))
bcc.hla.markers.plot



```

Result: HLA-DRB5 has greatest differential expression -- check to make sure same direction 

Violin Plots for Gene Expression 

```{r}

bcc.pdac.merge$patient[is.na(bcc.pdac.merge$patient)] <- "pdac"
Idents(bcc.pdac.merge) <- bcc.pdac.merge@meta.data$patient

Idents(BCCIO) <- BCCIO@meta.data$patient

HLADRB5.bcc.vlnplot <- VlnPlot(BCCIO, "HLA-DRB5")
HLADRB5.bcc.vlnplot

HLADRB5.pdac.vlnplot <- VlnPlot(PDAC155698.integrated, "HLA-DRB5")

View(PDAC155698.integrated@meta.data)

## Violin plots

Idents(BCCIO) <- BCCIO@meta.data$seurat_clusters

bcc.macrophages <- subset(BCCIO, idents = c("Macrophages"))
Idents(bcc.macrophages) <- bcc.macrophages@meta.data$patient


HLADRB5.bcc.macrophages.vlnplot <- VlnPlot(bcc.macrophages, "HLA-DRB5")
HLADRB5.bcc.macrophages.vlnplot

## Average expression for BCC

set.seed(001)

Idents(BCCIO) <- BCCIO@meta.data$patient
BCCIO.subset <- BCCIO[, sample(colnames(BCCIO), size=5000, replace=F)]

BCCIO.averages <- AverageExpression(BCCIO.subset)

head(BCCIO.averages[["RNA"]][c("HLA-DRB5"),])

## Average expression for PDAC

Idents(PDAC155698.integrated) <- PDAC155698.integrated@meta.data$orig.ident
PDAC.averages <- AverageExpression(PDAC155698.integrated)

head(PDAC.averages[["RNA"]][c("HLA-DRB5"),])

```



Conduct ANOVA tests

```{r}

## Create Seurat object with all macrophages from both datasets

Idents(BCCIO.pre) <- BCCIO.pre$seurat_clusters
bcc.macrophages <- subset(BCCIO.pre, idents = c("Macrophages"))

Idents(PDAC155698.integrated) <- PDAC155698.integrated$seurat_clusters
pdac.macrophages <- subset(PDAC155698.integrated, idents= c("Macrophages"))

bcc.pdac.macrophages <- merge(bcc.macrophages, pdac.macrophages, add.cell.ids = c("bcc", "pdac"))

## Average gene expression per patient for HLA 

hla.genes <- unique(hla.markers.by.patient$GeneNames)

Idents(bcc.pdac.macrophages) <- bcc.pdac.macrophages$patient
bcc.pdac.averages <- AverageExpression(bcc.pdac.macrophages)
hla.averages <- bcc.pdac.averages[["RNA"]][hla.genes,]

hla.averages <- rbind(hla.averages, c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC"))

rownames(hla.averages)[rownames(hla.averages) == "18"] <- "category"
hla.averages <- as.data.frame(t(hla.averages))

for (i in 1:18) {
  hla.averages[,i] <- as.character(hla.averages[,i])
  hla.averages[,i] <- as.numeric(hla.averages[,i])
}


save(hla.averages, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\hla.averages")

## ANOVA test for all HLA genes at 95% confidence

hla.anova.plots <- vector("list", length=17)

for (i in 1:length(hla.genes)) {
  colnames(hla.averages)[i] <- "gene"
  plot <- anova_test(hla.averages, hla.genes[i], "Category")
  hla.anova.plots[[i]] <- recordPlot()
  assign(paste0("hla.anova.plots.", hla.genes[i]), hla.anova.plots[[i]])
  plot.new()
  colnames(hla.averages)[i] <- hla.genes[i]
}

names(hla.anova.plots) <- hla.genes

hla.anova.plots <- list(hla.anova.plots[[1]], hla.anova.plots[[2]], hla.anova.plots[[3]], hla.anova.plots[[4]], hla.anova.plots[[5]], hla.anova.plots[[6]], hla.anova.plots[[7]], hla.anova.plots[[8]], hla.anova.plots[[9]], hla.anova.plots[[10]], hla.anova.plots[[11]], hla.anova.plots[[12]], hla.anova.plots[[13]], hla.anova.plots[[14]], hla.anova.plots[[15]], hla.anova.plots[[16]], hla.anova.plots[[17]])

grid.arrange(`hla.anova.plotsHLA-A`, `hla.anova.plotsHLA-B`)

## ANOVA test for all HLA genes at 90% confidence

hla.anova.90.plots <- vector("list", length=17)

for (i in 1:length(hla.genes)) {
  colnames(hla.averages)[i] <- "gene"
  plot <- anova_test(hla.averages, hla.genes[i], "Category", confidence = 0.90)
  hla.anova.90.plots[[i]] <- recordPlot()
  plot.new()
  colnames(hla.averages)[i] <- hla.genes[i]
}

names(hla.anova.90.plots) <- hla.genes

```

Result: no significant different in mean expression between BCC responders and nonresponders for any HLA genes individually

Analyze MHC I/II genes together 

```{r}

## MHC Class 1: HLA-A, HLA-B, HLA-C, HLA-E, HLA-G
## MHC Class 2: Everything else (HLA-D__)

## Arithmetic average score 

hla.averages.score <- data.frame(matrix(ncol=2, nrow=26))
colnames(hla.averages.score) <- c("MHC1", "MHC2")
rownames(hla.averages.score) <- rownames(hla.averages)

hla.averages.score$MHC1 <- (hla.averages$`HLA-A` + hla.averages$`HLA-B`+ hla.averages$`HLA-C` + hla.averages$`HLA-E` + hla.averages$`HLA-G`) / 5

hla.averages.score$MHC2 <- (hla.averages$`HLA-DOA`+ hla.averages$`HLA-DQB2` + hla.averages$`HLA-DMA` + hla.averages$`HLA-DQA2` + hla.averages$`HLA-DRA` + hla.averages$`HLA-DPA1` + hla.averages$`HLA-DQB1` + hla.averages$`HLA-DPB1` + hla.averages$`HLA-DRB1` + hla.averages$`HLA-DRB5` + hla.averages$`HLA-DQA1` + hla.averages$`HLA-DMB`) / 12

hla.averages.score$category <- hla.averages$category

hla.scores <- melt(hla.averages.score, id.var="category")
ggplot(data=hla.scores, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

## Normalized arithmetic average score

hla.averages.normalized <- hla.averages %>% mutate_at(hla.genes, ~(scale(.) %>% as.vector))

hla.averages.score.norm <- data.frame(matrix(ncol=2, nrow=26))
colnames(hla.averages.score.norm) <- c("MHC1", "MHC2")
rownames(hla.averages.score.norm) <- rownames(hla.averages.normalized)

hla.averages.score.norm$MHC1 <- (hla.averages.normalized$`HLA-A` + hla.averages.normalized$`HLA-B`+ hla.averages.normalized$`HLA-C` + hla.averages.normalized$`HLA-E` + hla.averages.normalized$`HLA-G`) / 5

hla.averages.score.norm$MHC2 <- (hla.averages.normalized$`HLA-DOA`+ hla.averages.normalized$`HLA-DQB2` + hla.averages.normalized$`HLA-DMA` + hla.averages.normalized$`HLA-DQA2` + hla.averages.normalized$`HLA-DRA` + hla.averages.normalized$`HLA-DPA1` + hla.averages.normalized$`HLA-DQB1` + hla.averages.normalized$`HLA-DPB1` + hla.averages.normalized$`HLA-DRB1` + hla.averages.normalized$`HLA-DRB5` + hla.averages.normalized$`HLA-DQA1` + hla.averages.normalized$`HLA-DMB`) / 12

hla.averages.score.norm$category <- hla.averages$category

hla.scores.norm <- melt(hla.averages.score.norm, id.var="category")
ggplot(data=hla.scores.norm, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

## Boxplots for each gene

hla.norm <- melt(hla.averages.normalized, id.var="category")
ggplot(data=hla.norm, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

## ANOVA Test on MHC Scores

mhc1.anova <- aov(MHC1~category, data=hla.averages.score)
mhc2.anova <- aov(MHC2~category, data=hla.averages.score)

mhc1.tukey <- TukeyHSD(mhc1.anova, conf.level = 0.95)
mhc2.tukey <- TukeyHSD(mhc2.anova, conf.level = 0.95)

tuk_plot(mhc1.tukey, title = "MHC I") + tuk_plot(mhc2.tukey, title = "MHC II")

## ANOVA Test on Normalized MHC Scores

mhc1.normalized.anova <- aov(MHC1~category, data=hla.averages.score.norm)
mhc2.normalized.anova <- aov(MHC2~category, data=hla.averages.score.norm)

mhc1.normalized.tukey <- TukeyHSD(mhc1.normalized.anova, conf.level=0.95)
mhc2.normalized.tukey <- TukeyHSD(mhc2.normalized.anova, conf.level=0.95)

tuk_plot(mhc1.normalized.tukey, title = "MHC I") + tuk_plot(mhc2.normalized.tukey, title = "MHC II")



```

Result: NR and PDAC have lower scores on MHC I compared to R; differences inconclusive for MHC II
Function of MHC Class I genes: display intracellular proteins to cytotoxic T cells (function to kill cancer cells)

Check MHC scores for malignant cell populations

```{r}

## Subset malignant cell populations for BCC, combine with all PDAC cells

Idents(BCCIO) <- BCCIO$seurat_clusters
bcc.malignant <- subset(BCCIO, idents = c("Macrophages", "Tumor 1", "Tumor 2"))

Idents(PDAC155698.integrated) <- PDAC155698.integrated$seurat_clusters
PDAC.subset <- subset(PDAC155698.integrated, idents = c("Acinar cells", "Epithelial + Endocrine"))
PDAC.subset$response <- "PDAC"

bcc.pdac.malignant <- merge(bcc.malignant, PDAC.subset, add.cell.ids = c("bcc", "pdac"))
bcc.pdac.malignant$category <- paste0(bcc.pdac.malignant$response, " ", bcc.pdac.malignant$seurat_clusters)


## Average gene expression per patient for HLA 

Idents(bcc.pdac.malignant) <- bcc.pdac.malignant$patient
bcc.pdac.malignant.averages <- AverageExpression(bcc.pdac.malignant)
hla.malignant.averages <- bcc.pdac.malignant.averages[["RNA"]][hla.genes,]

rownames(hla.malignant.averages)[rownames(hla.malignant.averages) == "18"] <- "category"
hla.malignant.averages <- as.data.frame(t(hla.malignant.averages))

for (i in 1:17) {
  hla.malignant.averages[,i] <- as.character(hla.malignant.averages[,i])
  hla.malignant.averages[,i] <- as.numeric(hla.malignant.averages[,i])
}

hla.malignant.averages <- cbind(hla.malignant.averages, c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC"))

colnames(hla.malignant.averages)[18] <- "category"

save(hla.malignant.averages, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\hla.malignant.averages")

## MHC Class I Score 

hla.malignant.averages.score <- data.frame(matrix(ncol=2, nrow=27))
colnames(hla.malignant.averages.score) <- c("MHC1", "MHC2")
rownames(hla.malignant.averages.score) <- rownames(hla.malignant.averages)[1:27]

hla.malignant.averages.score$MHC1 <- (hla.malignant.averages$`HLA-A` + hla.malignant.averages$`HLA-B`+ hla.malignant.averages$`HLA-C` + hla.malignant.averages$`HLA-E` + hla.malignant.averages$`HLA-G`) / 5

hla.malignant.averages.score$MHC2 <- (hla.malignant.averages$`HLA-DOA`+ hla.malignant.averages$`HLA-DQB2` + hla.malignant.averages$`HLA-DMA` + hla.malignant.averages$`HLA-DQA2` + hla.malignant.averages$`HLA-DRA` + hla.malignant.averages$`HLA-DPA1` + hla.malignant.averages$`HLA-DQB1` + hla.malignant.averages$`HLA-DPB1` + hla.malignant.averages$`HLA-DRB1` + hla.malignant.averages$`HLA-DRB5` + hla.malignant.averages$`HLA-DQA1` + hla.malignant.averages$`HLA-DMB`) / 12

hla.malignant.averages.score$category <- c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC")

hla.malignant.scores <- melt(hla.malignant.averages.score, id.var="category")
ggplot(data=hla.malignant.scores, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

## ANOVA tests

mhc1.malignant.anova <- aov(MHC1~category, data=hla.malignant.averages.score)
mhc2.malignant.anova <- aov(MHC2~category, data=hla.malignant.averages.score)

mhc1.malignant.tukey <- TukeyHSD(mhc1.malignant.anova, conf.level = 0.95)
mhc2.malignant.tukey <- TukeyHSD(mhc2.malignant.anova, conf.level = 0.95)

tuk_plot(mhc1.malignant.tukey, title = "MHC I") + tuk_plot(mhc2.malignant.tukey, title = "MHC II")

mhc1.malignant.tukey
mhc2.malignant.tukey

```

Result: MHC Class I and Class II lower in PDAC vs. responders and nonresponders 
MHC Class II in PDAC --> essentially no expression, much lower than responders or nonresponders



HSP Genes Differential Expression: Responders vs. PDAC 

```{r}

## Segregate the HSP genes

bcc.pdac.markers <- list(bcc.pdac.1.markers, bcc.pdac.2.markers, bcc.pdac.3.markers, bcc.pdac.4.markers, bcc.pdac.5.markers, bcc.pdac.6.markers, bcc.pdac.7.markers, bcc.pdac.8.markers, bcc.pdac.9.markers, bcc.pdac.10.markers, bcc.pdac.11A.markers, bcc.pdac.11B.markers, bcc.pdac.12.markers, bcc.pdac.13.markers, bcc.pdac.15.markers, bcc.pdac.16.markers)

patient.ids <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11A", "11B", "12", "13", "15", "16")

hsp.markers <- NULL

for (i in 1:length(bcc.pdac.markers)) {
  bcc.pdac.markers[[i]]$gene.header <- substr(bcc.pdac.markers[[i]]$GeneNames, 0, 3)
  bcc.pdac.markers[[i]]$patient <- (patient.ids[i])
  hsp.markers[[i]] <- subset(bcc.pdac.markers[[i]], gene.header == "HSP")
}

for (i in 1:length(hsp.markers)) {
  hsp.markers[[i]]$patient
  assign(paste0("hsp.markers.", i), hsp.markers[[i]])
}

## Merge datasets together

hsp.markers <- list(hsp.markers.1, hsp.markers.2, hsp.markers.3, hsp.markers.4, hsp.markers.5, hsp.markers.6, hsp.markers.7, hsp.markers.8, hsp.markers.9, hsp.markers.10, hsp.markers.11, hsp.markers.12, hsp.markers.13, hsp.markers.14, hsp.markers.15, hsp.markers.16)

hsp.markers.by.patient <- do.call("rbind", hsp.markers)

## Plot results

hsp.markers.by.patient$log_pvaladj <- -log(hsp.markers.by.patient$p_val_adj)

hsp.markers.plot <- ggplot(hsp.markers.by.patient, aes(x=log_pvaladj, y=avg_log2FC, color=GeneNames, size=1)) + theme_minimal() + ggtitle("Differential Expression of HSP genes by patient: \n BCC Responders vs. PDAC") + xlab("-log10(P)") + ylab("Log2 fold change") + geom_point() + theme(plot.title = element_text(size=20), legend.text = element_text(size = 10), legend.title = element_text(size = 14))
hsp.markers.plot

```



HSP Genes Differential Expression: Responders vs. Nonresponders

```{r}

## Differentially expressed genes by patient

patient.ids <- c("su005", "su006", "su007", "su008")
bcc.hsp.markers <- NULL

for (i in 1:length(bcc.pre.response.markers)) {
  bcc.pre.response.markers[[i]]$gene.header <- substr(bcc.pre.response.markers[[i]]$GeneNames, 0, 3)
  bcc.pre.response.markers[[i]]$patient <- (patient.ids[i])
  bcc.hsp.markers[[i]] <- subset(bcc.pre.response.markers[[i]], gene.header == "HSP")
}



for (i in 1:length(bcc.hsp.markers)) {
  assign(paste0("bcc.hsp.", patient.ids[i], ".markers"), bcc.hsp.markers[[i]])
}

bcc.hsp.markers <- list(bcc.hsp.su005.markers, bcc.hsp.su006.markers, bcc.hsp.su007.markers, bcc.hsp.su008.markers)
bcc.hsp.markers <- do.call("rbind", bcc.hsp.markers)

## Plot results

bcc.hsp.markers$log_pvaladj <- -log(bcc.hsp.markers$p_val_adj)

bcc.hsp.markers.plot <- ggplot(bcc.hsp.markers, aes(x=log_pvaladj, y=avg_log2FC, color=GeneNames, size=1.5)) + theme_minimal() + ggtitle("Differential Expression of HSP genes by patient: \n BCC Responders vs. Nonresponders") + xlab("-log10(P)") + ylab("Log2 fold change") + geom_point() + theme(plot.title = element_text(size=20), legend.text = element_text(size=10), legend.title = element_text(size = 14))
bcc.hsp.markers.plot

```

Result: no clear trends, opposite direction from responder vs. PDAC

HSP Score Analysis

```{r}

## Average gene expression per patient for HSP

hsp.genes <- unique(hsp.markers.by.patient$GeneNames)

Idents(bcc.pdac.macrophages) <- bcc.pdac.macrophages$patient
bcc.pdac.averages <- AverageExpression(bcc.pdac.macrophages)
hsp.averages <- bcc.pdac.averages[["RNA"]][hsp.genes,]

hsp.averages <- rbind(hsp.averages, c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC"))

rownames(hsp.averages)[rownames(hsp.averages) == "15"] <- "category"
hsp.averages <- as.data.frame(t(hsp.averages))

for (i in 1:14) {
  hsp.averages[,i] <- as.character(hsp.averages[,i])
  hsp.averages[,i] <- as.numeric(hsp.averages[,i])
}

colnames(hsp.averages)[15] <- "category"

save(hsp.averages, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\hsp.averages")

## Calculate HSP score (arithmetic average)

hsp.averages.score <- data.frame(matrix(ncol=2, nrow=25))
colnames(hsp.averages.score) <- c("HSP", "category")
rownames(hsp.averages.score) <- rownames(hsp.averages)

hsp.averages.score$HSP <- rowSums(hsp.averages[,-15]) / 14

hsp.averages.score$category <- hsp.averages$category

hsp.scores <- melt(hsp.averages.score, id.var="category")
ggplot(data=hsp.scores, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

hsp.anova <- aov(HSP~category, data=hsp.averages.score)
hsp.tukey <- TukeyHSD(hsp.anova, conf.level = 0.95)

tuk_plot(hsp.tukey, title = "HSP")

```



HSP Score Analysis: Malignant Cells

```{r}

## Average gene expression for HSP malignant cells

hsp.genes <- unique(hsp.markers.by.patient$GeneNames)

Idents(bcc.pdac.malignant) <- bcc.pdac.malignant$patient
bcc.pdac.malignant.averages <- AverageExpression(bcc.pdac.malignant)
hsp.malignant.averages <- bcc.pdac.malignant.averages[["RNA"]][hsp.genes,]

hsp.malignant.averages <- rbind(hsp.malignant.averages, c("R", "R", "R", "R", "NR", "NR", "NR", "NR", "R", "NR", "R", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC", "PDAC"))

rownames(hsp.malignant.averages)[rownames(hsp.malignant.averages) == "18"] <- "category"
hsp.malignant.averages <- as.data.frame(t(hsp.malignant.averages))

for (i in 1:17) {
  hsp.malignant.averages[,i] <- as.character(hsp.malignant.averages[,i])
  hsp.malignant.averages[,i] <- as.numeric(hsp.malignant.averages[,i])
}

colnames(hsp.malignant.averages)[18] <- "category"

save(hsp.malignant.averages, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\hsp.malignant.averages")

## Calculate HSP score (arithmetic average)

hsp.malignant.averages.score <- data.frame(matrix(ncol=2, nrow=27))
colnames(hsp.malignant.averages.score) <- c("HSP", "category")
rownames(hsp.malignant.averages.score) <- rownames(hsp.malignant.averages)

hsp.malignant.averages.score$HSP <- rowSums(hsp.malignant.averages[,-18]) / 17

hsp.malignant.averages.score$category <- hsp.malignant.averages$category

hsp.malignant.scores <- melt(hsp.malignant.averages.score, id.var="category")
ggplot(data=hsp.malignant.scores, aes(x=variable, y=value)) + geom_boxplot(aes(fill=category))

hsp.malignant.anova <- aov(HSP~category, data=hsp.malignant.averages.score)
hsp.malignant.tukey <- TukeyHSD(hsp.malignant.anova, conf.level = 0.95)

tuk_plot(hsp.malignant.tukey, title = "HSP")

```

##################################################################################################################

Compare pretreatment PDAC to BCC responders and nonresponders, pretreatment vs. posttreatment

```{r}

## Possible comparisons: CD8+ T cells, CD4+ T cells, Tregs, B cells, Macrophages, NK cells

BCC.PDAC.comparison.CD8 <- fiveway_diffexpression_MHCandHSP(c("CD8_activation", "CD8_exhausted", "CD8_memory"), "CD8+ T cells", "CD8+ T Cells")
BCC.PDAC.comparison.CD4 <- fiveway_diffexpression_MHCandHSP("CD4", "CD4+ T cells", "CD4+ T Cells")
BCC.PDAC.comparison.Tregs <- fiveway_diffexpression_MHCandHSP("Tregs", "Tregs", "Regulatory T Cells")
BCC.PDAC.comparison.Macrophages <- fiveway_diffexpression_MHCandHSP("Macrophages", "Macrophages", "Macrophages")
BCC.PDAC.comparison.B <- fiveway_diffexpression_MHCandHSP("B cells", "B cells","B Cells")
BCC.PDAC.comparison.NK <- fiveway_diffexpression_MHCandHSP("NK", "NK cells", "Natural Killer Cells")

## Use grid.draw() to see the final product 

```


Look at expression of PD-1 pathway again

```{r}

Idents(PDAC155698.integrated) <- PDAC155698.integrated$seurat_clusters

PDAC.averages <- AverageExpression(PDAC155698.integrated)[["RNA"]]
PDAC.genes <- rownames(PDAC.averages)

PDAC.averages.PD1 <- PDAC.averages[c("PDCD1", "CD274", "PDCD1LG2"),]

Idents(BCC.responder.pre) <- BCC.responder.pre$seurat_clusters
Idents(BCC.responder.post) <- BCC.responder.post$seurat_clusters
Idents(BCC.nonresponder.pre) <- BCC.nonresponder.pre$seurat_clusters
Idents(BCC.nonresponder.post) <- BCC.nonresponder.post$seurat_clusters

BCC.responder.pre.averages <- AverageExpression(BCC.responder.pre)[["RNA"]]
BCC.responder.post.averages <- AverageExpression(BCC.responder.post)[["RNA"]]
BCC.nonresponder.pre.averages <- AverageExpression(BCC.nonresponder.pre)[["RNA"]]
BCC.nonresponder.post.averages <- AverageExpression(BCC.nonresponder.post)[["RNA"]]

BCC.responder.pre.averages.PD1 <- BCC.responder.pre.averages[c("PDCD1", "CD274", "PDCD1LG2"),]
BCC.responder.post.averages.PD1 <- BCC.responder.post.averages[c("PDCD1", "CD274", "PDCD1LG2"),]
BCC.nonresponder.pre.averages.PD1 <- BCC.nonresponder.pre.averages[c("PDCD1", "CD274", "PDCD1LG2"),]
BCC.nonresponder.post.averages.PD1 <- BCC.nonresponder.post.averages[c("PDCD1", "CD274", "PDCD1LG2"),]



```



####################################################################################################################

Complete analysis using reclustered PDAC dataset

Compare malignant cells between PDAC and BCC 

```{r}

## Create merged BCC and PDAC object

Idents(BCCIO) <- BCCIO$seurat_clusters
Idents(PDAC.integrated) <- PDAC.integrated$seurat_clusters

PDAC.malignant <- subset(PDAC.integrated, idents=c("Ductal cells 1", "Ductal cells 2", "Ductal cells 3", "Ductal cells 4"))

BCC.malignant <- subset(BCCIO, idents=c("Tumor 1", "Tumor 2"))

BCC.PDAC.malignant <- merge(PDAC.malignant, BCC.malignant, add.cell.ids = c("PDAC", "BCC"))

BCC.PDAC.malignant@meta.data$sort <- NULL
BCC.PDAC.malignant@meta.data$RNA_snn_res.0.5 <- NULL
BCC.PDAC.malignant@meta.data$RNA_snn_res.0.4 <- NULL

BCC.PDAC.malignant@meta.data$response[is.na(BCC.PDAC.malignant@meta.data$response)] <- "PDAC"
BCC.PDAC.malignant@meta.data$treatment[is.na(BCC.PDAC.malignant@meta.data$treatment)] <- "PDAC"
BCC.PDAC.malignant@meta.data$celltype <- "Malignant cells"

## Enhanced volcano plot comparing PDAC to all BCC 

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$celltype

BCC.PDAC.malignant.volcano <- bioc_volcano(BCC.PDAC.malignant, celltype = "Malignant cells", group = "response", ident = "PDAC", ident2 = "BCC", title = "PDAC vs. BCC (All)")

## Enhanced volcano plot comparing PDAC to all BCC pretreatment

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$treatment

BCC.PDAC.malignant.pre <- subset(BCC.PDAC.malignant, idents = c("pre", "PDAC"))

Idents(BCC.PDAC.malignant.pre) <- BCC.PDAC.malignant.pre$celltype 

BCC.PDAC.malignant.pre.volcano <- bioc_volcano(BCC.PDAC.malignant.pre, celltype = "Malignant cells", group = "response", ident = "PDAC", ident2 = "BCC", title = "PDAC vs. BCC Pretreatment")

## Enhanced volcano plot comparing PDAC to BCC responders pretreatment

Idents(BCC.PDAC.malignant.pre) <- BCC.PDAC.malignant.pre$response

BCC.PDAC.malignant.preresponder <- subset(BCC.PDAC.malignant.pre, idents = c("Responsive", "PDAC"))

BCC.PDAC.malignant.preresponder.volcano <- bioc_volcano(BCC.PDAC.malignant.preresponder, celltype = "Malignant cells", group = "response", ident = "PDAC", ident2 = "BCC", title = "PDAC vs. BCC Responders Pretreatment")

```



Calculate MHC Class I scores for PDAC and BCC

```{r}

## Calculate scores by patient

Idents(BCC.PDAC.malignant.pre) <- BCC.PDAC.malignant.pre$response
Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$response


## Dot plots (by gene, by cluster)

mhc.genes <- c('HLA-A', 'HLA-B', 'HLA-C', 'HLA-E', 'HLA-F', 'HLA-G', 'HLA-DMA', 'HLA-DMB', 'HLA-DOA', 'HLA-DPA1', 'HLA-DPB1', 'HLA-DQA1', 'HLA-DQA2', 'HLA-DQB1', 'HLA-DQB2', 'HLA-DRA', 'HLA-DRB1', 'HLA-DRB5')

BCC.PDAC.malignant.expression <- AverageExpression(BCC.PDAC.malignant)
BCC.PDAC.malignant.mhc.expression <- BCC.PDAC.malignant.expression[["RNA"]][mhc.genes,]

BCC.PDAC.malignant.mhc.expression$Gene <- rownames(BCC.PDAC.malignant.mhc.expression)
BCC.PDAC.malignant.mhc.expression <- melt(BCC.PDAC.malignant.mhc.expression, id.vars="Gene")

BCC.PDAC.malignant.mhc.dotplot <- ggplot(BCC.PDAC.malignant.mhc.expression, aes(x=Gene, y=value, color=variable)) +
  geom_dotplot(binaxis='y', stackdir='center', binwidth=0) + 
  theme_light() +
  coord_flip() +
  labs(title = "Average Expression of MHC Genes", x = "Score", y = "Gene") +
  geom_point(aes(size=3))

BCC.PDAC.malignant.mhc.dotplot

## Dumbbell plots (by gene, by patient)

BCC.PDAC.malignant@meta.data$cancer <- BCC.PDAC.malignant@meta.data$response
Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$cancer

BCC.PDAC.malignant <- RenameIdents(object = BCC.PDAC.malignant,
                                   "PDAC" = "PDAC",
                                   "Responsive" = "BCC", 
                                   "Nonresponsive" = "BCC")
BCC.PDAC.malignant@meta.data$cancer <- Idents(BCC.PDAC.malignant)

BCC.PDAC.malignant.expression.cancer <- AverageExpression(BCC.PDAC.malignant)
BCC.PDAC.malignant.mhc.expression.cancer <- data.frame(t(BCC.PDAC.malignant.expression.cancer[["RNA"]][mhc.genes,]))
colnames(BCC.PDAC.malignant.mhc.expression.cancer) <- mhc.genes

BCC.PDAC.malignant.mhc.expression.cancer$Cancer <- rownames(BCC.PDAC.malignant.mhc.expression.cancer)
BCC.PDAC.malignant.mhc.expression.cancer <- melt(BCC.PDAC.malignant.mhc.expression.cancer, id.vars="Cancer")

BCC.PDAC.malignant.mhc.expression.cancer <- BCC.PDAC.malignant.mhc.expression.cancer %>% 
  mutate(paired = rep(1:(n()/2), each=2), Cancer=factor(Cancer))

BCC.PDAC.malignant.mhc.expression.cancer %>% ggplot(aes(x=value, y=variable)) +
  geom_line(aes(group=paired)) +
  geom_point(aes(color=Cancer), size=5) +
  theme(legend.position="top") +
  labs(title = "Average Expression of MHC Genes", x = "Score", y = "Gene") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

## Calculate MHC Class I and II scores

mhc1.genes <- c('HLA-A', 'HLA-B', 'HLA-C', 'HLA-E', 'HLA-F', 'HLA-G')
mhc2.genes <- c('HLA-DMA', 'HLA-DMB', 'HLA-DOA', 'HLA-DPA1', 'HLA-DPB1', 'HLA-DQA1', 'HLA-DQA2', 'HLA-DQB1', 'HLA-DQB2', 'HLA-DRA', 'HLA-DRB1', 'HLA-DRB5')

BCC.PDAC.malignant.mhc1.expression.cancer <- BCC.PDAC.malignant.expression.cancer[["RNA"]][mhc1.genes,]
BCC.PDAC.malignant.mhc2.expression.cancer <- BCC.PDAC.malignant.expression.cancer[["RNA"]][mhc2.genes,]

avg1 <- t(data.frame(colMeans(BCC.PDAC.malignant.mhc1.expression.cancer)))
BCC.PDAC.malignant.mhc1.expression.cancer <- rbind(BCC.PDAC.malignant.mhc1.expression.cancer, avg1)

avg2 <- t(data.frame(colMeans(BCC.PDAC.malignant.mhc2.expression.cancer)))
BCC.PDAC.malignant.mhc2.expression.cancer <- rbind(BCC.PDAC.malignant.mhc2.expression.cancer, avg2)

BCC.PDAC.malignant.mhc.expression.cancer <- rbind(BCC.PDAC.malignant.mhc1.expression.cancer, BCC.PDAC.malignant.mhc2.expression.cancer)

rownames(BCC.PDAC.malignant.mhc.expression.cancer)[rownames(BCC.PDAC.malignant.mhc.expression.cancer) == "colMeans.BCC.PDAC.malignant.mhc1.expression.cancer."] <- "MHC Class I"
rownames(BCC.PDAC.malignant.mhc.expression.cancer)[rownames(BCC.PDAC.malignant.mhc.expression.cancer) == "colMeans.BCC.PDAC.malignant.mhc2.expression.cancer."] <- "MHC Class II"

BCC.PDAC.malignant.mhc.expression.cancer <- data.frame(t(BCC.PDAC.malignant.mhc.expression.cancer))

# Execute 1198 - 1210

## Recreate dumbbell plots with both responders and nonresponders

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$response

BCC.PDAC.malignant.averages.response <- AverageExpression(BCC.PDAC.malignant)
BCC.PDAC.malignant.averages.response.mhc1 <- BCC.PDAC.malignant.averages.response[["RNA"]][mhc1.genes,]
BCC.PDAC.malignant.averages.response.mhc2 <- BCC.PDAC.malignant.averages.response[["RNA"]][mhc2.genes,]

avg1 <- t(data.frame(colMeans(BCC.PDAC.malignant.averages.response.mhc1)))
BCC.PDAC.malignant.averages.response.mhc1 <- rbind(BCC.PDAC.malignant.averages.response.mhc1, avg1)

avg2 <- t(data.frame(colMeans(BCC.PDAC.malignant.averages.response.mhc2)))
BCC.PDAC.malignant.averages.response.mhc2 <- rbind(BCC.PDAC.malignant.averages.response.mhc2, avg2)

BCC.PDAC.malignant.averages.response.mhc <- rbind(BCC.PDAC.malignant.averages.response.mhc1, BCC.PDAC.malignant.averages.response.mhc2)

rownames(BCC.PDAC.malignant.averages.response.mhc)[rownames(BCC.PDAC.malignant.averages.response.mhc) == "colMeans.BCC.PDAC.malignant.averages.response.mhc1."] <- "MHC Class I"
rownames(BCC.PDAC.malignant.averages.response.mhc)[rownames(BCC.PDAC.malignant.averages.response.mhc) == "colMeans.BCC.PDAC.malignant.averages.response.mhc2."] <- "MHC Class II"

BCC.PDAC.malignant.averages.response.mhc <- data.frame(t(BCC.PDAC.malignant.averages.response.mhc))

BCC.PDAC.malignant.averages.response.mhc$Response <- rownames(BCC.PDAC.malignant.averages.response.mhc)
BCC.PDAC.malignant.averages.response.mhc <- melt(BCC.PDAC.malignant.averages.response.mhc, id.vars="Response")

BCC.PDAC.malignant.averages.response.mhc <- BCC.PDAC.malignant.averages.response.mhc %>% 
  mutate(paired = rep(1:(n()/3), each=3), Response=factor(Response))

BCC.PDAC.malignant.averages.response.mhc %>% ggplot(aes(x=value, y=variable)) +
  geom_line(aes(group=paired)) +
  geom_point(aes(color=Response), size=5) +
  theme(legend.position="top") +
  labs(title = "Average Expression of MHC Genes", x = "Score", y = "Gene") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

## Recreate dumbbell plots, separating pretreatment and posttreatment

BCC.PDAC.malignant$restreat <- "PDAC"
BCC.PDAC.malignant$restreat[BCC.PDAC.malignant$treatment == "pre" & BCC.PDAC.malignant$response == "Responsive"] <- "Responsive Pretreatment"
BCC.PDAC.malignant$restreat[BCC.PDAC.malignant$treatment == "post" & BCC.PDAC.malignant$response == "Responsive"] <- "Responsive Posttreatment"
BCC.PDAC.malignant$restreat[BCC.PDAC.malignant$treatment == "pre" & BCC.PDAC.malignant$response == "Nonresponsive"] <- "Nonresponsive Pretreatment"
BCC.PDAC.malignant$restreat[BCC.PDAC.malignant$treatment == "post" & BCC.PDAC.malignant$response == "Nonresponsive"] <- "Nonresponsive Posttreatment"

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$restreat

BCC.PDAC.malignant.averages.restreat <- AverageExpression(BCC.PDAC.malignant)
BCC.PDAC.malignant.averages.restreat.mhc1 <- BCC.PDAC.malignant.averages.restreat[["RNA"]][mhc1.genes,]
BCC.PDAC.malignant.averages.restreat.mhc2 <- BCC.PDAC.malignant.averages.restreat[["RNA"]][mhc2.genes,]

avg1 <- t(data.frame(colMeans(BCC.PDAC.malignant.averages.restreat.mhc1)))
BCC.PDAC.malignant.averages.restreat.mhc1 <- rbind(BCC.PDAC.malignant.averages.restreat.mhc1, avg1)

avg2 <- t(data.frame(colMeans(BCC.PDAC.malignant.averages.restreat.mhc2)))
BCC.PDAC.malignant.averages.restreat.mhc2 <- rbind(BCC.PDAC.malignant.averages.restreat.mhc2, avg2)

BCC.PDAC.malignant.averages.restreat.mhc <- rbind(BCC.PDAC.malignant.averages.restreat.mhc1, BCC.PDAC.malignant.averages.restreat.mhc2)

rownames(BCC.PDAC.malignant.averages.restreat.mhc)[rownames(BCC.PDAC.malignant.averages.restreat.mhc) == "colMeans.BCC.PDAC.malignant.averages.restreat.mhc1."] <- "MHC Class I"
rownames(BCC.PDAC.malignant.averages.restreat.mhc)[rownames(BCC.PDAC.malignant.averages.restreat.mhc) == "colMeans.BCC.PDAC.malignant.averages.restreat.mhc2."] <- "MHC Class II"

BCC.PDAC.malignant.averages.restreat.mhc <- data.frame(t(BCC.PDAC.malignant.averages.restreat.mhc))

BCC.PDAC.malignant.averages.restreat.mhc$Response <- rownames(BCC.PDAC.malignant.averages.restreat.mhc)
BCC.PDAC.malignant.averages.restreat.mhc <- melt(BCC.PDAC.malignant.averages.restreat.mhc, id.vars="Response")

BCC.PDAC.malignant.averages.restreat.mhc <- BCC.PDAC.malignant.averages.restreat.mhc %>% 
  mutate(paired = rep(1:(n()/5), each=5), Response=factor(Response))

BCC.PDAC.malignant.averages.restreat.mhc %>% ggplot(aes(x=value, y=variable)) +
  geom_line(aes(group=paired)) +
  geom_point(aes(color=Response), size=5) +
  theme(legend.position="top") +
  labs(title = "Average Expression of MHC Genes", x = "Score", y = "Gene") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

```



Investigate MHC expression in different clusters and patients 

```{r}

BCC.PDAC.malignant <- preprocessing(BCC.PDAC.malignant)
BCC.PDAC.malignant <- clustering(BCC.PDAC.malignant, resolution=0.5)

BCC.PDAC.malignant$clusters <- save$seurat_clusters
Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$clusters
DimPlot(BCC.PDAC.malignant, reduction = "umap")

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$seurat_clusters
BCC.PDAC.malignant.averages <- AverageExpression(BCC.PDAC.malignant, return.seurat=TRUE, add.ident="restreat")

mhc.genes <- c(mhc1.genes, mhc2.genes)

BCC.PDAC.mhc.heatmap <- DoHeatmap(BCC.PDAC.malignant.averages, features=mhc.genes, size=3, draw.lines=FALSE)

BCC.PDAC.heatmap <- DoHeatmap(BCC.PDAC.malignant.averages, features = unlist(TopFeatures(BCC.PDAC.malignant[['pca']], balanced=TRUE)), size=3, draw.lines=FALSE)
BCC.PDAC.heatmap

## Investigate expression per patient

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$response

BCC.PDAC.malignant.averages <- AverageExpression(BCC.PDAC.malignant, return.seurat=TRUE, add.ident="patient")
BCC.PDAC.patient.heatmap <- DoHeatmap(BCC.PDAC.malignant.averages, features=mhc.genes, size=3, draw.lines=TRUE)

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$treatment
BCC.PDAC.malignant.pre <- subset(BCC.PDAC.malignant, idents=c("PDAC", "pre"))

Idents(BCC.PDAC.malignant.pre) <- BCC.PDAC.malignant.pre$response
BCC.PDAC.malignant.pre.averages <- AverageExpression(BCC.PDAC.malignant.pre, return.seurat=TRUE, add.ident="patient")
BCC.PDAC.pre.patient.heatmap <- DoHeatmap(BCC.PDAC.malignant.pre.averages, features=mhc.genes, size=3, draw.lines=TRUE)

## Throw out patient 1

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$patient
BCC.PDAC.malignant.p1 <- subset(BCC.PDAC.malignant, idents="pdac1", invert=TRUE)

Idents(BCC.PDAC.malignant.p1) <- BCC.PDAC.malignant.p1$response
BCC.PDAC.malignant.p1.averages <- AverageExpression(BCC.PDAC.malignant.p1, return.seurat=TRUE, add.ident="patient")
BCC.PDAC.patient.p1.heatmap <- DoHeatmap(BCC.PDAC.malignant.p1.averages, features=mhc.genes, size=3, draw.lines=TRUE)

Idents(BCC.PDAC.malignant.p1) <- BCC.PDAC.malignant.p1$treatment
BCC.PDAC.malignant.p1.pre <- subset(BCC.PDAC.malignant.p1, idents=c("PDAC", "pre"))

Idents(BCC.PDAC.malignant.p1.pre) <- BCC.PDAC.malignant.p1.pre$response
BCC.PDAC.malignant.p1.pre.averages <- AverageExpression(BCC.PDAC.malignant.p1.pre, return.seurat=TRUE, add.ident="patient")
BCC.PDAC.patient.p1.pre.heatmap <- DoHeatmap(BCC.PDAC.malignant.p1.pre.averages, features=mhc.genes, size=3, draw.lines=TRUE)

## Separate by patient and treatment

Idents(BCC.PDAC.malignant) <- BCC.PDAC.malignant$treatment
BCC.malignant <- subset(BCC.PDAC.malignant, idents=c("pre", "post"))

Idents(BCC.malignant) <- BCC.malignant$patient
BCC.malignant.averages <- AverageExpression(BCC.malignant, return.seurat=TRUE, add.ident="treatment")

BCC.malignant.averages.mhc1 <- data.frame(BCC.malignant.averages[["RNA"]][mhc1.genes,])
BCC.malignant.averages.mhc2 <- data.frame(BCC.malignant.averages[["RNA"]][mhc2.genes,])

BCC.malignant.averages.mhc1$su001_post <- 0
BCC.malignant.averages.mhc1$su002_post <- 0
BCC.malignant.averages.mhc1$su010_pre <- 0

BCC.malignant.averages.mhc2$su001_post <- 0
BCC.malignant.averages.mhc2$su002_post <- 0
BCC.malignant.averages.mhc2$su010_pre <- 0

BCC.malignant.averages.mhc1 <- BCC.malignant.averages.mhc1[, c(1, 16, 2, 17, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 18, 15)]
BCC.malignant.averages.mhc2 <- BCC.malignant.averages.mhc2[, c(1, 16, 2, 17, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 18, 15)]

avg1 <- t(data.frame(colMeans(BCC.malignant.averages.mhc1)))
avg2 <- t(data.frame(colMeans(BCC.malignant.averages.mhc2)))

BCC.malignant.averages.mhc <- rbind(avg1, avg2)

BCC.malignant.averages.mhc.pre <- BCC.malignant.averages.mhc[, c(1, 3, 5, 7, 9, 11, 13, 15, 17)]
BCC.malignant.averages.mhc.post <- BCC.malignant.averages.mhc[, c(2, 4, 6, 8, 10, 12, 14, 16, 18)]
colnames(BCC.malignant.averages.mhc.pre) <- c('su001', 'su002', 'su003', 'su004', 'su005', 'su006', 'su007', 'su008', 'su010')
rownames(BCC.malignant.averages.mhc.pre) <- c('MHC Class 1 Pre-treatment', 'MHC Class 2 Pre-treatment')
colnames(BCC.malignant.averages.mhc.post) <- c('su001', 'su002', 'su003', 'su004', 'su005', 'su006', 'su007', 'su008', 'su010')
rownames(BCC.malignant.averages.mhc.post) <- c('MHC Class 1 Post-treatment', 'MHC Class 2 Post-treatment')
BCC.malignant.averages.mhc <- data.frame(rbind(BCC.malignant.averages.mhc.pre, BCC.malignant.averages.mhc.post))

save <- BCC.malignant.averages.mhc

BCC.malignant.averages.mhc$Response <- rownames(BCC.malignant.averages.mhc)
BCC.malignant.averages.mhc <- melt(BCC.malignant.averages.mhc, id.vars="Response")

BCC.malignant.averages.mhc <- BCC.malignant.averages.mhc %>% 
  mutate(paired = rep(1:(n()/4), each=4), Response=factor(Response))

BCC.malignant.averages.mhc %>% ggplot(aes(x=value, y=variable)) +
  geom_line(aes(group=paired)) +
  geom_point(aes(color=Response), size=5) +
  theme(legend.position="top") +
  labs(title = "Average Expression of MHC Genes", x = "Score", y = "Patient") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

## Five-way boxplot to compare MHC scores

BCCIO$restreat <- "filler"

BCCIO$restreat[BCCIO$treatment == "pre" & BCCIO$response == "Responsive"] <- "Responsive Pretreatment"
BCCIO$restreat[BCCIO$treatment == "post" & BCCIO$response == "Responsive"] <- "Responsive Posttreatment"
BCCIO$restreat[BCCIO$treatment == "pre" & BCCIO$response == "Nonresponsive"] <- "Nonresponsive Pretreatment"
BCCIO$restreat[BCCIO$treatment == "post" & BCCIO$response == "Nonresponsive"] <- "Nonresponsive Posttreatment"

Idents(BCCIO) <- BCCIO$restreat

BCC.responder.pre <- subset(BCCIO, idents="Responsive Pretreatment")
BCC.responder.post <- subset(BCCIO, idents="Responsive Posttreatment")
BCC.nonresponder.pre <- subset(BCCIO, idents="Nonresponsive Pretreatment")
BCC.nonresponder.post <- subset(BCCIO, idents="Nonresponsive Posttreatment")

PDAC155698.integrated <- PDAC.integrated

Idents(BCCIO) <- BCCIO$seurat_clusters
Idents(PDAC155698.integrated) <- PDAC155698.integrated$seurat_clusters

BCC.PDAC.comparison.malignant <- fiveway_diffexpression_MHCandHSP(c("Tumor 1", "Tumor 2"), c("Ductal cells 1", "Ductal cells 2", "Ductal cells 3", "Ductal cells 4"), "Malignant Cells")
grid.draw(BCC.PDAC.comparison.malignant)

```



################################################################################################################

Compare gene expression between malignnat cells in BCC vs. PDAC

```{r}

## Isolate malignant/ductal cells into its own Seurat object and integrate

Idents(BCCIO) <- BCCIO$seurat_clusters
Idents(PDAC.all) <- PDAC.all$cluster

BCC.malignant <- subset(BCCIO, idents=c("Tumor 1", "Tumor 2"))

PDAC.malignant <- subset(PDAC.all, idents=c("Ductal cells: Healthy", "Ductal cells: Malignant 1", "Ductal cells: Malignant 2", "Ductal cells: Normal"))

malignant.anchors <- FindIntegrationAnchors(object.list=list(BCC.malignant, PDAC.malignant), dims=1:50)
malignant <- IntegrateData(anchorset=malignant.anchors, dims=1:50)
DefaultAssay(malignant) <- "RNA"

malignant@meta.data$seurat_clusters[is.na(malignant@meta.data$seurat_clusters)] <- malignant@meta.data$cluster[is.na(malignant@meta.data$seurat_clusters)]

Idents(malignant) <- malignant$seurat_clusters

malignant <- RenameIdents(object=malignant, 
                          'Tumor 1' = 'BCC Malignant 1',
                          'Tumor 2' = 'BCC Malignant 2',
                          'Ductal cells: Malignant 1' = 'PDAC DC Malignant 1',
                          'Ductal cells: Malignant 2' = 'PDAC DC Malignant 2',
                          'Ductal cells: Normal'= 'PDAC DC Normal',
                          'Ductal cells: Healthy' = 'PDAC DC Healthy')

malignant$imported_clusters <- Idents(malignant)

malignant <- preprocessing(malignant, heatmap=FALSE)
malignant <- clustering(malignant, resolution=0.3)

malignant$imported_clusters[rownames(BCC.malignant@meta.data)] <- paste(BCC.malignant$seurat_clusters)
malignant$imported_clusters[rownames(PDAC.malignant@meta.data)] <- paste(PDAC.malignant$cluster)

Idents(malignant) <- malignant$imported_clusters

malignant <- RenameIdents(object=malignant,
                          'Tumor 1' = 'BCC Malignant 1',
                          'Tumor 2' = 'BCC Malignant 2',
                          'Ductal cells: Malignant 1' = 'PDAC DC Malignant 1',
                          'Ductal cells: Malignant 2' = 'PDAC DC Malignant 2',
                          'Ductal cells: Normal' = 'PDAC DC Normal',
                          'Ductal cells: Healthy' = 'PDAC DC Healthy')

malignant$imported_clusters <- Idents(malignant)

malignant <- RenameIdents(object=malignant,
                          'BCC Malignant 1' = 'BCC',
                          'BCC Malignant 2' = 'BCC',
                          'PDAC DC Malignant 1' = 'PDAC',
                          'PDAC DC Malignant 2' = 'PDAC',
                          'PDAC DC Normal' = 'PDAC',
                          'PDAC DC Healthy' = 'PDAC')

malignant$cancer <- Idents(malignant)

malignant$healthy <- malignant$imported_clusters
Idents(malignant) <- malignant$healthy

malignant <- RenameIdents(object=malignant,
                          'BCC Malignant 1' = 'BCC',
                          'BCC Malignant 2' = 'BCC',
                          'PDAC DC Malignant 1' = 'Malignant',
                          'PDAC DC Malignant 2' = 'Malignant',
                          'PDAC DC Normal' = 'Normal',
                          'PDAC DC Healthy' = 'Normal')

malignant$healthy <- Idents(malignant)

## Look at differential expression between the populations 

DimPlot(malignant, group.by="imported_clusters", label=TRUE, shuffle=TRUE) + NoLegend()
DimPlot(malignant, group.by="seurat_clusters", label=TRUE, shuffle=TRUE) + NoLegend()

ggplot_barcharts(malignant$imported_clusters, malignant$seurat_clusters)
ggplot_barcharts(malignant$cancer, malignant$seurat_clusters)

malignant$filler <- "everything"
Idents(malignant) <- malignant$filler
malignant.volcano <- bioc_volcano(malignant, "everything", "cancer", "PDAC", "BCC", title="Malignant cells")

Idents(malignant) <- malignant$cancer
pdac.volcano <- bioc_volcano(malignant, "PDAC", "healthy", "Malignant", "Normal", title="PDAC Ductal Cells")

Idents(malignant) <- malignant$imported_clusters
pdacupregulated.filter <- c('S100P', 'CEACAM6', 'CLDN18', 'GPRC5A', 'PHLDA2', 'CEACAM5', 'LAMC2', 'SFN', 'AHNAK2', 'TFF1', 'SDR16C5', 'LAMA3', 'IFI27', 'TGM2', 'NQO1', 'TPM2', 'S100A4', 'TRIM29', 'CD55', 'TMPRSS4', 'FOXQ1', 'CAPN8', 'PLAUR', 'ADAM28', 'FXYD3', 'TSPAN1', 'FXYD5', 'KCNK1')
DotPlot(malignant, features = pdacupregulated.filter, assay = 'RNA', cols = c('White', 'Blue')) + RotatedAxis() + xlab('Marker Genes')+ ylab('Clusters')

mhc.genes <- c('HLA-A', 'HLA-B', 'HLA-C', 'HLA-E', 'HLA-F', 'HLA-G', 'HLA-DMA', 'HLA-DMB', 'HLA-DOA', 'HLA-DPA1', 'HLA-DPB1', 'HLA-DQA1', 'HLA-DQA2', 'HLA-DQB1', 'HLA-DQB2', 'HLA-DRA', 'HLA-DRB1', 'HLA-DRB5')
DotPlot(malignant, features = mhc.genes, assay = 'RNA', cols = c('White', 'Blue')) + RotatedAxis() + xlab('Marker Genes')+ ylab('Clusters')

```



Investigate MHC expression between BCC and PDAC 

```{r}

## MHC score comparison using all BCC cells 

mhc1.genes <- c('HLA-A', 'HLA-B', 'HLA-C', 'HLA-E', 'HLA-F', 'HLA-G')
mhc2.genes <- c('HLA-DMA', 'HLA-DMB', 'HLA-DOA', 'HLA-DPA1', 'HLA-DPB1', 'HLA-DQA1', 'HLA-DQA2', 'HLA-DQB1', 'HLA-DQB2', 'HLA-DRA', 'HLA-DRB1', 'HLA-DRB5')

exp <- AverageExpression(malignant)
mhc1.exp <- t(exp$RNA[mhc1.genes,])
mhc2.exp <- t(exp$RNA[mhc2.genes,])

data <- as.data.frame(t(cbind(rownames(exp), mhc1.exp, mhc2.exp, rowMeans(mhc1.exp), rowMeans(mhc2.exp))))
rownames(data) <- c(mhc1.genes, mhc2.genes, "MHC Class I", "MHC Class II")

data$gene <- rownames(data)
data <- melt(data, id.vars="gene")
data <- data %>% mutate(paired = rep(1:(n()/length(rownames(data))), each=length(rownames(data))), gene=factor(gene))

data %>% ggplot(aes(x=value, y=gene)) +
  geom_line(aes(group=gene)) +
  geom_point(aes(color=variable), size=5) +
  theme(legend.position="top") +
  labs(title = "MHC Gene Expression", x = "Expression", y = "Gene") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

## MHC score comparison separating responsive and nonresponsive BCC cells, pretreatment and posttreatment

malignant$refined_clusters <- as.character(malignant$imported_clusters)
malignant@meta.data[is.na(malignant@meta.data)] <- "NA"

refined_clusters <- c()
for (i in 1:nrow(malignant@meta.data)) {
  if (malignant$treatment[i] == "pre" && malignant$response[i] == "Responsive") {
    refined_clusters[i] <- "BCC Responsive Pre-treatment"
  }
  else if (malignant$treatment[i] == "pre" && malignant$response[i] == "Nonresponsive") {
    refined_clusters[i] <- "BCC Nonresponsive Pre-treatment"
  }
  else if (malignant$treatment[i] == "post" && malignant$response[i] == "Responsive") {
    refined_clusters[i] <- "BCC Responsive Post-treatment"
  }
  else if (malignant$treatment[i] == "post" && malignant$response[i] == "Nonresponsive") {
    refined_clusters[i] <- "BCC Nonresponsive Post-treatment" 
  }
  else {
    refined_clusters[i] <- malignant$imported_clusters[i]
  }
}

malignant$refined_clusters <- refined_clusters
Idents(malignant) <- malignant$refined_clusters

malignant <- RenameIdents(object=malignant,
                          "3" = "PDAC DC Malignant 1",
                          "4" = "PDAC DC Malignant 2", 
                          "5" = "PDAC DC Normal",
                          "6" = "PDAC DC Healthy")

malignant$refined_clusters <- Idents(malignant)

for (i in 1:nrow(malignant@meta.data)) {
  if (malignant$orig.ident[i] == "H1") {
    malignant$patient[i] <- "pdacH1"
  } else if (malignant$orig.ident[i] == "H2") {
    malignant$patient[i] <- "pdacH2"
  } else if (malignant$orig.ident[i] == "H3") {
    malignant$patient[i] <- "pdacH3"
  }
}

save(malignant, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC PDAC Malignant Cells")

exp <- AverageExpression(malignant)
mhc1.exp <- t(exp$RNA[mhc1.genes,])
mhc2.exp <- t(exp$RNA[mhc2.genes,])

data <- as.data.frame(t(cbind(rownames(exp), mhc1.exp, mhc2.exp, rowMeans(mhc1.exp), rowMeans(mhc2.exp))))
rownames(data) <- c(mhc1.genes, mhc2.genes, "MHC Class I", "MHC Class II")

data$gene <- rownames(data)
data <- melt(data, id.vars="gene")
data <- data %>% mutate(paired = rep(1:(n()/length(rownames(data))), each=length(rownames(data))), gene=factor(gene))

data %>% ggplot(aes(x=value, y=gene)) +
  geom_line(aes(group=gene)) +
  geom_point(aes(color=variable), size=5) +
  theme(legend.position="top") +
  labs(title = "MHC Gene Expression", x = "Expression", y = "Gene") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

```



Look at MHC gene expression by individual cell

```{r}

Idents(malignant) <- malignant$refined_clusters

malignant <- RenameIdents(object=malignant,
                          "PDAC DC Malignant 1" = "PDAC M1",
                          "PDAC DC Malignant 2" = "PDAC M2",
                          "PDAC DC Normal" = "PDAC N",
                          "PDAC DC Healthy" = "PDAC H",
                          "BCC Responsive Pre-treatment" =  "BCC RPre",
                          "BCC Responsive Post-treatment" = "BCC RPost",
                          "BCC Nonresponsive Pre-treatment" = "BCC NRPre",
                          "BCC Nonresponsive Post-treatment" = "BCC NRPost")

malignant$refined_clusters <- Idents(malignant)

## Look at MHC genes individually

VlnPlot(malignant, features=mhc.genes, pt.size=0)

## Look at MHC scores (all cells) -- CDF

malignant.counts <- malignant@assays$RNA@counts

malignant.mhc1.counts <- data.frame(malignant.counts[mhc1.genes,])
malignant.mhc2.counts <- data.frame(malignant.counts[mhc2.genes,])

malignant.mhc1.counts["MHC Class I",] <- colSums(malignant.mhc1.counts)
malignant.mhc2.counts["MHC Class II",] <- colSums(malignant.mhc2.counts)

malignant.mhc1.counts <- data.frame(t(malignant.mhc1.counts))
malignant.mhc2.counts <- data.frame(t(malignant.mhc2.counts))

mhc1.cdf <- ggplot(malignant.mhc1.counts, aes(MHC.Class.I+1)) + stat_ecdf(geom = "step", size=1) +
  labs(title="MHC Class I Score", x="Sum of Normalized Count Matrix", y="Cumulative Proportion") +
  scale_x_log10() + theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

mhc2.cdf <- ggplot(malignant.mhc2.counts, aes(MHC.Class.II+1)) + stat_ecdf(geom = "step", size=1) +
  labs(title="MHC Class II Score", x="Sum of Count Normalized Matrix", y="Cumulative Proportion") +
  scale_x_log10() + theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

grid.arrange(mhc1.cdf, mhc2.cdf, ncol=2)

## Look at MHC scores divided by cluster -- CDF

Idents(malignant) <- malignant$refined_clusters
clusters <- levels(Idents(malignant))

for (i in 1:length(clusters)) {
  assign(paste0("malignant.", clusters[i]), subset(malignant, idents=clusters[i]))
}

clusters <- list(`malignant.PDAC M1`, `malignant.PDAC M2`, `malignant.PDAC N`, `malignant.PDAC H`, `malignant.BCC RPre`, `malignant.BCC RPost`, `malignant.BCC NRPre`, `malignant.BCC NRPost`)

MHC.Class.I <- c(0)
MHC.Class.II <- c(0)
id <- c(0)

mhc1.cdf.data <- data.frame(MHC.Class.I, id)
mhc2.cdf.data <- data.frame(MHC.Class.II, id)

for (i in 1:length(clusters)) {
  counts <- clusters[[i]]@assays$RNA@data
  mhc1.counts <- data.frame(counts[mhc1.genes,])
  mhc2.counts <- data.frame(counts[mhc2.genes,])
  mhc1.counts["MHC Class I",] <- colSums(mhc1.counts)
  mhc2.counts["MHC Class II",] <- colSums(mhc2.counts)
  mhc1.counts["id",] <- levels(Idents(malignant))[i]
  mhc2.counts["id",] <- levels(Idents(malignant))[i]
  mhc1.counts <- data.frame(t(mhc1.counts))
  mhc2.counts <- data.frame(t(mhc2.counts))
  mhc1.cdf.data <- rbind(mhc1.cdf.data, mhc1.counts[,c("MHC.Class.I", "id")])
  mhc2.cdf.data <- rbind(mhc2.cdf.data, mhc2.counts[,c("MHC.Class.II", "id")])
}

mhc1.cdf.data <- mhc1.cdf.data[-1,]
mhc2.cdf.data <- mhc2.cdf.data[-1,]
mhc1.cdf.data$MHC.Class.I <- as.numeric(mhc1.cdf.data$MHC.Class.I)
mhc2.cdf.data$MHC.Class.II <- as.numeric(mhc2.cdf.data$MHC.Class.II)
mhc1.cdf.data$id <- as.factor(mhc1.cdf.data$id)
mhc2.cdf.data$id <- as.factor(mhc2.cdf.data$id)

mhc1.cdf <- ggplot(mhc1.cdf.data, aes(x=MHC.Class.I, col=id)) + stat_ecdf(geom = "step", size=1) +
  labs(title="MHC Class I Score", x="Sum of Normalized Count Matrix", y="Cumulative Proportion") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

mhc2.cdf <- ggplot(mhc2.cdf.data, aes(x=MHC.Class.II, col=id)) + stat_ecdf(geom = "step", size=1) +
  labs(title="MHC Class II Score", x="Sum of Normalized Count Matrix", y="Cumulative Proportion") +
  theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

ggarrange(mhc1.cdf, mhc2.cdf, ncol=2, common.legend=TRUE, legend="bottom")

## Construct PDF / Violin Plots for MHC expression 

mhc1.pdf <- ggplot(mhc1.cdf.data, aes(x=MHC.Class.I+1, fill=id)) + geom_density(alpha=0.3) +
  labs(title="MHC Class I Score", x="Sum of Count Matrix",y = "Probability Density") +
  scale_x_log10() + theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))
  
mhc2.pdf <- ggplot(mhc2.cdf.data, aes(x=MHC.Class.II+1, fill=id)) + geom_density(alpha=0.3) +
  labs(title="MHC Class II Score", x="Sum of Count Matrix",y = "Probability Density") +
  scale_x_log10() + theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

ggarrange(mhc1.pdf, mhc2.pdf, ncol=2, common.legend=TRUE, legend="bottom")

mhc1.vln <- ggplot(mhc1.cdf.data, aes(x=MHC.Class.I+1, factor(id))) + geom_violin() +
  labs(title="MHC Class I Score", x = "Sum of Normalized Count Matrix", y = "Cluster") +
  aes(fill=factor(id)) + scale_x_log10() + theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

mhc2.vln <- ggplot(mhc2.cdf.data, aes(x=MHC.Class.II+1, factor(id))) + geom_violin() +
  labs(title="MHC Class II Score", x = "Sum of Normalized Count Matrix", y = "Cluster") +
  aes(fill=factor(id)) + scale_x_log10() + theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

ggarrange(mhc1.vln, mhc2.vln, ncol=2, common.legend=TRUE, legend="bottom")

```



Look at per-patient MHC scores

```{r}

Idents(malignant) <- malignant$cancer
malignant.pdac <- subset(malignant, idents="PDAC")

Idents(malignant.pdac) <- malignant.pdac$orig.ident
patients <- levels(Idents(malignant.pdac))

for (i in 1:length(patients)) {
  assign(paste0("malignant.", patients[i]), subset(malignant.pdac, idents=patients[i]))
}

## Look at distribution of cells per patient

ggplot_barcharts(malignant.pdac$refined_clusters, malignant.pdac$orig.ident)

## Construct CDF

infected.patients <- list(malignant.I1, malignant.I2, malignant.I3, malignant.I4, malignant.I5, malignant.I6, malignant.I7, malignant.I8, malignant.I9, malignant.I10, malignant.I11, malignant.I12, malignant.I13, malignant.I14, malignant.I15, malignant.I16)

healthy.patients <- list(malignant.H1, malignant.H2, malignant.H3)

infected.patient.names <- c('I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9', 'I10', 'I11A', 'I11B', 'I12', 'I13', 'I15', 'I16')
healthy.patient.names <- c('H1', 'H2', 'H3')

mhc1.icdf <- genes_cdf(datasets=infected.patients, genes=mhc1.genes, clusters=infected.patient.names, title="MHC Class I Score: Cancerous Samples") + xlim(0, 21)
mhc1.hcdf <- genes_cdf(datasets=healthy.patients, genes=mhc1.genes, clusters=healthy.patient.names, title="MHC Class I Score: Non-Cancerous Samples") + xlim(0, 21)

mhc2.icdf <- genes_cdf(datasets=infected.patients, genes=mhc2.genes, clusters=infected.patient.names, title="MHC Class II Score: Cancerous Samples") + xlim(0, 34)
mhc2.hcdf <- genes_cdf(datasets=healthy.patients, genes=mhc2.genes, clusters=healthy.patient.names, title="MHC Class II Score: Non-Cancerous Samples") + xlim(0, 34)

ggarrange(mhc1.icdf, mhc1.hcdf, mhc2.icdf, mhc2.hcdf, ncol=2, nrow=2, common.legend=TRUE, legend="bottom")

## Construct violion plots

mhc1.vln <- ggplot(mhc1.cdf.data, aes(x=MHC.Class.I, factor(id))) + geom_violin() +
  labs(title="MHC Class I Score", x = "Sum of Count Matrix", y = "Patient") +
  aes(fill=factor(id)) + theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

mhc2.vln <- ggplot(mhc2.cdf.data, aes(x=MHC.Class.II, factor(id))) + geom_violin() +
  labs(title="MHC Class II Score", x = "Sum of Count Matrix", y = "Patient") +
  aes(fill=factor(id)) + theme_light() +
  theme(legend.position="top", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

ggarrange(mhc1.vln, mhc2.vln, ncol=2, common.legend=TRUE, legend="bottom")

## Compare subclusters per patient

levels <- c("M1", "M2", "N")

for (i in 1:length(patients)) {
  Idents(infected.patients[[i]]) <- infected.patients[[i]]$refined_clusters
  m1 <- subset(infected.patients[[i]], idents="PDAC DC Malignant 1")
  m2 <- subset(infected.patients[[i]], idents="PDAC DC Malignant 2")
  n <- subset(infected.patients[[i]], idents="PDAC DC Normal")
  
  subclusters <- list(m1, m2, n)
  assign(paste0(patient.names[i], ".cdf1"), gene_cdf(datasets=subclusters, genes=mhc1.genes, clusters=levels, title=paste0("MHC I Class Score:  ", patient.names[i])))
  assign(paste0(patient.names[i], ".cdf2"), gene_cdf(datasets=subclusters, genes=mhc2.genes, clusters=levels, title=paste0("MHC Class II Score: ", patient.names[i])))
  print(patient.names[i])
}

patients.cdf1 <- list(I1.cdf1, I2.cdf1, I3.cdf1, I4.cdf1, I5.cdf1, I6.cdf1, I7.cdf1, I8.cdf1, I9.cdf1, I11A.cdf1, I11B.cdf1, I13.cdf1, I15.cdf1, I16.cdf1)

patients.cdf2 <- list(I1.cdf2, I2.cdf2, I3.cdf2, I4.cdf2, I5.cdf2, I6.cdf2, I7.cdf2, I8.cdf2, I9.cdf2, I11A.cdf2, I11B.cdf2, I13.cdf2, I15.cdf2, I16.cdf2)

ggarrange(plotlist=patients.cdf1, ncol=4, nrow=4, common.legend=TRUE, legend="bottom")
ggarrange(plotlist=patients.cdf2, ncol=4, nrow=4, ommon.legend=TRUE, legend="bottom")

```



Statistical comparison of MHC Class I and Class II scores, calculated by patient 

```{r}

## Construct boxplot

clusters <- list(`malignant.PDAC M1`, `malignant.PDAC M2`, `malignant.PDAC N`, `malignant.PDAC H`, `malignant.BCC RPre`, `malignant.BCC RPost`, `malignant.BCC NRPre`, `malignant.BCC NRPost`)

clusternames <- c('PDAC.M1','PDAC.M2', 'PDAC.N', 'PDAC.H', 'BCC.RPre', 'BCC.RPost', 'BCC.NRPre', 'BCC.NRPost')

pdacclusters <- list(`malignant.PDAC M1`, `malignant.PDAC M2`, `malignant.PDAC N`, `malignant.PDAC H`)

for (i in 1:length(pdacclusters)) {
  clusters[[i]]$patient <- clusters[[i]]$orig.ident
}

for (i in 1:length(clusters)) {
  Idents(clusters[[i]]) <- clusters[[i]]$patient
  assign(paste0(clusternames[i], ".exp"), AverageExpression(clusters[[i]]))
}

cluster.averages <- list(PDAC.M1.exp, PDAC.M2.exp, PDAC.N.exp, PDAC.H.exp, BCC.RPre.exp, BCC.RPost.exp, BCC.NRPre.exp, BCC.NRPost.exp)

for (i in 1:length(cluster.averages)) {
  assign(paste0(clusternames[i], ".exp.mhc1"), cluster.averages[[i]]$RNA[mhc1.genes, ])
  assign(paste0(clusternames[i], ".exp.mhc2"), cluster.averages[[i]]$RNA[mhc2.genes, ])
}

cluster.averages.mhc1 <- list(PDAC.M1.exp.mhc1, PDAC.M2.exp.mhc1, PDAC.N.exp.mhc1, PDAC.H.exp.mhc1, BCC.RPre.exp.mhc1, BCC.RPost.exp.mhc1, BCC.NRPre.exp.mhc1, BCC.NRPost.exp.mhc1)

cluster.averages.mhc2 <- list(PDAC.M1.exp.mhc2, PDAC.M2.exp.mhc2, PDAC.N.exp.mhc2, PDAC.H.exp.mhc2, BCC.RPre.exp.mhc2, BCC.RPost.exp.mhc2, BCC.NRPre.exp.mhc2, BCC.NRPost.exp.mhc2)

for (i in 1:length(cluster.averages.mhc1)) {
  cluster.averages.mhc1[[i]] <- as.data.frame(colMeans(cluster.averages.mhc1[[i]]))
  cluster.averages.mhc1[[i]]$patient <- rownames(cluster.averages.mhc1[[i]])
  cluster.averages.mhc1[[i]]$cluster <- clusternames[i]
  colnames(cluster.averages.mhc1[[i]]) <- c("score", "patient", "cluster")
}

for (i in 1:length(cluster.averages.mhc2)) {
  cluster.averages.mhc2[[i]] <- as.data.frame(colMeans(cluster.averages.mhc2[[i]]))
  cluster.averages.mhc2[[i]]$patient <- rownames(cluster.averages.mhc2[[i]])
  cluster.averages.mhc2[[i]]$cluster <- clusternames[i]
  colnames(cluster.averages.mhc2[[i]]) <- c("score", "patient", "cluster")
}

cluster.averages.mhc1 <- rbind(cluster.averages.mhc1[[1]], cluster.averages.mhc1[[2]], cluster.averages.mhc1[[3]], cluster.averages.mhc1[[4]], cluster.averages.mhc1[[5]], cluster.averages.mhc1[[6]], cluster.averages.mhc1[[7]], cluster.averages.mhc1[[8]])

cluster.averages.mhc2 <- rbind(cluster.averages.mhc2[[1]], cluster.averages.mhc2[[2]], cluster.averages.mhc2[[3]], cluster.averages.mhc2[[4]], cluster.averages.mhc2[[5]], cluster.averages.mhc2[[6]], cluster.averages.mhc2[[7]], cluster.averages.mhc2[[8]])

comparisons <- list(c("BCC.RPre", "BCC.RPost"), c("BCC.NRPre", "BCC.NRPost"), c("BCC.RPre", "BCC.NRPre"), c("BCC.RPost", "BCC.NRPost"), c("PDAC.M1", "BCC.RPre"), c("PDAC.M1", "BCC.NRPre"), c("PDAC.M2", "BCC.RPre"), c("PDAC.M2", "BCC.NRPre"))

mhc1.score.comparison <- ggboxplot(cluster.averages.mhc1, x="cluster", y="score", color="cluster", title="MHC Class I Score", add="jitter") +
  stat_compare_means(comparisons = comparisons) +
  theme(legend.position="none", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

mhc2.score.comparison <- ggboxplot(cluster.averages.mhc2, x="cluster", y="score", color="cluster", title="MHC Class II Score", add="jitter") +
  stat_compare_means(comparisons = comparisons) +
  theme(legend.position="none", plot.title = element_text(color="black", size=18, face="bold", hjust = 0.5))

```



Comparison of MHC Class I scores based on percentage of malignant cells

```{r}

## Calculate percentage of malignant cells

Idents(malignant.pdac) <- malignant.pdac$refined_clusters
malignant.pdac <- RenameIdents(object=malignant.pdac,
                               'PDAC M1' = 'Malignant',
                               'PDAC M2' = 'Malignant', 
                               'PDAC N' = 'Non-malignant',
                               'PDAC H' = 'Non-cancerous')

malignant.pdac$malignant <- Idents(malignant.pdac)

pdac.proportions <- ggplot_barcharts(malignant.pdac$malignant, malignant.pdac$orig.ident)

pdac.m1m2.proportions <- ggplot_barcharts(malignant.pdac$refined_clusters, malignant.pdac$orig.ident) +
  theme_classic() + theme(legend.position="bottom") +
  labs(y="Cell Proportion Normalized by Patient", x="Patient", fill="Cell Type", legend.position="bottom") +
  scale_x_discrete(limits=c('I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9', 'I10', 'I11', 'I12', 'I13', 'I14', 'I15', 'I16'))


## Create scatter plot of proportion of malignant cells vs. MHC Class I Score

pdac.patients <- list(malignant.I1, malignant.I10, malignant.I11, malignant.I12, malignant.I13, malignant.I14, malignant.I15, malignant.I16, malignant.I2, malignant.I3, malignant.I4, malignant.I5, malignant.I6, malignant.I7, malignant.I8, malignant.I9, malignant.H1, malignant.H2, malignant.H3)

status <- c('Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Cancerous', 'Normal', 'Normal', 'Normal')

proportion <- rep(0, length(pdac.patients))
data <- pdac.proportions$plot_env$normalized_dataset
for (i in 1:16) {
  proportion[i] <- data$n[2*i+2] / (data$n[2*i+2] + data$n[2*i+3])
}

mhc1.scores <- rep(0, length(pdac.patients))
mhc2.scores <- rep(0, length(pdac.patients))
for (i in 1:length(pdac.patients)) {
  Idents(pdac.patients[[i]]) <- pdac.patients[[i]]$orig.ident
  avg <- AverageExpression(pdac.patients[[i]])
  mhc1.scores[i] <- mean(avg$RNA[mhc1.genes,])
  mhc2.scores[i] <- mean(avg$RNA[mhc2.genes,])
}

mhc1.scatter.data <- data.frame(status, proportion, mhc1.scores)

mhc1.scatter <- ggplot(mhc1.scatter.data, aes(x=proportion, y=mhc1.scores, color=status)) + geom_point(size=6) +
  geom_smooth(data=subset(mhc1.scatter.data, proportion>0), method="lm", se=FALSE, size=2, color="black") + 
  stat_regline_equation(label.y=12.7, label.x=0.06, aes(label = ..eq.label..), size=8) +
  stat_regline_equation(label.y=12.1, label.x=0.06, aes(label = ..rr.label..), size=8) +
  scale_fill_manual(values=c("blue", "red")) +
  labs(y="MHC Class I Score", x="Proportion of Malignant Ductal Cells", color="Status") +
  theme_bw(base_size=28) +
  scale_color_manual(values=c("maroon", "blue"))
           
mhc2.scatter.data <- data.frame(status, proportion, mhc2.scores)

mhc2.scatter <- ggplot(mhc2.scatter.data, aes(x=proportion, y=mhc2.scores, color=status)) + geom_point(size=6) +
  geom_smooth(data=subset(mhc2.scatter.data, proportion>0), method="lm", se=FALSE, size=2, color="black") + 
  stat_regline_equation(label.y=3.3, label.x=0.06, aes(label = ..eq.label..), size=8) +
  stat_regline_equation(label.y=3.1, label.x=0.06, aes(label = ..rr.label..), size=8) +
  scale_fill_manual(values=c("blue", "red")) +
  labs(y="MHC Class II Score", x="Proportion of Malignant Ductal Cells", color="Status") +
  theme_bw(base_size=28) +
  scale_color_manual(values=c("maroon", "blue"))
           
ggarrange(mhc1.scatter, mhc2.scatter, ncol=2, common.legend=TRUE, legend="bottom")

## Examine distribution of MHC Class I scores as a function of number of malignant cells 

Idents(malignant.pdac) <- malignant.pdac$malignant

malignant.pdac.lv1 <- merge(malignant.I4, y=c(malignant.I11, malignant.I9, malignant.I13))
malignant.pdac.lv2 <- merge(malignant.I8, y=c(malignant.I10, malignant.I5, malignant.I7))
malignant.pdac.lv3 <- merge(malignant.I12, y=c(malignant.I6, malignant.I3, malignant.I2))
malignant.pdac.lv4 <- merge(malignant.I15, y=c(malignant.I14, malignant.I16, malignant.I1))

malignant.levels <- list(`malignant.PDAC H`, malignant.pdac.lv1, malignant.pdac.lv2, malignant.pdac.lv3, malignant.pdac.lv4)
percentages <- c("Non-cancerous", "33% - 44%", "44% - 60%", "71% - 86%", "92% - 99%")

malignant.levels.mhc1.cdf <- genes_cdf(datasets=malignant.levels, clusters=percentages, genes=mhc1.genes, title="MHC Class I Score")
malignant.levels.mhc2.cdf <- genes_cdf(datasets=malignant.levels, clusters=percentages, genes=mhc2.genes, title="MHC Class II Score")

ggarrange(malignant.levels.mhc1.cdf, malignant.levels.mhc2.cdf, ncol=2, common.legend=TRUE, legend="bottom")

mhc1.levels.data <- malignant.levels.mhc1.cdf$data
colnames(mhc1.levels.data) <- c("MHC Class I", "Proportion")
mhc1.levels.data$`MHC Class I` <- mhc1.levels.data$`MHC Class I`

malignant.levels.mhc1.pdf <- ggplot(mhc1.levels.data, aes(x=`MHC Class I`, y=factor(Proportion), fill=Proportion)) + geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
  labs(x = "MHC Class I Score", y = "Cluster") +
  theme(legend.position="none") + theme_bw()

mhc2.levels.data <- malignant.levels.mhc2.cdf$data
colnames(mhc2.levels.data) <- c("MHC Class II", "Proportion")
mhc2.levels.data$`MHC Class II` <- mhc2.levels.data$`MHC Class II`

malignant.levels.mhc2.pdf <- ggplot(mhc2.levels.data, aes(x=`MHC Class II`, y=factor(Proportion), fill=Proportion)) + geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
  labs(x = "MHC Class II Score", y = "Cluster") +
  theme(legend.position="none") + theme_bw()

```



Examine differences in expression of the PD-1 pathway

```{r}

## PD-1 genes for malignant cells

Idents(malignant) <- malignant$refined_clusters

clusters <- list(`malignant.PDAC M1`, `malignant.PDAC M2`, `malignant.PDAC N`, `malignant.PDAC H`, `malignant.BCC RPre`, `malignant.BCC RPost`, `malignant.BCC NRPre`, `malignant.BCC NRPost`)

clusternames <- c("PDAC M1", "PDAC M2", "PDAC N", "PDAC H", "BCC RPre", "BCC RPost", "BCC NRPre", "BCC NRPost")

alldata <- malignant@assays$RNA@data
pd1.data <- data.frame(alldata["PDCD1",])
colnames(pd1.data) <- "gene"

ggplot(pd1.data, aes(gene)) + stat_ecdf(geom="step") + scale_x_log10()

pd1.cdf <- gene_cdf(datasets=clusters, genename="PDCD1", clusternames=clusternames, title="PD-1 Gene Expression")
pdl1.cdf <- gene_cdf(datasets=clusters, genename="CD274", clusternames=clusternames, title="PD-L1 Gene Expression")

ggarrange(pd1.cdf + scale_y_log10(), pdl1.cdf + scale_y_log10(), ncol=2, common.legend=TRUE, legend="bottom")

## PD-1 pathway expression for all PDAC cells

Idents(PDAC.all) <- PDAC.all$cluster
pdac.clusters <- levels(Idents(PDAC.all))

for (i in 1:length(pdac.clusters)) {
  assign(paste0("pdac.", pdac.clusters[i]), subset(PDAC.all, idents=pdac.clusters[i]))
  print(pdac.clusters[i])
}

pdac.clusters <- list(`pdac.Acinar cells`, `pdac.B cells`, `pdac.CD4+ T Cells`, `pdac.CD8+ T Cells`, `pdac.Ductal cells: Healthy`, `pdac.Ductal cells: Malignant 1`, `pdac.Ductal cells: Malignant 2`, `pdac.Ductal cells: Normal`, `pdac.Endothelial cells`, `pdac.Fibroblasts`, `pdac.Granulocytes`, `pdac.Macrophages`, `pdac.Mast cells`, `pdac.Myeloid cells 1`, `pdac.Myeloid cells 2`, `pdac.NK Cells`, `pdac.pDCs`, `pdac.Perivascular cells`, `pdac.Plasma cells`, `pdac.Proliferating Tregs`, `pdac.Tregs`)

pdac.clusternames <- c("Acinar", "B cells", "CD4+ T cells", "CD8+ T cells", "Ductal: Healthy", "Ductal: Mal1", "Ductal: Mal2", "Ductal: Normal", "Endothelial", "Fibroblasts", "Granulocytes", "Macrophages", "Mast", "Myeloid 1", "Myeloid 2", "NK cells", "pDCs", "Perivascular", "Plasma", "Prolif. Tregs", "Tregs")

pd1.cdf.all <- gene_cdf(datasets=pdac.clusters, genename="PDCD1", clusternames=pdac.clusternames, title="PD-1 Gene Expression")

pdl1.cdf.all <- gene_cdf(datasets=pdac.clusters, genename="CD274", clusternames=pdac.clusternames, title="PD-L1 Gene Expression")

pdl2.cdf.all <- gene_cdf(datasets=pdac.clusters, genename="PDCD1LG2", clusternames=pdac.clusternames, title="PD-L2 Gene Expression")

ggarrange(pd1.cdf.all + scale_y_log10(), pdl1.cdf.all+ scale_y_log10(), pdl2.cdf.all + scale_y_log10(), ncol=3, common.legend=TRUE, legend="bottom")

zap70.cdf.all <- gene_cdf(datasets=pdac.clusters, genename="ZAP70", clusternames=pdac.clusternames, title="ZAP70 Gene Expression") + scale_y_log10()

cd28.cdf.all <- gene_cdf(datasets=pdac.clusters, genename="CD28", clusternames=pdac.clusternames, title="CD28 Gene Expression") + scale_y_log10()

shp2.cdf.all <- gene_cdf(datasets=pdac.clusters, genename="PTPN11", clusternames=pdac.clusternames, title="SHP2 Gene Expression") + scale_y_log10()

batf.cdf.all <- gene_cdf(datasets=pdac.clusters, genename="BATF", clusternames=pdac.clusternames, title="BATF Gene Expression") + scale_y_log10()

ggarrange(zap70.cdf.all, cd28.cdf.all, shp2.cdf.all, batf.cdf.all, ncol=4, common.legend=TRUE, legend="bottom")

## PD-1 pathway expression for all BCC cells

Idents(BCCIO) <- BCCIO$seurat_clusters
bcc.clusters <- levels(Idents(BCCIO))

for (i in 1:length(bcc.clusters)) {
  assign(paste0("bcc.", bcc.clusters[i]), subset(BCCIO, idents=bcc.clusters[i]))
  print(bcc.clusters[i])
}

bcc.clusters <- list(`bcc.B cells`, `bcc.CAFs`, `bcc.CD4`, `bcc.CD8_activation`, `bcc.CD8_exhausted`, `bcc.CD8_memory`, `bcc.DCs`, `bcc.Endothelial`, `bcc.Macrophages`, `bcc.Melanocytes`, `bcc.Misc.`, `bcc.Myofibroblasts`, `bcc.NK`, `bcc.pDCs`, `bcc.Plasma`, `bcc.T_proliferation`, `bcc.Tregs`, `bcc.Tumor 1`, `bcc.Tumor 2`)

Idents(BCCIO) <- BCCIO$treatment
BCC.pre <- subset(BCCIO, idents="pre")

Idents(BCC.pre) <- BCC.pre$response
BCC.pre.R <- subset(BCC.pre, idents="Responsive")
BCC.pre.NR <- subset(BCC.pre, idents="Nonresponsive")

Idents(BCC.pre.R) <- BCC.pre.R$seurat_clusters
Idents(BCC.pre.NR) <- BCC.pre.NR$seurat_clusters

for (i in 1:length(bcc.clusters)) {
  assign(paste0("bcc.preR", bcc.clusters[i]), subset(BCC.pre.R, idents=bcc.clusters[i]))
  assign(paste0("bcc.preNR", bcc.clusters[i]), subset(BCC.pre.NR, idents=bcc.clusters[i]))
  print(bcc.clusters[i])
}

bcc.clusternames <- c("B cells", "CAFs", "CD4+ T cells", "CD8+ T cells: Activated", "CD8+ T cells: Exhausted", "CD8+ T cells: Memory", "Dendritic", "Endothelial", "Macrophages", "Melanocytes", "Miscellaneous", "Myofibroblasts", "NK cells", "pDCs", "Plasma", "Proliferating T cells", "Tregs", "Tumor 1 ", "Tumor 2")

pd1.cdf.bcc <- gene_cdf(datasets=bcc.clusters, genename="PDCD1", clusternames=bcc.clusternames, title="PD-1 Gene Expression")

pdl1.cdf.bcc <- gene_cdf(datasets=bcc.clusters, genename="CD274", clusternames=bcc.clusternames, title="PD-L1 Gene Expression")

pdl2.cdf.bcc <- gene_cdf(datasets=bcc.clusters, genename="PDCD1LG2", clusternames=bcc.clusternames, title="PD-L2 Gene Expression")

ggarrange(pd1.cdf.bcc + scale_y_log10(), pdl1.cdf.bcc+ scale_y_log10(), pdl2.cdf.bcc + scale_y_log10(), ncol=3, common.legend=TRUE, legend="bottom")

zap70.cdf.bcc <- gene_cdf(datasets=bcc.clusters, genename="ZAP70", clusternames=bcc.clusternames, title="ZAP70 Gene Expression") + scale_y_log10()

cd28.cdf.bcc <- gene_cdf(datasets=bcc.clusters, genename="CD28", clusternames=bcc.clusternames, title="CD28 Gene Expression") + scale_y_log10()

shp2.cdf.bcc <- gene_cdf(datasets=bcc.clusters, genename="PTPN11", clusternames=bcc.clusternames, title="SHP2 Gene Expression") + scale_y_log10()

batf.cdf.bcc <- gene_cdf(datasets=bcc.clusters, genename="BATF", clusternames=bcc.clusternames, title="BATF Gene Expression") + scale_y_log10()

ggarrange(zap70.cdf.bcc, cd28.cdf.bcc, shp2.cdf.bcc, batf.cdf.bcc, ncol=4, common.legend=TRUE, legend="bottom")

```



Differential gene expression between BCC and PDAC: PD-1 pathway

```{r}
## Specific gene-by-gene comparisons BCC vs. PDAC for certain celltypes

bcc.CD8 <- merge(bcc.CD8_activation, y=c(bcc.CD8_exhausted, bcc.CD8_memory))
bcc.preRCD8 <- merge(bcc.preRCD8_activation, y=c(bcc.preRCD8_exhausted, bcc.preRCD8_memory))
bcc.preNRCD8 <- merge(bcc.preNRCD8_activation, y=c(bcc.preNRCD8_exhausted, bcc.preNRCD8_memory))

pdac.allTregs <- merge(`pdac.Proliferating Tregs`, pdac.Tregs)

genes <- c("PDCD1", "CD274", "PDCD1LG2", "PTPN11", "BATF", "ZAP70", "KRAS", "HRAS", "NRAS")
genenames <- c("PD-1", "PD-L1", "PD-L2", "SHP2", "BATF", "ZAP70", "KRAS", "HRAS", "NRAS")

tcells <- list(bcc.preRCD4, bcc.preRCD8, bcc.preRTregs, bcc.preNRCD4, bcc.preNRCD8, bcc.preNRTregs, `pdac.CD4+ T Cells`, `pdac.CD8+ T Cells`, pdac.Tregs)
tcellnames <- c("R BCC CD4+", "R BCC CD8+", "R BCC Tregs", "NR BCC CD4+", "NR BCC CD8+", "NR BCC Tregs", "PDAC CD4+", "PDAC CD8+", "PDAC Tregs")

for (i in 1:length(genes)) {
  assign(paste0("tcell.cdf.", genenames[i]), gene_cdf(datasets=tcells, genename=genes[i], clusternames=tcellnames, title=paste0(genenames[i], " Gene Expression")))
  print(genenames[i])
}

ggarrange(`tcell.cdf.PD-1` + ylim(0.7, 1.00), `tcell.cdf.PD-L1` + ylim(0.875, 1.00), `tcell.cdf.PD-L2` + ylim(0.96, 1.00), ncol=3, common.legend=TRUE, legend="bottom")
ggarrange(`tcell.cdf.SHP2` + ylim(0.80, 1.00), `tcell.cdf.BATF` + ylim(0.1, 1.00), ncol=2, common.legend=TRUE, legend="bottom")
ggarrange(`tcell.cdf.KRAS` + ylim(0.65, 1.00), `tcell.cdf.HRAS` + ylim(0.88, 1.00), `tcell.cdf.NRAS` + ylim(0.82, 1.00), `tcell.cdf.ZAP70` + ylim(0.64, 1.00), nrow=2, ncol=2, common.legend=TRUE, legend="bottom")

pdac.tcells <- merge(`pdac.CD4+ T Cells`, y=c(`pdac.CD8+ T Cells`, pdac.Tregs))
bcc.r.tcells <- merge(bcc.preRCD4, y=c(bcc.preRCD8, bcc.preRTregs))
bcc.nr.tcells <- merge(bcc.preNRCD4, y=c(bcc.preNRCD8, bcc.preNRTregs))

tcells.grouped <- list(pdac.tcells, bcc.r.tcells, bcc.nr.tcells)
tcells.grouped.names <- c("PDAC", "BCC Responsive", "BCC Nonresponsive")

upgenes <- c("ZAP70", "KRAS", "HRAS", "NRAS")
upgenes.cdf <- genes_cdf(datasets=tcells.grouped, clusters=tcells.grouped.names, genes=upgenes, title="Down-Regulated Genes") + ylim(0.36, 1.00) + theme_bw()

downgenes <- c("PTPN11", "BATF")
downgenes.cdf <- genes_cdf(datasets=tcells.grouped, clusters=tcells.grouped.names, genes=downgenes, title="Up-Regulated Genes") + ylim(0.38, 1.00) + theme_bw()

ggarrange(upgenes.cdf, downgenes.cdf, ncol=2, common.legend=TRUE, legend="bottom")

comparisons <- list(c("BCC Nonresponsive", "BCC Responsive"), c("BCC Responsive", "PDAC"), c("BCC Nonresponsive", "PDAC"))

upgenes.boxplot <- ggboxplot(upgenes.cdf$data, x="cluster", y="gene", color="cluster", palette=c("#F8766D", "#00BA38", "#619CFF")) +
  stat_compare_means(comparisons = comparisons) +
  labs(y="Gene Expression Score", x="Population", title="Down-Regulated Genes") + 
  theme_bw() + theme(legend.position="none", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

downgenes.boxplot <- ggboxplot(downgenes.cdf$data, x="cluster", y="gene", color="cluster", palette=c("#F8766D", "#00BA38", "#619CFF")) +
  stat_compare_means(comparisons = comparisons) +
  labs(y="Gene Expression Score", x="Population", title="Up-Regulated Genes") + 
  theme_bw() + theme(legend.position="none", plot.title = element_text(color="black", size=18, face="bold", hjust=0.5))

upgenes.plot <- ggarrange(upgenes.boxplot, upgenes.cdf + NoLegend(), ncol=2)
downgenes.plot <- ggarrange(downgenes.boxplot, downgenes.cdf + NoLegend(), ncol=2)

annotate_figure(upgenes.plot, top=text_grob("Down-Regulated Genes", color="black", face="bold", size=18))
annotate_figure(downgenes.plot, top=text_grob("Up-Regulated Genes", color="black", face="bold", size=18))


```