---
title: "BCCIO / PDAC Comparison"
output: html_notebook
---

Load Packages and Data

```{r}
## Load packages
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(cowplot)
library(pracma)
library(RColorBrewer)
library(EnhancedVolcano)

memory.limit(size = 16000000)

## Integrated PDAC datasets 
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 155698\\PDAC155698.integrated")

## BCCIO dataset
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\GSE 123813\\BCCIOwithUMAP")

## Merged dataset (5000 cells from PDAC and BCC, random sample)
load("C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.PDAC.merge")

## Load functions
source("ClusteringAnalysisFunctions.R", echo=TRUE)
source("SeuratClusteringFunctions.R", echo=TRUE)

```



Merge PDAC and BCC datasets (subset)

```{r}

## Subset data 

BCCIO@meta.data$cell.names <- rownames(BCCIO@meta.data)
PDAC155698.integrated$cell.names <- rownames(PDAC155698.integrated@meta.data)

cells.to.sample <- 5000
set.seed(111)
BCCIO.sampled.cells <- sample(x = BCCIO@meta.data$cell.names, size=cells.to.sample, replace=F)
PDAC.sampled.cells <- sample(x = PDAC155698.integrated@meta.data$cell.names, size=cells.to.sample, replace=F)

BCCIO.subset <- subset(BCCIO, cells = BCCIO.sampled.cells)
PDAC155698.integrated.subset <- subset(PDAC155698.integrated, cells = PDAC.sampled.cells)

BCCIO.subset$cancer <- c("bcc")
PDAC155698.integrated.subset$cancer <- c("pdac")

## Merge datasets

bcc.pdac.merge <- merge(BCCIO.subset, PDAC155698.integrated.subset, add.cell.ids = c("bcc", "pdac"))
Idents(bcc.pdac.merge) <- bcc.pdac.merge@meta.data$seurat_clusters

bcc.pdac.merge@meta.data <- bcc.pdac.merge@meta.data[, -which(colnames(bcc.pdac.merge@meta.data) %in% c('cell', 'cell.names', 'sub_cluster'))]
bcc.pdac.merge$orig.clusters <- bcc.pdac.merge$seurat_clusters

## Save merged dataset

save(bcc.pdac.merge, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.PDAC.merge")

```



Enhanced Volcano Plots (PDAC vs. BCC)

```{r}

## How to read volcano plots: 
## X-axis: difference in expression (log2 fold change) of gene between ident.1 and ident.2
## Y-axis: statistical significance of difference in expression (-Log10 probability) 
## Genes of interest: located on upper corners of plot


## CD8 T cells 

bcc.pdac.merge <- RenameIdents(bcc.pdac.merge, 'CD8_memory' = 'CD8+ T cells', 'CD8_activation' = 'CD8+ T cells', 'CD8_exhausted' = 'CD8+ T cells', 'CD4' = 'CD4+ T cells')

CD8.markers <- FindMarkers(bcc.pdac.merge, subset.ident='CD8+ T cells', group.by='cancer', ident.1='bcc')
CD8.markers_tibble <- as_tibble(CD8.markers, rownames = "GeneNames")
CD8.markers_tibble <- arrange(CD8.markers_tibble, avg_logFC)

CD8.volcano <- EnhancedVolcano(CD8.markers_tibble,
                               lab = CD8.markers_tibble$GeneNames,
                               x = 'avg_logFC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'CD8+ T Cell DE: BCC vs. PDAC')
CD8.volcano

CD8.volcano.pd1 <- EnhancedVolcano(CD8.markers_tibble,
                                   lab = CD8.markers_tibble$GeneNames,
                                   x = 'avg_logFC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'CD8+ T Cell DE: BCC vs. PDAC')
CD8.volcano.pd1


## CD4 T cells

bcc.pdac.merge <- RenameIdents(bcc.pdac.merge, 'CD8_memory' = 'CD8+ T cells', 'CD8_activation' = 'CD8+ T cells', 'CD8_exhausted' = 'CD8+ T cells', 'CD4' = 'CD4+ T cells')

CD4.markers <- FindMarkers(bcc.pdac.merge, subset.ident='CD4+ T cells', group.by='cancer', ident.1='bcc')
CD4.markers_tibble <- as_tibble(CD4.markers, rownames = "GeneNames")
CD4.markers_tibble <- arrange(CD4.markers_tibble, avg_logFC)

CD4.volcano <- EnhancedVolcano(CD4.markers_tibble,
                               lab = CD4.markers_tibble$GeneNames,
                               x = 'avg_logFC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'CD4+ T Cell DE: BCC vs. PDAC')
CD4.volcano

CD4.volcano.pd1 <- EnhancedVolcano(CD4.markers_tibble,
                                   lab = CD4.markers_tibble$GeneNames,
                                   x = 'avg_logFC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'CD4+ T Cell DE: BCC vs. PDAC')
CD4.volcano.pd1


## Macrophages

macrophage.markers <- FindMarkers(bcc.pdac.merge, subset.ident='Macrophages', group.by='cancer', ident.1='bcc')
macrophage.markers_tibble <- as_tibble(macrophage.markers, rownames = "GeneNames")
macrophage.markers_tibble <- arrange(macrophage.markers_tibble, avg_logFC)

macrophage.volcano <- EnhancedVolcano(macrophage.markers_tibble,
                               lab = macrophage.markers_tibble$GeneNames,
                               x = 'avg_logFC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'Macrophage DE: BCC vs. PDAC')
macrophage.volcano

macrophage.volcano.pd1 <- EnhancedVolcano(macrophage.markers_tibble,
                                   lab = macrophage.markers_tibble$GeneNames,
                                   x = 'avg_logFC', y = 'p_val_adj',
                                   pCutoff= 0.05, FCcutoff = 0.5,
                                   selectLab = c('PDCD1', 'CD274', 'PDCD1LG2', 'TCF7', 'SIRPA', 'HAVCR2', 'LGALS9', 'CD80', 'HLA-DQA1', 'ICOSLG', 'CD40', 'PVR', 'TNFSF18', 'TBFSF4', 'CSF1', 'TIGIT', 'TNFRSF18', 'LAG3', 'CD47', 'CD70', 'CD27', 'TNFRSF4', 'CTLA4', 'CD40LG', 'ICOS', 'CD28'),
                                   title= 'Macrophage DE: BCC vs. PDAC')
macrophage.volcano.pd1

```



Enhanced Volcano Plots (PDAC)

```{r}

pdac.macrophage.markers <- FindMarkers(PDAC155698.integrated, subset.ident='Macrophages', group.by='orig.ident', ident.1='PDAC155698.1')
pdac.macrophage.markers_tibble <- as_tibble(pdac.macrophage.markers, rownames = "GeneNames")
pdac.macrophage.markers_tibble <- arrange(pdac.macrophage.markers_tibble, avg_logFC)

pdac.macrophage.volcano <- EnhancedVolcano(pdac.macrophage.markers_tibble,
                               lab = pdac.macrophage.markers_tibble$GeneNames,
                               x = 'avg_logFC', y = 'p_val_adj',
                               pCutoff = 0.05, FCcutoff = 0.5,
                               title= 'Macrophage DE: PDAC Patient 1 vs. Rest of Patients')
pdac.macrophage.volcano

```



Compare BCC Responders vs. Nonresponders

```{r}

## Label BCC dataset with responders vs. nonresponders

responders <- c("su001", "su002", "su003", "su004", "su009", "su012")
nonresponders <- c("su005", "su006", "su007", "su008", "su010")

for (patient in responders) {
  BCCIO@meta.data$response[BCCIO$patient == patient] = "Responsive"
}
for (patient in nonresponders) {
  BCCIO@meta.data$response[BCCIO$patient == patient] = "Nonresponsive"
}

## Enhanced Volcano Plots for T cells, B cells, and macrophages

Idents(BCCIO) <- BCCIO$seurat_clusters

bcc.cd4.responder.volcano <- bioc_volcano(BCCIO, c("CD4"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd4.responder.volcano

bcc.cd8_memory.responder.volcano <- bioc_volcano(BCCIO, c("CD8_memory"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_memory.responder.volcano

bcc.cd8_exhausted.responder.volcano <- bioc_volcano(BCCIO, c("CD8_exhausted"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_exhausted.responder.volcano

bcc.cd8_activation.responder.volcano <- bioc_volcano(BCCIO, c("CD8_activation"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.cd8_activation.responder.volcano

bcc.tregs.responder.volcano <- bioc_volcano(BCCIO, c("Tregs"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.tregs.responder.volcano

bcc.bcell.responder.volcano <- bioc_volcano(BCCIO, c("B cells"), c("response"), c("Responsive"), c("Nonresponsive"))
bcc.bcell.responder.volcano

```



Look specifically at BCC pre-treatment 
Goal: Determine predictors of PD1-blockade responsiveness

```{r}

## Split dataset into pre-treatment and post-treatment

Idents(BCCIO) <- BCCIO$treatment

BCCIO.pre <- subset(BCCIO, idents=c("pre"))
BCCIO.post <- subset(BCCIO, idents=c("post"))

## Enhanced Volcano plots (CD4, CD8, B, macrophage)

interestingcelltypes <- c("CD8_memory", "CD8_activation", "CD8_exhausted", "CD4", "Tregs", "T_proliferation", "B cells", "Macrophages")

Idents(BCCIO.pre) <- BCCIO.pre$seurat_clusters

for(celltype in interestingcelltypes) {
  bcc.pretreatment.responder.volcano <- bioc_volcano(BCCIO.pre, celltype, c("response"), c("Responsive"), c("Nonresponsive"))
  assign(paste0("bcc.pretreatment.", celltype, ".responder.volcano"), bcc.pretreatment.responder.volcano)
  bcc.pretreatment.responder.volcano
}

`bcc.pretreatment.B cells.responder.volcano`
## Pretty much no difference

`bcc.pretreatment.CD4.responder.volcano`
## Nothing of interest (CXCL13 higher in responders -> bad clustering?)

`bcc.pretreatment.CD8_activation.responder.volcano`
## Nothing of interest

`bcc.pretreatment.CD8_exhausted.responder.volcano`
## Lots of significant differences
## Higher amongst responders: CCL5 (associated with higher proliferation in melanoma)
## Higher amongst nonresponders: HSP group 

`bcc.pretreatment.CD8_memory.responder.volcano`
## Higher amongst responders: VIM (EMT cell marker -> leads to metastasis)
## Higher amongst nonresponders: HSP group, CCL4 (similar to CCL5)

`bcc.pretreatment.Macrophages.responder.volcano`
## Higher amongst responders: HLA genes (major role in immune regulation)

`bcc.pretreatment.Tregs.responder.volcano`
## Nothing of interest

`bcc.pretreatment.T_proliferation.responder.volcano`
## Higher amongst responders: HLA genes 

```



Conclusion?: look for higher expression of HLA genes in macrophages and T_proliferation 

Use Gene Ontology to investigate markers and gene pathways 

```{r}

## Macrophage markers

macrophage.markers <- FindMarkers(BCCIO.pre, subset.ident='Macrophages', group.by='response', ident.1='Responsive')
macrophage.markers_tibble <- as_tibble(macrophage.markers, rownames = "GeneNames")

macrophage.markers_tibble <- arrange(macrophage.markers_tibble, avg_logFC)
 
macrophage.nonresponder.markers <- macrophage.markers_tibble[macrophage.markers_tibble$p_val_adj < 0.05 &  macrophage.markers_tibble$avg_logFC < -0.5, ]

macrophage.responder.markers <- macrophage.markers_tibble[macrophage.markers_tibble$p_val_adj < 0.05 &  macrophage.markers_tibble$avg_logFC > 0.5, ]

## Proliferated T cells markers 

T_proliferation.markers <- FindMarkers(BCCIO.pre, subset.ident='T_proliferation', group.by='response', ident.1='Responsive')
T_proliferation.markers_tibble <- as_tibble(T_proliferation.markers, rownames = "GeneNames")

T_proliferation.markers_tibble <- arrange(T_proliferation.markers_tibble, avg_logFC)
 
T_proliferation.nonresponder.markers <- T_proliferation.markers_tibble[T_proliferation.markers_tibble$p_val_adj < 0.05 &  T_proliferation.markers_tibble$avg_logFC < -0.5, ]

T_proliferation.responder.markers <- T_proliferation.markers_tibble[T_proliferation.markers_tibble$p_val_adj < 0.05 &  T_proliferation.markers_tibble$avg_logFC > 0.5, ]

```



Merge BCC pretreatment data with PDAC data

```{r}

BCCIO.pre@meta.data$cell.names <- rownames(BCCIO.pre@meta.data)
PDAC155698.integrated$cell.names <- rownames(PDAC155698.integrated@meta.data)

cells.to.sample <- 5000
set.seed(111)
BCCIO.pre.sampled.cells <- sample(x = BCCIO.pre@meta.data$cell.names, size=cells.to.sample, replace=F)
PDAC.sampled.cells <- sample(x = PDAC155698.integrated@meta.data$cell.names, size=cells.to.sample, replace=F)

BCCIO.pre.subset <- subset(BCCIO.pre, cells = BCCIO.pre.sampled.cells)
PDAC155698.integrated.subset <- subset(PDAC155698.integrated, cells = PDAC.sampled.cells)

BCCIO.pre.subset$cancer <- c("bcc.pre")
PDAC155698.integrated.subset$cancer <- c("pdac")

## Merge datasets

bcc.pre.pdac.merge <- merge(BCCIO.pre.subset, PDAC155698.integrated.subset, add.cell.ids = c("bcc.pre", "pdac"))
Idents(bcc.pre.pdac.merge) <- bcc.pdac.merge@meta.data$seurat_clusters

bcc.pre.pdac.merge@meta.data <- bcc.pre.pdac.merge@meta.data[, -which(colnames(bcc.pre.pdac.merge@meta.data) %in% c('cell', 'cell.names', 'sub_cluster'))]
bcc.pre.pdac.merge$orig.clusters <- bcc.pre.pdac.merge$seurat_clusters

## Save merged dataset

save(bcc.pre.pdac.merge, file="C:\\Users\\Ryan\\Documents\\Nie Research Project\\BCC.pre.PDAC.merge")

```


Enhanced Volcano plots for BCC pretreatment (responder) vs. PDAC

```{r}

## Subset dataset to only include BCC pretreatment repsonders

Idents(bcc.pre.pdac.merge) <- bcc.pre.pdac.merge$response
bcc.pre.pdac.merge$response[is.na(bcc.pre.pdac.merge$response)] <- "PDAC"
Idents(bcc.pre.pdac.merge) <- bcc.pre.pdac.merge$response

bcc.pre.responders.pdac.merge <- subset(bcc.pre.pdac.merge, idents = c("PDAC", "Responsive"))

interestingcelltypes <- c("T_proliferation", "Macrophages")

Idents(bcc.pre.responders.pdac.merge) <- bcc.pre.responders.pdac.merge$seurat_clusters

for(celltype in interestingcelltypes) {
  bcc.pdac.responder.volcano <- bioc_volcano(bcc.pre.responders.pdac.merge, celltype, c("response"), c("Responsive"), c("PDAC"))
  assign(paste0("bcc.pdac.", celltype, ".responder.volcano"), bcc.pdac.responder.volcano)
  bcc.pdac.responder.volcano
}

## Macrophages
`bcc.pdac.Macrophages.responder.volcano`

## Proliferated T cells
`bcc.pdac.T_proliferation.responder.volcano`


```



Enhanced Volcano plots for PDAC by patient vs. BCC responders

```{r}

## Add PDAC patient column

bcc.pre.responders.pdac.merge$pdac.patient <- as.vector(bcc.pre.responders.pdac.merge$orig.ident)

bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su001.pre.tumor.cd45'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su002.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su003.pre'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su004.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su004.pre.tumor.cd45'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su009.pre.tcell'] <- 'BCC'
bcc.pre.responders.pdac.merge$pdac.patient[bcc.pre.responders.pdac.merge$pdac.patient == 'bcc.su012.pre.tcell'] <- 'BCC'

## Create separate objects for each PDAC patient

Idents(bcc.pre.responders.pdac.merge) <- bcc.pre.responders.pdac.merge$pdac.patient

bcc.pdac.1 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.1', 'BCC'))
bcc.pdac.2 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.2', 'BCC'))
bcc.pdac.3 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.3', 'BCC'))
bcc.pdac.4 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.4', 'BCC'))
bcc.pdac.5 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.5', 'BCC'))
bcc.pdac.6 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.6', 'BCC'))
bcc.pdac.7 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.7', 'BCC'))
bcc.pdac.8 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.8', 'BCC'))
bcc.pdac.9 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.9', 'BCC'))
bcc.pdac.10 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.10', 'BCC'))
bcc.pdac.11A <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.11A', 'BCC'))
bcc.pdac.11B <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.11B', 'BCC'))
bcc.pdac.12 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.12', 'BCC'))
bcc.pdac.13 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.13', 'BCC'))
bcc.pdac.15 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.15', 'BCC'))
bcc.pdac.16 <- subset(bcc.pre.responders.pdac.merge, idents = c('PDAC155698.16', 'BCC'))

Idents(bcc.pdac.1) <- bcc.pdac.1$seurat_clusters
Idents(bcc.pdac.2) <- bcc.pdac.2$seurat_clusters
Idents(bcc.pdac.3) <- bcc.pdac.3$seurat_clusters
Idents(bcc.pdac.4) <- bcc.pdac.4$seurat_clusters
Idents(bcc.pdac.5) <- bcc.pdac.5$seurat_clusters
Idents(bcc.pdac.6) <- bcc.pdac.6$seurat_clusters
Idents(bcc.pdac.7) <- bcc.pdac.7$seurat_clusters
Idents(bcc.pdac.8) <- bcc.pdac.8$seurat_clusters
Idents(bcc.pdac.9) <- bcc.pdac.9$seurat_clusters
Idents(bcc.pdac.10) <- bcc.pdac.10$seurat_clusters
Idents(bcc.pdac.11A) <- bcc.pdac.11A$seurat_clusters
Idents(bcc.pdac.11B) <- bcc.pdac.11B$seurat_clusters
Idents(bcc.pdac.12) <- bcc.pdac.12$seurat_clusters
Idents(bcc.pdac.13) <- bcc.pdac.13$seurat_clusters
Idents(bcc.pdac.15) <- bcc.pdac.15$seurat_clusters
Idents(bcc.pdac.16) <- bcc.pdac.16$seurat_clusters

## Enhanced Volcano Plots (for loop doesn't work)

bcc.pdac.1.volcano <- bioc_volcano(bcc.pdac.1, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 1"))
bcc.pdac.2.volcano <- bioc_volcano(bcc.pdac.2, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 2"))
bcc.pdac.3.volcano <- bioc_volcano(bcc.pdac.3, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 3"))
bcc.pdac.4.volcano <- bioc_volcano(bcc.pdac.4, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 4"))
bcc.pdac.5.volcano <- bioc_volcano(bcc.pdac.5, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 5"))
bcc.pdac.6.volcano <- bioc_volcano(bcc.pdac.6, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 6"))
bcc.pdac.7.volcano <- bioc_volcano(bcc.pdac.7, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 7"))
bcc.pdac.8.volcano <- bioc_volcano(bcc.pdac.8, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 8"))
bcc.pdac.9.volcano <- bioc_volcano(bcc.pdac.9, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 9"))
bcc.pdac.10.volcano <- bioc_volcano(bcc.pdac.10, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 10"))
bcc.pdac.11A.volcano <- bioc_volcano(bcc.pdac.11A, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 11A"))
bcc.pdac.11B.volcano <- bioc_volcano(bcc.pdac.11B, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 11B"))
bcc.pdac.12.volcano <- bioc_volcano(bcc.pdac.12, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 12"))
bcc.pdac.13.volcano <- bioc_volcano(bcc.pdac.13, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 13"))
bcc.pdac.15.volcano <- bioc_volcano(bcc.pdac.15, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 15"))
bcc.pdac.16.volcano <- bioc_volcano(bcc.pdac.16, c("Macrophages"), c("pdac.patient"), c("BCC"), c("PDAC Patient 16"))

```



Check DE gene expression of HLA genes by patient 

```{r}
## Differentially expressed genes

bcc.pdac.subsets <- list(bcc.pdac.1, bcc.pdac.2, bcc.pdac.3, bcc.pdac.4, bcc.pdac.5, bcc.pdac.6, bcc.pdac.7, bcc.pdac.8, bcc.pdac.9, bcc.pdac.10, bcc.pdac.11A, bcc.pdac.11B, bcc.pdac.12, bcc.pdac.13, bcc.pdac.15, bcc.pdac.16)

bcc.pdac.markers <- lapply(bcc.pdac.subsets, function(i) bioc_markers(i, c("Macrophages"), c("pdac.patient"), c("BCC")))

for (i in 1:length(bcc.pdac.markers)) {
  assign(paste0("bcc.pdac.", i, ".markers"), bcc.pdac.markers[[i]])
}

bcc.pdac.11.markers <- bcc.pdac.11A.markers
bcc.pdac.12.markers <- bcc.pdac.11B.markers
bcc.pdac.13.markers <- bcc.pdac.12.markers
bcc.pdac.14.markers <- bcc.pdac.13.markers

## Segregate the HLA genes

bcc.pdac.markers <- list(bcc.pdac.1.markers, bcc.pdac.2.markers, bcc.pdac.3.markers, bcc.pdac.4.markers, bcc.pdac.5.markers, bcc.pdac.6.markers, bcc.pdac.7.markers, bcc.pdac.8.markers, bcc.pdac.9.markers, bcc.pdac.10.markers, bcc.pdac.11A.markers, bcc.pdac.11B.markers, bcc.pdac.12.markers, bcc.pdac.13.markers, bcc.pdac.15.markers, bcc.pdac.16.markers)

patient.ids <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11A", "11B", "12", "13", "15", "16")

hla.markers <- NULL

for (i in 1:length(bcc.pdac.markers)) {
  bcc.pdac.markers[[i]]$gene.header <- substr(bcc.pdac.markers[[i]]$GeneNames, 0, 4)
  bcc.pdac.markers[[i]]$patient <- (patient.ids[i])
  hla.markers[[i]] <- subset(bcc.pdac.markers[[i]], gene.header == "HLA-")
}

for (i in 1:length(hla.markers)) {
  assign(paste0("hla.markers.", i), hla.markers[[i]])
}

## Merge datasets together

hla.markers <- list(hla.markers.1, hla.markers.2, hla.markers.3, hla.markers.4, hla.markers.5, hla.markers.6, hla.markers.7, hla.markers.8, hla.markers.9, hla.markers.10, hla.markers.11, hla.markers.12, hla.markers.13, hla.markers.14, hla.markers.15, hla.markers.16)

hla.markers.by.patient <- do.call("rbind", hla.markers)



```



Miscellaneous things

```{r}

BCCIO.pre.responder <- subset(BCCIO.pre, idents=c("Responsive"))




```